<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US" xml:lang="en-US">
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="pandoc" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
  <meta name="author" content="Benjamín J. Sánchez" />
  <meta name="author" content="Daniela C. Soto" />
  <meta name="dcterms.date" content="2021-02-15" />
  <meta name="keywords" content="markdown, publishing, manubot" />
  <title>Manuscript Title</title>
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    span.underline{text-decoration: underline;}
    div.column{display: inline-block; vertical-align: top; width: 50%;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
  </style>
  <!--[if lt IE 9]>
    <script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv-printshiv.min.js"></script>
  <![endif]-->
  <!--
  Manubot generated metadata rendered from header-includes-template.html.
  Suggest improvements at https://github.com/manubot/manubot/blob/master/manubot/process/header-includes-template.html
  -->
  <meta name="dc.format" content="text/html" />
  <meta name="dc.title" content="Manuscript Title" />
  <meta name="citation_title" content="Manuscript Title" />
  <meta property="og:title" content="Manuscript Title" />
  <meta property="twitter:title" content="Manuscript Title" />
  <meta name="dc.date" content="2021-02-15" />
  <meta name="citation_publication_date" content="2021-02-15" />
  <meta name="dc.language" content="en-US" />
  <meta name="citation_language" content="en-US" />
  <meta name="dc.relation.ispartof" content="Manubot" />
  <meta name="dc.publisher" content="Manubot" />
  <meta name="citation_journal_title" content="Manubot" />
  <meta name="citation_technical_report_institution" content="Manubot" />
  <meta name="citation_author" content="Benjamín J. Sánchez" />
  <meta name="citation_author_institution" content="Department of Bioengineering, Technical University of Denmark, Kgs. Lyngby, 2800, Denmark" />
  <meta name="citation_author_orcid" content="0000-0001-6093-4110" />
  <meta name="twitter:creator" content="@BenjaSanchez" />
  <meta name="citation_author" content="Daniela C. Soto" />
  <meta name="citation_author_institution" content="Genome Center, MIND Institute, and Department of Biochemistry &amp; Molecular Medicine, Davis, CA 95616,USA" />
  <meta name="citation_author_orcid" content="0000-0002-6292-655X" />
  <link rel="canonical" href="https://computer-aided-biotech.github.io/better-cb/" />
  <meta property="og:url" content="https://computer-aided-biotech.github.io/better-cb/" />
  <meta property="twitter:url" content="https://computer-aided-biotech.github.io/better-cb/" />
  <meta name="citation_fulltext_html_url" content="https://computer-aided-biotech.github.io/better-cb/" />
  <meta name="citation_pdf_url" content="https://computer-aided-biotech.github.io/better-cb/manuscript.pdf" />
  <link rel="alternate" type="application/pdf" href="https://computer-aided-biotech.github.io/better-cb/manuscript.pdf" />
  <link rel="alternate" type="text/html" href="https://computer-aided-biotech.github.io/better-cb/v/c0d4c2495966f15821efc542e8a26cd57a3f2abf/" />
  <meta name="manubot_html_url_versioned" content="https://computer-aided-biotech.github.io/better-cb/v/c0d4c2495966f15821efc542e8a26cd57a3f2abf/" />
  <meta name="manubot_pdf_url_versioned" content="https://computer-aided-biotech.github.io/better-cb/v/c0d4c2495966f15821efc542e8a26cd57a3f2abf/manuscript.pdf" />
  <meta property="og:type" content="article" />
  <meta property="twitter:card" content="summary_large_image" />
  <link rel="icon" type="image/png" sizes="192x192" href="https://manubot.org/favicon-192x192.png" />
  <link rel="mask-icon" href="https://manubot.org/safari-pinned-tab.svg" color="#ad1457" />
  <meta name="theme-color" content="#ad1457" />
  <!-- end Manubot generated metadata -->
</head>
<body>
<header id="title-block-header">
<h1 class="title">Manuscript Title</h1>
</header>
<p><small><em>
This manuscript
(<a href="https://computer-aided-biotech.github.io/better-cb/v/c0d4c2495966f15821efc542e8a26cd57a3f2abf/">permalink</a>)
was automatically generated
from <a href="https://github.com/computer-aided-biotech/better-cb/tree/c0d4c2495966f15821efc542e8a26cd57a3f2abf">computer-aided-biotech/better-cb@c0d4c24</a>
on February 15, 2021.
</em></small></p>
<h2 id="authors">Authors</h2>
<ul>
<li><p><strong>Benjamín J. Sánchez</strong><br>
<img src="images/orcid.svg" class="inline_icon" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-6093-4110">0000-0001-6093-4110</a>
· <img src="images/github.svg" class="inline_icon" alt="GitHub icon" />
<a href="https://github.com/BenjaSanchez">BenjaSanchez</a>
· <img src="images/twitter.svg" class="inline_icon" alt="Twitter icon" />
<a href="https://twitter.com/BenjaSanchez">BenjaSanchez</a><br>
<small>
Department of Bioengineering, Technical University of Denmark, Kgs. Lyngby, 2800, Denmark
</small></p></li>
<li><p><strong>Daniela C. Soto</strong><br>
<img src="images/orcid.svg" class="inline_icon" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-6292-655X">0000-0002-6292-655X</a>
· <img src="images/github.svg" class="inline_icon" alt="GitHub icon" />
<a href="https://github.com/dcsoto">dcsoto</a><br>
<small>
Genome Center, MIND Institute, and Department of Biochemistry &amp; Molecular Medicine, Davis, CA 95616,USA
</small></p></li>
</ul>
<h2 class="page_break_before" id="abstract">Abstract</h2>
<p>This should be the abstract.</p>
<h2 id="introduction">Introduction</h2>
<p>Since Margaret Dayhoff pioneered the field of bioinformatics back in the sixties, the application of computational tools to the field of biology has vastly grown in scope and impact. Nowadays, biotechnological and biomedical research are routinely fed by the insights arising from novel computational approaches, machine learning algorithms and mathematical models. The ever increasing amount of biological data and the exponential growth in computing power will amplify this trend in the years to come.</p>
<p>The use of computers to address biological matters encompasses a wide array of applications usually grouped under the terms of “computational biology” and “bioinformatics”. Although distinct definitions have been delineated for each one <span class="citation" data-cites="Xcr0K7I5 jn2NeK3H">[<a href="#ref-Xcr0K7I5" role="doc-biblioref">1</a>,<a href="#ref-jn2NeK3H" role="doc-biblioref">2</a>]</span>, here we will consider both under the umbrella term “computational biology”, alluding to any application that involves the intersection of computing and biological data. As such, a computational biologist can be a data analyst, a data engineer, a statistician, a mathematical modeler, a software developer, and many others. In praxis, the modern computational biologist will be a “scientist of many hats”, taking on several of the duties listed above. But first and foremost, we will consider a computational biologist as a scientist whose ultimate goal is answering a biological question or addressing a need in the life sciences, by means of computation.</p>
<p>Scientific computing requires following specific practices to enable shareable, reproducible and sustainable outputs that stand the test of time. Computing-heavy disciplines such as software engineering and data science have already adopted practices addressing the need for collaboration, visualization, project management, and strengthening of online communities. However, as a highly interdisciplinary and evolving field, computational biology has yet to acquire a set of universal “best practices”. As most computational biologists come from diverse backgrounds and oftentimes lack formal computational training, the absence of guidelines can lead to disparate and unsustainable practices that hinder reproducibility and collaboration, and slow down biomedical and biotechnological research.</p>
<p>Over the last decade, several researchers have published advice directed to bench scientists starting in either scientific computing <span class="citation" data-cites="k9JCaCug 15pOpsdFR">[<a href="#ref-k9JCaCug" role="doc-biblioref">3</a>,<a href="#ref-15pOpsdFR" role="doc-biblioref">4</a>]</span> or computational biology <span class="citation" data-cites="SGTWzj9t">[<a href="#ref-SGTWzj9t" role="doc-biblioref">5</a>]</span>. The advice encompasses a wide range of topics ranging from programming to project organization to manuscript writing. Other works have adopted a different approach, fixating in one powerful tool and diving deeper into its scope and applications, such as the software development and version control cloud service GitHub <span class="citation" data-cites="kEX5dgzK">[<a href="#ref-kEX5dgzK" role="doc-biblioref">6</a>]</span> and the web-application Jupyter Notebooks <span class="citation" data-cites="eYWjWdLd">[<a href="#ref-eYWjWdLd" role="doc-biblioref">7</a>]</span>. Similarly, recent works have chosen to comprehensively address one specific need in computational biology such as workflow automation <span class="citation" data-cites="1HhbGD4jJ">[<a href="#ref-1HhbGD4jJ" role="doc-biblioref">8</a>]</span> and software library development <span class="citation" data-cites="16iU4PKI1">[<a href="#ref-16iU4PKI1" role="doc-biblioref">9</a>]</span>. Although this advice proves immensely helpful, several aspects of the computational biology “journey” remain uncovered. Specifically, there is a lack of a clear roadmap regarding best practices from fundamental to advanced topics, in addition to practical examples tackling the complexities of this kind of research.</p>
<p>We premise that best practices in computational biology lie within a continuum that traverses three “levels”: the individual’s internal practices, the collaborative practices of a team, and the practices that allow a broader scientific community to access and engage over time with the research (Figure <a href="#fig:levels-fig">1</a>). Each one of these levels has a different set of needs and challenges. Here, we compile a selected list of relevant practices and related tools advisable for each one of these levels, emphasizing their time and place in a computational biology research project. Finally, we illustrate the utility of these practices in three case studies covering the broader spectrum of computational biology research.</p>
<div id="fig:levels-fig" class="fignos">
<figure>
<img src="images/Figure1.png" style="width:75.0%" alt="" /><figcaption><span>Figure 1:</span> The three “levels” of computational biology include your personal research, your group and collaborators, and your scientific field.</figcaption>
</figure>
</div>
<h2 id="level-1-personal-research">Level 1: Personal Research</h2>
<p>The computational biology “journey” begins with you—specifically, with the set of skills, tools and practices that you have put in place to conduct your research. Taking the time to optimally establish these building blocks will have high payoffs later, when you need to check your previous research (which will certainly happen). Consider that your most important collaborator is your future self, either of tomorrow or ten years from now. To consider all the aspects that make up a solid ground for any computational biology project, we devised a framework involving four main sequential steps (Table 1).</p>
<h3 id="step-1-choose-your-programming-languages">Step 1: Choose your programming languages</h3>
<p>Different programming languages serve distinctive purposes and have different idiosyncrasies. As such, choosing the right programming language for a project depends on your research goals and personal preference or skill. Additionally, communities usually favor the usage and training of some programming languages over others; utilizing such languages may facilitate integrating your work within the existing ecosystem.</p>
<p>As computational biology becomes a data intensive discipline, interacting with high-performance computing (HPC) operating systems has become a hallmark of the field. HPC infrastructures commonly use Unix/Linux distributions as their operating system. To interact with these platforms, you need to use a the command line interpreter known as the “Unix shell”. There are multiple versions of Unix shells, being Bash one of the most widely adopted. Besides providing an “user interface”, the shell is also a scripting language that allows manipulating files and executing programs through “shell scripts”. Unix/Linux operating systems have other interesting perks, such as powerful and fast commands for searching words and manipulating files (e.g. <code>sed</code>, <code>grep</code> or <code>join</code>) as well as the language AWK, that can perform quick text processing, including arithmetic operations.</p>
<p>One of the most common task of any computational biologist is data analysis. The job of a data analyst involves data cleaning, exploration, manipulation, and visualization. Currently, Python is the most widely used programming language for data analysis worldwide <span class="citation" data-cites="lkZcKu7R 1A4rxBI47">[<a href="#ref-lkZcKu7R" role="doc-biblioref">10</a>,<a href="#ref-1A4rxBI47" role="doc-biblioref">11</a>]</span>. Computational biology research has followed this trend, making Python one of the most popular languages among researchers. As machine learning and deep learning are more widely adopted in biological research, Python usage will keep growing. Python usage has been facilitated by the availability of packages for biological data analysis accessible through package managers such as Pip or Conda. Similarl to Python, R is the other prominent language in the field. Arguably, one of R main strengths is its wide array of tools for statistical analysis. Of particular interest is the Bioconductor repository, where many gold-standard tools for biological data analysis have been published. R usage in data science has deeply benefited from the Tidyverse packages, increasing the readability of the R syntax for both data manipulation via <code>dplyr</code>, and visualization via <code>ggplot2</code>.</p>
<p>Oftentimes, computational biologists require coding their own sets of instructions for processing data via scripts or programs. A script can be described as lightweight single-file program developed to tackle a narrow purpose. They are likely written in an interpreted programming language instead of a compiled one. Interpreted programming languages execute the program directly, without previous compilation, meaning each statement is run individually. They are quick to edit and can be run interactively, at expense of computational performance. In computational biology, the current most common multi-purpose scripting languages are Python and R. When working in a HPC, Shell/Bash scripting is also widely used. Perl and Matlab are also a popular language among bioinformatics and systems biology, respectively. A program, in the other hand, is a larger tool that usually combines multiple scripts and works as a “black box” to the user. It is designed to tackle computationally intensive problems, thus, a compiled language is preferred. Several tools designed for high-weight biological data processing have been written in C/C++. In recent years, however, scientists have been turning to Rust because of its speed, safety and friendly community <span class="citation" data-cites="wu0sr8q2">[<a href="#ref-wu0sr8q2" role="doc-biblioref">12</a>]</span>. If computational performance is not a concern, Python and R are suitable alternatives for coding programs.</p>
<p>Biological data processing is rarely a one-step process. To go from raw data to useful insights, several steps need to be taken in a specific order, accompanied by a plethora of decisions regarding parameters. Computational biologists have addressed this need by embracing workflow management systems to automate data analysis pipelines. A pipeline can be written as a Shell script where a handful of commands are written one after another, using Shell variables and Shell scripting syntax when needed. Although effective, this approach provides little control over the workflow, and lacks features to run isolated parts of the pipeline or track changes in input and output files. To overcome these limitations, a Bash script can be “upgraded” using the GNU Make program, which was originally designed to automate compilation and installation of software, but that it is flexible enough for building other types of workflows. Nowadays, however, several dedicated bioinformatics workflow managers have been developed. Snakemake is a workflow management system written in Python, allowing to incorporate the syntax of the tool with standard Python code. Similarly, Nextflow was build as an extension of the Groovy—a programming language for the Java virtual machine—and can execute Groovy code. These tools not provide control over any step of the pipeline, but also offer features like interacting with job schedulers, software environment managers, and cloud computing support. Alternatively, workflows can be written in the Common Workflow Language (CWL)—a declarative standard to define workflows with the goal of enabling portability and reproducibility. Workflows written in CWL can be run in any CWL-enabled engine.</p>
<h3 id="step-2-define-your-project-structure">Step 2: Define your project structure</h3>
<p>After choosing your programming languages and before starting any coding, we advice you to come up with a well-thought-out project structure. This design should be intentional and tailored to the present and future needs of your project—remember to be kind to your future self. A computational biology project requires, at the very least, a folder structure that supports code, data, and documentation. Although tempting, cramming all kind of files in a unique folder is unsustainable. Instead, separate each one on different folders and subfolders if needed. As additional principles consider documentation as a requirement and raw data as immutable. To simplify this process, you can base your project structure on research templates available off-the-rack. For data science projects, the python package Cookiecutter Data Science cut down the effort to the very minimum <span class="citation" data-cites="j6HfMqoR">[<a href="#ref-j6HfMqoR" role="doc-biblioref">13</a>]</span>. Running the package prompts a questionnaire in the terminal where you can input the project name, authors and other basic information. Then, the program generates a folder structure to store data—raw and processed—separated from notebooks and source code, as well as pre-made files for documentation such as a “readme”, a docs folder and a license. Similarly, the Reproducible Research Project Initialization (rr-init) offers a template folder structure that can be cloned from a GitHub repository and modified by the user <span class="citation" data-cites="7D5DgTkH">[<a href="#ref-7D5DgTkH" role="doc-biblioref">14</a>]</span>. Although the later is simpler than the former, both follow an akin philosophy aimed at research correctness and reproducibility <span class="citation" data-cites="3Gx1ESE4">[<a href="#ref-3Gx1ESE4" role="doc-biblioref">15</a>]</span>. For workflow automation projects, we advice a similar folder structure. Snakemake recommends to store each workflow in a dedicated folder separated into workflow-related files—the Snakefile, rules and scripts—results and configuration <span class="citation" data-cites="jq4WP878">[<a href="#ref-jq4WP878" role="doc-biblioref">16</a>]</span>. In all cases, the folder must be initialize as a git repo for version control (See Step 4).</p>
<p>Beyond files and folders, the software and dependencies needed to run an analysis, workflow or program, are also part of the project structure itself. The intricacies of software installation and dependency management are not to be underestimated. Fortunately, package and virtual environment managers significantly reduce this burden. A package manager is a system that automates the installation, upgrading, configuration and removing of community-developed programs; a virtual environment manager, in the other hand, is a tool that generates isolated “environments” containing programs and dependencies that are functionally independent from other environments or the default operating system. Once a virtual environment is activated, a package manager can be used to install third-party programs.</p>
<p>We believe that a computational biology project must start with its own virtual environment. The main reason is reproducibility: environments save the project’s dependencies and can restore them at will so the code can be run in any other computer. There are multiple options for both package and virtual environment management—some are language-specific; others, language-agnostic. If you are working with Python, you can initialize a Python environment using virtualenv or pipenv (where different Python versions can be installed). Inside the environment, you can use the Python package manager Pip to add Python code submitted to the Python Package Index (PyPI), GitHub or locally. For the R language, R-specific environments can be created using <code>renv</code>. After initializing the environment, packages can be installed via <code>install.packages</code> function from the Comprehensive R Archive Network (CRAN) and CRAN-like repositories. R also has BiocManager to install packages from the Bioconductor repository, which contains relevant software for high-throughput genomic sequencing analysis. To fully manage a dependencies beyond installation, RStudio developed the RStudio Package Manager which works with third-party code available in CRAN, Bioconductor, GitHub or locally. A language-agnostic alternative is Conda—an increasingly popular package manager and a virtual environment manager. It supports program installation from the Anaconda repository, which contains the channel Bioconda specifically tailored to bioinformatics software. Also, if Python is installed, Python dependencies can be installed via pip. Conda is particularly helpful when working with third-part code in all sorts of languages—a common predicament for the computational biologist. Conda package and environment manager is included in both Miniconda and Anaconda distributions. The former is a minimal version of Anaconda, containing only Conda, Python, and a few useful packages.</p>
<h3 id="step-3-choose-your-working-set-up">Step 3: Choose your working set-up</h3>
<p>With the foundation in place, the next step is to start coding. However, a more practical question needs to be answer first: where to code. The simplest tool available for this purpose are text editors. Since writing code is ultimately writing text, any tool where characters can be typed fulfills this purpose. However, coding can be streamlined by additional features as those found in code editors—text editors especially developed for writing code. Crucial features to facilitate coding include syntax highlight, indentation or autocompletion. Commonly used desktop editors include Atom, Sublime, Visual Studio Code, and Notepad++ (Windows only), all which a myriad of plugins available to enhance the coding experience. Command line text editors are also suitable options for coding, being Vim and Emacs the most powerful ones. All of these tools share the advantage of being language agnostic, allowing easy switching between languages, especially handy for the polyglot computational biologist.</p>
<p>In addition to text editors, integrated development environment (IDE) are also popular options for coding. As its essence, IDE are supercharged text editors, i.e. with multiple other features that make writing code easier. The main parts of an IDE are a code editor (with syntax highlight, indentation and suggestions), a debugger, a folder structure, and a way to execute your code (a compiler or interpreter). IDEs are not language agnostic, meaning that they allow to code in one language. The array of features also comes at a cost—IDEs usually use more memory and imply more visual clutter, if that is a concern of yours. For Python, Jupyter Lab, Spyder and JetBrains’ PyCharm are popular options, while for R, RStudio is the gold-standard. Notably, the differences between an IDE and a code editor are somewhat blurry, especially when enough plugins have been added to a code editor.</p>
<p>In the latest years, notebooks have acquired relevance in computational biology research. A notebook is an interactive application that combines live code (read-print-eval loop or REPL), narrative, equations and visualizations. Common notebooks use an interpreted languages such as Python or R, and narrative follows markdown syntax. Data analysis greatly benefits from using notebooks instead of plain text editors or even IDEs: the combination of visuals and texts allows researcher to tell compelling stories about their data, and the interactivity of its code enables quick testing of different strategies. Jupyter notebook is a popular web-based interactive notebook developed originally for Python coding, but also accepts R and other programming languages upon installation of their kernels—a computing engine that executes the notebook’s live code “under the hood”. Jupyter notebook can also be run in the cloud using platforms such as Google CoLab and Amazon WebServices, taking advantage of the current trend of cloud computing. RStudio also allows the generation of R-based notebooks known as R Markdown, which is specially well-suited for generating reports.</p>
<h3 id="step-4-follow-good-coding-practices">Step 4: Follow good coding practices</h3>
<p>After dealing with steps one to three, finally comes writing code. Coding, however, requires good practices to ensure correctness, sustainability and reproducibility for you, your future self, your collaborators (as we will discuss in Level 2) and the whole community (as we will discuss in Level 3). First and foremost, you need to make sure your code works correctly. In computational biology, correctness implies biological and statistical soundness. Both are big topics beyond the scope of this manuscript. To achieve the former, however, a useful approach is to design positive and negative controls in your program, analysis or workflow. In an experiment, a positive control is a control group that is expected to produce results; a negative control is expected to produce no results. The same approach can be applied to computation, using input data whose output is previously known. Biological soundness can also be tested by quickly assessing expected orders of magnitude in both, intermediate and final files.</p>
<p>Beyond the correct functioning of your code, you will need to pay attention to the way your code looks, also know as “coding style”. This includes a series of small and ubiquitous decisions regarding where and how to add comments; indentation and white spaces usage; variable, function and class naming; and overall code organization. It is true that, as in writing, there is a lot of your own personality in the way you code. However, sticking to existing coding style rules facilitates collaborations with your future self and others. Indeed, as we sometimes have trouble reading our own handwriting, we can also struggle reading our own code if we overlook any guidelines. At the very least, your code must display internal consistency. Even better, you can follow any of the multiple coding style rules that have been published. Although arbitrary, most of these rules have been developed with readability in mind. A good place to start, however, are style guides from software development teams. Google, for example, has published guidelines for Python, R, Shell, C++, and HTML/CSS <span class="citation" data-cites="4PJ4Kllx">[<a href="#ref-4PJ4Kllx" role="doc-biblioref">17</a>]</span>. Also, a series of guidelines for the Python programming language have been published with the name of Python Enhancement Proposal (PEP), where the most widely adopted is PEP-8 <span class="citation" data-cites="1D76el83m">[<a href="#ref-1D76el83m" role="doc-biblioref">18</a>]</span>. To aid flagging stylistic errors in your code, tools called “linters” are usually included with code editors and IDEs or provided as plugins.</p>
<p>In the matter of code styling, two topics merit additional attention: variable naming and comments. Variable names should be descriptive enough to convey an idea about the variable, function or class’ content and use. The goal is to produce “self-documented” code that reads close to plain English. To do so, use multi-words variable names if necessary. In such cases, the most common conventions include Camel Case, where the second and subsequents words are capitalized (“camelCase”); Pascal Case, where all words are capitalized (“PascalCase”); and Snake Case, where words are separated by underscores (“snake_case”). Notably, all these conventions can be used in a same coding style to differentiate variables, functions and classes. For example, PEP-8 recommends Snake Case for functions and variables, and Pascal Case for class names. In addition to master variable naming, code comments—explanatory human-readable statements not evaluated by the program—are necessary to enhance the code’s readability. No matter how beautiful and well-organized your code is, high-level code decisions will not be obvious unless stated. As a corollary, code explanations that can be deduced from the syntax itself should be omitted. Comments can span a single line or several ones, forming a block, and can be found in three strategic parts: at the top of the program file (“header comment”), which contains the author and the date of the code and what it accomplishes; above every function (“function header”), which contains the purpose and behavior of the function; and in line, next to tricky code whose behavior is not obvious or warrant a remark.</p>
<p>When working with a sizable code base, a good practice related is to strive for modularity—splitting your code’s functionalities into independent entities known as modules. Modularity enhances code readability and reusability—enabling your code to be used by you or others in future applications—and expedites maintenance. In Python, subdivisions are defined as follow: a module is a collection of functions and global variables, a package a collection of modules, a library a collection of packages, and a framework a collection of libraries. Modules are simply files with the <code>.py</code> extension. Packages, in the other hand, must be indicated to the Python interpreter adding a file named <code>__init__.py</code> (which could be empty or not).</p>
<p>Code styling rules also apply to data science notebooks. However, when writing notebooks you must also engage in “literate programming”—a programming paradigm where the code is accompanied by human-readable explanation of its logic and purpose. In other words, notebooks must tell a story about the analysis, connecting the dots between the code, the results and the figures. Human-readable language is often written in Markdown—a lightweight markup language. Little has been written about good practices for literate programming, but we advice you to explain the purpose of each chunk of code and provide some interpretation of its results.</p>
<p>Equally as important as writing good code is to use version control—the practice of tracking and managing changes in your code. This is a good and necessary practice even if your only collaborator is your future self. One of the main advantages of version control is keeping a change log of your files that can be utilized to go back to previous versions of your code, and remind you of previous approaches disregarded in newer versions. The most widely used version control system, Git, achieves these tasks with the command <code>git checkout</code>. Additionally, version control also allows you to safely try new functionalities using “branches”—carbon copies of the main original branch (known as “master”) where you can add code independently and optionally merge it to the original one. Git creates branches using the command <code>git branch</code> and the same <code>git checkout</code> can be used to switch among them. We will discuss the utility and implementation of branches in collaborative projects in the next section (See Level 2: Collaboration). Nowadays, there are multiple code repositories that also provide version control with Git, such as GitHub, GitLab or Bitbucket. They have the additional benefit of backing up your code and code history in the cloud, keeping your work safe and shareable. You can interact with these platforms using the browser or via graphic user interfaces (GUI) such as GitHub Desktop or GitKraken.</p>
<p><strong>Table 1</strong>: Steps to start any computational biology project.</p>
<table>
<thead>
<tr>
<th>
Step
</th>
<th>
Use case
</th>
<th>
Common tools
</th>
</tr>
</thead>
<tbody>
<tr>
<td rowspan="5">
<b>Step 1:</b> Choose your programming languages
</td>
<td rowspan="1">
Interacting with a Unix/Linux HPC
</td>
<td>
• Shell/Bash
</td>
</tr>
<tr>
<td rowspan="1">
Data analysis
</td>
<td>
• Python, R
</td>
</tr>
<tr>
<td rowspan="1">
Scripts and programs:
</td>
<td>
• Interpreted: Python, R, Perl<br />• Compiled: C/C++, Rust
</td>
</tr>
<tr>
<td rowspan="1">
Workflows
</td>
<td>
• Snakemake (Python), Nextflow (Groovy), CWL
</td>
</tr>
</tbody>
<tbody>
<tr>
<td rowspan="5">
<b>Step 2:</b> Define your project structure
</td>
<td rowspan="1">
Project structure
</td>
<td>
• Templates: Cookiecutter Data Science, rr-init<br />• Workflows: Snakemake structure
</td>
</tr>
<tr>
<td rowspan="1">
Virtual environment managers
</td>
<td>
• Language-specific: pipenv (Python), virtualenv (Python), renv (R) <br />• Language agnostic: Conda
</td>
</tr>
<tr>
<td rowspan="1">
Package managers
</td>
<td>
• Language-specific: pip (Python),BiocManager (R), R Studio package manager (R)<br />• Language-agnostic: Conda
</td>
</tr>
</tbody>
<tbody>
<tr>
<td rowspan="5">
<b>Step 3:</b> Choosing your working set-up
</td>
<td rowspan="1">
Text editors
</td>
<td>
• Desktop applications: Atom, Sublime, Visual Studio Code, Notepad++<br />• Command line: Vim, Emacs
</td>
</tr>
<tr>
<td rowspan="1">
IDEs
</td>
<td>
• For Python: Jupyter Lab, JetBrains/PyCharm, Spyder<br />• For R: R Studio
</td>
</tr>
<tr>
<td rowspan="1">
Notebooks
</td>
<td>
• Jupyter (Python, R), R Markdown (R)
</td>
</tr>
</tbody>
<tbody>
<tr>
<td rowspan="5">
<b>Step 4:</b> Follow good coding practices
</td>
<td rowspan="1">
Coding style
</td>
<td>
• Styling guides: PEP-8 (Python), Google (Python, R) <br />• Linters
</td>
</tr>
<tr>
<td rowspan="1">
Literate programming
</td>
<td>
• Markdown
</td>
</tr>
<tr>
<td rowspan="1">
Version control
</td>
<td>
• Version control system: Git<br />• Code repositories: GitHub, GitLab, Bitbucket<br />• Git GUIs: GitHub Desktop, GitKraken
</td>
</tr>
</tbody>
</table>
<h2 id="level-2-collaboration">Level 2: Collaboration</h2>
<p>Collaboration is a key aspect of scientific research, but it is especially relevant in computational biology, where interdisciplinary knowledge is often needed. Collaborators can take different forms: your boss or advisor, colleagues or lab mates, other laboratories, people from academia or industry, or your future self (as discussed in Level 1). Although collaborators can have a wide range of involvement with your project—from co-authors to commenters—they all share a direct relationship with you and your research, comprising a group of a dozen of people at most (contrary to a community, which is an open group of a large number of people, as we will discuss in Level 3). Each type of collaboration requires its own set of good practices, which we will cover in the next paragraphs.</p>
<h3 id="share-code">2.1 Share code</h3>
<p>Sharing code is one of the most common practices in software development, where large teams work together developing highly complex functions and scripts. Although computational biology projects are usually not as big, proper ways of sharing code are still essential, as it is not desired to have file conflicts as soon as two different researchers change the same piece of code. As mentioned in the previous section, hosting services such as GitHub <span class="citation" data-cites="dmEZPbB4">[<a href="#ref-dmEZPbB4" role="doc-biblioref">19</a>]</span>, GitLab <span class="citation" data-cites="16Oz9VFEk">[<a href="#ref-16Oz9VFEk" role="doc-biblioref">20</a>]</span> and Bitbucket <span class="citation" data-cites="1E58VA9bS">[<a href="#ref-1E58VA9bS" role="doc-biblioref">21</a>]</span> (Table ??) allow for having a Git repository stored online, by creating a copy of the repository known as the “remote”, which becomes the official version of the repository. The key advantage of using a remote is that there will be no direct interaction between different local copies of the repository, also known as a “clones”, but instead each clone only will interact with the remote, and can only update the remote if there are no conflicts. This way, if a collaborator updated the repository with some changes, another collaborator will not be able to send their changes until they make sure to update their local copy with the changes already present online.</p>
<p>Tools for collaborative research. {#tbl:collaboration-tools}</p>
<table>
<colgroup>
<col style="width: 23%" />
<col style="width: 76%" />
</colgroup>
<thead>
<tr class="header">
<th>Goal</th>
<th>Tools</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Share code</td>
<td>• <em>Hosting services</em>: <strong>GitHub</strong> <span class="citation" data-cites="dmEZPbB4">[<a href="#ref-dmEZPbB4" role="doc-biblioref">19</a>]</span>, <strong>GitLab</strong> <span class="citation" data-cites="16Oz9VFEk">[<a href="#ref-16Oz9VFEk" role="doc-biblioref">20</a>]</span>, <strong>Bitbucket</strong> <span class="citation" data-cites="1E58VA9bS">[<a href="#ref-1E58VA9bS" role="doc-biblioref">21</a>]</span>. <br />• <em>Git branching strategies:</em> <strong>Github flow</strong> <span class="citation" data-cites="qjg93S3G">[<a href="#ref-qjg93S3G" role="doc-biblioref">22</a>]</span>.<br />• <em>Tests:</em> correctness (e.g. <strong>pytest</strong> <span class="citation" data-cites="15kUlshEi">[<a href="#ref-15kUlshEi" role="doc-biblioref">23</a>]</span>), coverage (e.g. <strong>codecov</strong> <span class="citation" data-cites="cEpTMlZc">[<a href="#ref-cEpTMlZc" role="doc-biblioref">24</a>]</span>), automation (e.g. <strong>tox</strong> <span class="citation" data-cites="1Bi82XR9D">[<a href="#ref-1Bi82XR9D" role="doc-biblioref">25</a>]</span>, <strong>Travis CI</strong> <span class="citation" data-cites="SM1pqdjN">[<a href="#ref-SM1pqdjN" role="doc-biblioref">26</a>]</span>, <strong>Github Actions</strong> <span class="citation" data-cites="NUXbp429">[<a href="#ref-NUXbp429" role="doc-biblioref">27</a>]</span>).<br />• <em>Code reviews</em>: <strong>Github</strong> <span class="citation" data-cites="RCh2gb01">[<a href="#ref-RCh2gb01" role="doc-biblioref">28</a>]</span>, <strong>Crucible</strong> <span class="citation" data-cites="6woNTKZi">[<a href="#ref-6woNTKZi" role="doc-biblioref">29</a>]</span>, <strong>Upsource</strong> <span class="citation" data-cites="LJc9aT8R">[<a href="#ref-LJc9aT8R" role="doc-biblioref">30</a>]</span>.</td>
</tr>
<tr class="even">
<td>Share data</td>
<td>• <em>FAIR principles</em> <span class="citation" data-cites="6DjakjNS">[<a href="#ref-6DjakjNS" role="doc-biblioref">31</a>]</span>. <br />• <em>Tidy data</em> <span class="citation" data-cites="Gtnr8c2U">[<a href="#ref-Gtnr8c2U" role="doc-biblioref">32</a>]</span>. <br />• <em>Data version control</em> <span class="citation" data-cites="JkrlcwLO">[<a href="#ref-JkrlcwLO" role="doc-biblioref">33</a>]</span>.</td>
</tr>
<tr class="odd">
<td>Share data science notebooks</td>
<td>• <em>Static:</em> <strong>GitHub</strong>, <strong>GitLab</strong>, <strong>NBviewer</strong> <span class="citation" data-cites="8Cn0lgXi">[<a href="#ref-8Cn0lgXi" role="doc-biblioref">34</a>]</span>. <br />• <em>Interactive:</em> <strong>Binder</strong> <span class="citation" data-cites="1dqfprG7">[<a href="#ref-1dqfprG7" role="doc-biblioref">35</a>]</span>, <strong>Google CoLab</strong> <span class="citation" data-cites="mr26BIpF">[<a href="#ref-mr26BIpF" role="doc-biblioref">36</a>]</span>.<br />• <em>Comparative:</em> <strong>ReviewNB</strong> <span class="citation" data-cites="12ovVgVju">[<a href="#ref-12ovVgVju" role="doc-biblioref">37</a>]</span>.</td>
</tr>
<tr class="even">
<td>Share workflows</td>
<td>• <em>General hosting services</em>: <strong>GitHub</strong>, <strong>GitLab</strong>, <strong>Bitbucket</strong><br />• <em>Dedicated workflow repositories:</em> <strong>WorkflowHub</strong> <span class="citation" data-cites="6YE0Zqqy">[<a href="#ref-6YE0Zqqy" role="doc-biblioref">38</a>]</span></td>
</tr>
<tr class="odd">
<td>Share manuscripts</td>
<td>• <em>General-purpose word processors:</em> <strong>Google Docs</strong> <span class="citation" data-cites="18tOsSTai">[<a href="#ref-18tOsSTai" role="doc-biblioref">39</a>]</span>, <strong>Office 365</strong> <span class="citation" data-cites="sP8f9fNq">[<a href="#ref-sP8f9fNq" role="doc-biblioref">40</a>]</span>.<br />• <em>Scholarly word processors:</em> <strong>Authorea</strong> <span class="citation" data-cites="UeFHiqcp">[<a href="#ref-UeFHiqcp" role="doc-biblioref">41</a>]</span>.<br />• <em>Online applications supporting Markup Languages:</em> <strong>Overleaf</strong> (LaTeX) <span class="citation" data-cites="J12A3elW">[<a href="#ref-J12A3elW" role="doc-biblioref">42</a>]</span>, <strong>Manubot</strong> (Markdown + GitHub) <span class="citation" data-cites="petulEc2">[<a href="#ref-petulEc2" role="doc-biblioref">43</a>]</span>.</td>
</tr>
</tbody>
</table>
<p>In order to guarantee that different collaborators can work simultaneously in the same repository, a good idea is to implement some type of branching strategy in the repository (Table ??). In a small team of collaborators, the most common strategy is to have a single <code>master</code> branch and generate from it branches that each different developer can work on. Then, whenever the developer is ready, they can request to combine, or “merge”, the changes from their branch into the master branch, in a process known as “pull request”, or PR for short. Once a PR has been opened, collaborators can review it, and if it fulfills their criteria, approve it so that it can be merged into the master branch: any succeeding branches will now have those commits already included as part of their history. This branching strategy is sometimes referred to as Github flow <span class="citation" data-cites="qjg93S3G">[<a href="#ref-qjg93S3G" role="doc-biblioref">22</a>]</span> and will suffice for most projects. For more complex branching systems, see Section <a href="#make-your-research-sustainable">3.3: Make your research sustainable</a>.</p>
<p>Using Git hosting services for collaboration has many additional benefits. The commit history not only shows what was done at each point in time, but also which collaborator did it, so that if e.g. a bug was introduced, commands such as <code>git blame</code> will show which collaborator caused it. Collaborators can also create “forks”, i.e. full copies of repositories under their own possession, for e.g. having different a version of the software that works for a different purpose. Git hosting services can be accessed interactively online, or from the terminal with tools such as GitHub CLI <span class="citation" data-cites="uRbOUixg">[<a href="#ref-uRbOUixg" role="doc-biblioref">44</a>]</span>. Finally, Git hosting services also allow collaborators to open issues <span class="citation" data-cites="jIIrCDT3">[<a href="#ref-jIIrCDT3" role="doc-biblioref">45</a>]</span> for listing pending to-do’s and/or asking questions, acting as an open forum for development discussions, which has the advantage of remaining accessible for the future, if new collaborators would join the research project later (as opposed to closed e-mail discussions). We will discuss additional advantages of using Git hosting services, in terms of interacting with a user base, in <a href="#level-3-community">Level 3: Community</a>.</p>
<p>Another important concept to internalize when developing code, especially together with other collaborators, is to develop unit tests (Table ??). Unit tests are scripts that will run to determine if specific modules/functions work as intended within the codebase, so that if later the function grows in scope, its proper basic functioning is ensured. For instance, if a function was defined for adding numbers, a simple test would be to asses if the function outputs 13 when the inputs 6 and 7 are provided. Tools such as <code>pytest</code> <span class="citation" data-cites="15kUlshEi">[<a href="#ref-15kUlshEi" role="doc-biblioref">23</a>]</span> for Python and <code>testthat</code> <span class="citation" data-cites="E7mOdOEw">[<a href="#ref-E7mOdOEw" role="doc-biblioref">46</a>]</span> for R exist to then detect said scripts, and run all of them to display if any specific section is failing. It is a very good practice to develop tests at the same time you develop code (at the personal research level), as adding tests <em>a posteriori</em> is significantly harder (but sometimes inevitable). Going beyond testing correctness, tools such as <code>flake8</code> <span class="citation" data-cites="ZNYSO1m0">[<a href="#ref-ZNYSO1m0" role="doc-biblioref">47</a>]</span> will test styling preferences (for complying with PEP8), <code>safety</code> <span class="citation" data-cites="iFiB8Hir">[<a href="#ref-iFiB8Hir" role="doc-biblioref">48</a>]</span> will test for vulnerabilities among the software’s dependencies, and Codecov <span class="citation" data-cites="cEpTMlZc">[<a href="#ref-cEpTMlZc" role="doc-biblioref">24</a>]</span> will test which percentage of the codebase is tested, given that as a rule of thumb, the more code lines tested, the more reliable a software is. All these different types of tests can be funneled into a single testing pipeline that can run automatically whenever desired. This process is known as Continuous Integration (CI), and can be tuned to run locally whenever commits are made, or online whenever a pull request is opened and/or merged. When running locally, an environment manager / command line tool such as <code>tox</code> <span class="citation" data-cites="1Bi82XR9D">[<a href="#ref-1Bi82XR9D" role="doc-biblioref">25</a>]</span> helps to ensure all tests are ran under e.g. different python versions. For setting up the CI cycle online, different dedicated CI tools such as Travis CI <span class="citation" data-cites="SM1pqdjN">[<a href="#ref-SM1pqdjN" role="doc-biblioref">26</a>]</span> or Circle CI <span class="citation" data-cites="T03Api6e">[<a href="#ref-T03Api6e" role="doc-biblioref">49</a>]</span> exist, and more recently GitHub actions <span class="citation" data-cites="NUXbp429">[<a href="#ref-NUXbp429" role="doc-biblioref">27</a>]</span> has been introduced for running the integration directly from Github.</p>
<p>Having tests is a great way of ensuring code fulfills a certain level of correctness and styling. However, it is no replacement for a human assessment to see if the code is correct, necessary and useful. Therefore, code reviewing is essential whenever developing code in collaboration (Table ??). Tools such as Crucible <span class="citation" data-cites="6woNTKZi">[<a href="#ref-6woNTKZi" role="doc-biblioref">29</a>]</span> and Upsource <span class="citation" data-cites="LJc9aT8R">[<a href="#ref-LJc9aT8R" role="doc-biblioref">30</a>]</span> exist for making in-line reviews of each file, but the most common approach, if the software is stored in a repository, is to directly review using the online review tools from the hosting service. In the case of Github <span class="citation" data-cites="RCh2gb01">[<a href="#ref-RCh2gb01" role="doc-biblioref">28</a>]</span>, this not only allows the reviewer to open a comment in any line of the code (which creates a thread for the original author to reply in), but also to suggest changes, for the author to in turn approve/dismiss. When reviewing, there are a series of things to look for (from functionality to documentation), and good practices to keep in mind (such as phrasing the comments in a constructive way), which are outside of the scope of this review but presented in detail elsewhere <span class="citation" data-cites="FTO2RnTd HnmFW2rl">[<a href="#ref-FTO2RnTd" role="doc-biblioref">50</a>,<a href="#ref-HnmFW2rl" role="doc-biblioref">51</a>]</span>.</p>
<h3 id="share-data">2.2 Share data</h3>
<p>The practices of sharing data stem from the same place as with sharing code: we should store our dataset and any changes to it in a repository, and ensure it complies with standards by testing its quality. However, due to data having a more consistent structure than code, as data is often processed and outputted by machines in standard formats, there are additional criteria that should be considered when we share it with collaborators (and later on with the community). The main set of guidelines that represent these criteria were outlined a few years ago in what is known as the FAIR principles <span class="citation" data-cites="6DjakjNS">[<a href="#ref-6DjakjNS" role="doc-biblioref">31</a>]</span>: data should be Findable, i.e. easy to find online; Accessible, i.e. easy to access once found; Interoperable, i.e. easy to integrate with other data/applications/workflows/etc.; and Reusable, i.e. presented in a way that allows for others to use it for the same purpose or different settings.</p>
<p>For making data findable, research repositories such as Zenodo <span class="citation" data-cites="Z6ni32OG">[<a href="#ref-Z6ni32OG" role="doc-biblioref">52</a>]</span> and Figshare <span class="citation" data-cites="1EMndz0j9">[<a href="#ref-1EMndz0j9" role="doc-biblioref">53</a>]</span> allow you to assign a digital object identifier (DOI) to any group of files you upload, including data and/or code. Alternatively, regular code repositories like Github can be used instead, as you can use specific commits and/or releases to identify specific versions of the data (see <a href="#make-your-research-accessible">3.1. Make your research accessible</a>), in combination with extensions for Large File Storage such as git LFS <span class="citation" data-cites="11GtZ7icJ">[<a href="#ref-11GtZ7icJ" role="doc-biblioref">54</a>]</span>, in the case of data files larger than 100 MB <span class="citation" data-cites="varWEBV8">[<a href="#ref-varWEBV8" role="doc-biblioref">55</a>]</span>. A final alternative is the Data Version Control (DVC) initiative <span class="citation" data-cites="JkrlcwLO">[<a href="#ref-JkrlcwLO" role="doc-biblioref">33</a>]</span>, which is especially useful when doing machine learning, as it can keep track of data, machine learning models and even scoring metrics.</p>
<p>For making data accessible, we encourage as much as possible to make your repositories open access so that everything is accessible to everyone, but in cases in which you or your collaborators prefer some restrictions, you can create guest accounts to provide access to private repositories. For making data interoperable, distinctions between raw and clean data have been made <span class="citation" data-cites="10.1371/journal.pcbi.1000424">[<span class="citeproc-not-found" data-reference-id="10.1371/journal.pcbi.1000424"><strong>???</strong></span>]</span>, with the raw data being the same files that came out of the measuring device, and the clean data the files that are ready to be used for any computational analysis. An important characteristic that clean data should have is to be “tidy”, which is reviewed in detail elsewhere <span class="citation" data-cites="Gtnr8c2U">[<a href="#ref-Gtnr8c2U" role="doc-biblioref">32</a>]</span>. Finally, for making data reusable, thorough documentation of the data - including experimental design, measurements units and sources of error - is required.</p>
<h3 id="share-data-science-notebooks">2.3 Share data science notebooks</h3>
<p>As we previously discussed, Jupyter Notebook have become a fundamental tool of data analytics. Accordingly, you will likely need to share your notebook with collaborators at some point. To do this, there are static and interactive options. The former, as the name indicates, share computational notebooks as a rendered text, written internally in HTML. Static notebooks are a good option when you want to avoid any modifications and can work as an archive of past analyses, but interacting with its content is cumbersome—the file must be downloaded and run in a local Jupyter installation. Git-based code repositories, such as GitHub and GitLab, automatically render notebooks that can be later shared pointing collaborators to the GitHub repository. To ease this process, the Project Jupyter provides a web application called NBviewer, where you can paste a Jupyter Notebook’s URL, publicly hosted in GitHub or elsewhere, and renders the file into a static HTML web page with a stable link.</p>
<p>Interactive notebooks, in the other hand, not only render the file but also allow collaborators to fully interact with it, tinkering parameters or trying new input data—no installation required. The Binder Projects (which is also part of the Project Jupyter) offers the Binder service, where any publicly hosted Git-based repository can be open with a Jupyter Notebook interface. The user can fully interact with any notebook within the repository, although changes will not be saved to the original file. The platforms supports Python and R among other languages, and any additional packages required to run the analysis need to be specified in a configuration file within the repository. Similarly, Jupyter Notebooks can be run interactively using Google CoLaboratory (CoLab), which is available to anyone with a Google account. Notebooks can be updated locally, from any public GitHub repository, or from Google Drive, where also imported files are also saved. In both cases, the machines provided by these services are comparable to a modern laptop. Thus, these tools are not suitable for some computing-intensive tasks common to computational biology problems.</p>
<p>Besides sharing notebooks, oftentimes computational biologists need to work and edit a notebook together. In those cases, notebooks need to be treated as any other piece of code: updates from different collaborators must be managed with version control in a platform such as GitHub. The problem, however, is that Git-based hosting services deal with notebooks as if they were HTML text, where changes between versions are hard to visualize. To better compare these changes, there is NBreview, which renders and display in parallel the old and new versions of a notebook for easy comparison. The tool can be easily installed using your GitHub account, and notebooks can be reviewed from their website.</p>
<h3 id="share-computational-workflows">2.4 Share computational workflows</h3>
<p>Computational biology projects often demands sharing multi-step analyses with dozens of third-party software and dependencies. Although these steps can be shared as documentation, complex workflows are better shared as stand-alone code that can be easily run with minimal file manipulation from collaborators. Doing so can safeguard the reproducibility and replicability of the analysis, leading to better science and less issues down the road.</p>
<p>The simplest way to share a pipeline is to generate a Bash script that receives input files from the command line, thus, allowing to run it with different input data. However, Bash scripts offer little control over the overall workflow and cannot re-run specific parts of the pipeline. To address these issues, pipelines are better shared using a workflow automation system. Theoretically, all the instructions regarding the workflow could be written in the main pipeline file—in Snakemake, the <code>.smk</code> file (or Snakefile); in Nextflow, te <code>.nf</code> file; and in CWL, <code>.cwl</code> file. However, to ensure reproducibility, it is a good practice to share complete pipelines, meaning folder structures, additional files and software specification, as well all custom scripts developed for the analysis. These files can be shared using the same tools as other forms of code, namely GitHub or any other Git hosting services. Alternatively, they can be uploaded to hosting services specialized in workflows, like WorkflowHub <span class="citation" data-cites="6YE0Zqqy">[<a href="#ref-6YE0Zqqy" role="doc-biblioref">38</a>]</span>, currently in beta.</p>
<p>When sharing workflows, consider that sharing software versioning is necessary for your collaborators to reproduce your pipeline using their own computing setup. Conda environments, for example, can be easily created from an environment file (in YAML language), which can be exported from an existing environment. Notably, Snakemake and Nexflow can be configured to automatically build isolated environments for each rule or step, enabling running different versions of a program within the same pipeline (which is especially helpful when needing both Python 2 and 3). Besides sharing the specifications of an environment, it is possible to share the environment itself via containers, using platforms like Docker and Singularity, which are especially helpful to share environments with a broader community, as we will discuss in Level 3.</p>
<h3 id="write-manuscripts-collaboratively">2.5 Write manuscripts collaboratively</h3>
<p>Writing articles is arguably the main way where we can share our research with the scientific community and for that purpose, the world. However, in a highly interdisciplinary field as computational biology, writing manuscripts is also a collaborative effort, where multiple people is directly involved in the crafting of a manuscripts. The traditional computer tools for writing documents, therefore, are oftentimes not suitable for this type of collaboration, resulting in files with different names jumping from one e-mail inbox to another, resulting in multiple and likely contradictory final versions. Let’s avoid this by streamline collaborative manuscript writing with tools made for that purpose.</p>
<p>Big companies have become aware of the need for collaborative writing, developing online applications that can be simultaneously edited by multiple people. Google’s GDocs and Microsoft’s Office 365 are well-known word processors designed for this purpose, where the text is displayed with the exact appearance than in a printout (know as <em>What-You-See-Is-What-You-Get</em> , or WYSIWYG) and the text can be formatted making use of the internal features of the application. The advantage of these technologies is that they are extremely user-friendly, and require no additional knowledge. They are a good option when one or more of your collaborators seeks simplicity, but they are not specifically tailored for the needs of scientific writing, such as adding references, equations and figures. Fortunately, third-party companies have developed plugins for these applications to add references to your document. Companies like Authorea have developed their own online application specifically designed for writing manuscripts. Authorea, in particular, offers templates for different type of research projects, allows you to manage collaborators from the same platform, and easily add reference from using identifiers (DOI, PubMed, etc.). Consider, nonetheless, that some collaborators may not want to adopt a new tool exclusively for writing manuscripts.</p>
<p>In addition to word editors, text editors are a competitive option to write manuscripts when combined with a markup language—a human-readable computer language that uses tags to delineate formatting elements in a document that will be later rendered. Since the formatting process is internally handled by the application, styling elements (headers, text formatting, equations) can be easily written in text, even achieving greater consistency than word processors. Disciplines closely related to computational biology, such as statistics and mathematics, have historically used the markup language LaTeX for writing articles. This language has a specific syntax to write mathematics constructs as simple text, making it a sound choice for papers with lots of equations. To aid collaborative writing, platforms like Overleaf provide online LaTeX editors, supporting features like real-time editing. In addition to LaTeX, an emerging trend in collaborative writing is to use the lightweight markup language Markdown within the GitHub infrastructure. The software Manubot provides a set of functionalities to write scholarly articles within a GitHub repository, leveraging all the advantages of Git version control and the GitHub hosting platform <span class="citation" data-cites="YuJbg3zO">[<a href="#ref-YuJbg3zO" role="doc-biblioref">56</a>]</span>. For example, it provides cloud storage, version control, and facilitates the maintainers’ work by managing updates via pull requests. The GitHub user interface also allows off-line discussions about the manuscript using issues" and task assignment (See level 3 for tips on project management). Manubot, in particular, accepts citations using manuscripts identifiers and renders automatically renders the article in PDF and HTML formats. As a drawback, it requires technical expertise in Git and familiarity with GitHub; as an upside, its reliable infrastructure scales well to large and open collaborative projects.</p>
<h2 id="level-3-community">Level 3: Community</h2>
<p>The third and final step of this journey is when you want to present your research to the community. This can already be done shortly after the start of your project, or towards the end, when you are close to publish your results. Either way, the main goal when doing it should be to develop and maintain an open and reproducible computational biology project, with community engagement over time. Here, we can distinguish 3 sub-goals: Make your research 1) accessible, 2) reproducible and 3) sustainable. The latter is especially relevant when part of the research involves developing code that will be used by others in the future (e.g. a tool or workflow), but we believe that many of those ideas are still relevant to any computational biology project as well.</p>
<h3 id="make-your-research-accessible">3.1. Make your research accessible</h3>
<p>Making your research accessible is the first step towards open research. This include ensuring that anyone (inside or outside of academia) can access your research, and that your research stays available long after your paper is published. It is extremely frustrating for any researcher to look for software or a set of scripts from a paper published a few years ago, only to find a “404 error” where the code used to be provided. Even more frustrating is whenever authors offer code as “available on request”; this often leads to an email from the supervisor of the project explaining that the developer of the code left years ago so the code is not available anymore.</p>
<p>There are three main ways in which people publish accompanying code to publications: As supplementary material in the publication it was used, via privately-owned domains, or via public repositories. The first option (publishing the code as supplementary material) has low accessibility if the paper is not open access, as only members of institutions that pay the journal membership will be able to access it. Moreover, it is completely static and cannot ever be updated if a new feature wants to be introduced, or a bug is found. The second option (privately-owned domains) lacks sustainability, as it requires certain maintenance, and either the domain can expire or the website might migrate somewhere else. Therefore, here we argue for the third option (public repositories) because of its accessibility and sustainability over time, as owners can update the code (if necessary) in a git-compliant way, and they don’t need to rely on maintaining a domain and/or servers by themselves. As mentioned before (see <a href="#level-2-collaboration">Level 2: Collaboration</a>), there are several hosting services for this purpose <span class="citation" data-cites="dmEZPbB4 16Oz9VFEk 1E58VA9bS">[<a href="#ref-dmEZPbB4" role="doc-biblioref">19</a>,<a href="#ref-16Oz9VFEk" role="doc-biblioref">20</a>,<a href="#ref-1E58VA9bS" role="doc-biblioref">21</a>]</span> (Table <a href="#tbl:community-tools-1">1</a>), all equally valid depending on where people in your specific field usually publish their tools. As the authors have more experience with Github <span class="citation" data-cites="dmEZPbB4">[<a href="#ref-dmEZPbB4" role="doc-biblioref">19</a>]</span>, in the following several repository-specific tools will be exemplified with Github.</p>
<div id="tbl:community-tools-1" class="tablenos">
<table>
<caption><span>Table 1:</span> Tools for making your research accessible. For each highlighted goal, different tools that achieve that goal, together with additional remarks, are shown. </caption>
<colgroup>
<col style="width: 17%" />
<col style="width: 35%" />
<col style="width: 47%" />
</colgroup>
<thead>
<tr class="header">
<th>Goal</th>
<th>Tool options</th>
<th>Additional remarks</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Publish your code</td>
<td>• <strong>Github</strong> <span class="citation" data-cites="dmEZPbB4">[<a href="#ref-dmEZPbB4" role="doc-biblioref">19</a>]</span><br/>• <strong>Gitlab</strong> <span class="citation" data-cites="16Oz9VFEk">[<a href="#ref-16Oz9VFEk" role="doc-biblioref">20</a>]</span><br/>• <strong>Bitbucket</strong> <span class="citation" data-cites="1E58VA9bS">[<a href="#ref-1E58VA9bS" role="doc-biblioref">21</a>]</span></td>
<td>All three options allow you to host your repository online for free. Choose whichever is more common in your own field.</td>
</tr>
<tr class="even">
<td>Introduce your code</td>
<td>• <strong>README file</strong> <span class="citation" data-cites="16wWVekW1">[<a href="#ref-16wWVekW1" role="doc-biblioref">57</a>]</span>: First file that shows up in a repository. <br/>• <strong>Github Pages</strong> <span class="citation" data-cites="nR2NTYy9">[<a href="#ref-nR2NTYy9" role="doc-biblioref">58</a>]</span>: Separate website.</td>
<td>Provide a landing page to any repository with a short overview of the code (installation, usage, acknowledgments, etc).</td>
</tr>
<tr class="odd">
<td>Share your code</td>
<td>• Several licensing options <span class="citation" data-cites="Lbp8MQSL">[<a href="#ref-Lbp8MQSL" role="doc-biblioref">59</a>]</span>.</td>
<td>Indicate with a license file what restrictions apply when using your code. If you don’t include this, you will loose many users.</td>
</tr>
<tr class="even">
<td>Archive your code</td>
<td>• <strong>Github Releases</strong> <span class="citation" data-cites="146GfpGTF">[<a href="#ref-146GfpGTF" role="doc-biblioref">60</a>]</span><br/>• <strong>Zenodo</strong> <span class="citation" data-cites="Z6ni32OG">[<a href="#ref-Z6ni32OG" role="doc-biblioref">52</a>]</span>: Provides DOI.<br/>• <strong>figshare</strong> <span class="citation" data-cites="1EMndz0j9">[<a href="#ref-1EMndz0j9" role="doc-biblioref">53</a>]</span>: Provides DOI.</td>
<td>Share progressive stable versions of your code as you develop it. Use <strong>semantic versioning</strong> <span class="citation" data-cites="13UfgtMhu">[<a href="#ref-13UfgtMhu" role="doc-biblioref">61</a>]</span> for assigning standard identifiers to your releases.</td>
</tr>
<tr class="odd">
<td>Publish a tool</td>
<td>• <strong>PyPI</strong> <span class="citation" data-cites="Mduzvxei">[<a href="#ref-Mduzvxei" role="doc-biblioref">62</a>]</span>: Python.<br/>• <strong>CRAN</strong> <span class="citation" data-cites="rWFDdwIs">[<a href="#ref-rWFDdwIs" role="doc-biblioref">63</a>]</span>: R.<br/>• <strong>Bioconductor</strong> <span class="citation" data-cites="10CJ2sJo8">[<a href="#ref-10CJ2sJo8" role="doc-biblioref">64</a>]</span>: R.<br/>• <strong>Bioconda</strong> <span class="citation" data-cites="AvIrTFAn">[<a href="#ref-AvIrTFAn" role="doc-biblioref">65</a>]</span>: Language-agnostic.</td>
<td>Produce a package easy to install and use. Especially useful if you think you could have a userbase that will run the same analysis as you on other datasets and/or conditions.</td>
</tr>
<tr class="even">
<td>Publish an interactive web app</td>
<td>• <strong>Dash</strong> <span class="citation" data-cites="dk80YsaO">[<a href="#ref-dk80YsaO" role="doc-biblioref">66</a>]</span>: Python.<br/>• <strong>R-Shiny</strong> <span class="citation" data-cites="B0TdOflx">[<a href="#ref-B0TdOflx" role="doc-biblioref">67</a>]</span>: R.</td>
<td>Provide easy and interactive data exploration to your users. Especially useful if you have large datasets that can be explored in different ways.</td>
</tr>
</tbody>
</table>
</div>
<p>When publishing any piece of code online, there are two files that are fundamental to include: A readme file, and a license. Including a readme file <span class="citation" data-cites="16wWVekW1">[<a href="#ref-16wWVekW1" role="doc-biblioref">57</a>]</span> is about welcoming your users, and introducing them to your code (Table <a href="#tbl:community-tools-1">1</a>): it should include a description about its main intended use, an overview of the installation, the most common commands, a way to contact the developers if problems arise (see <a href="#make-your-research-sustainable">3.3. Make your research sustainable</a> for more details on this), and acknoledgments (if appropiate). We recommend to keep it short: You can include all the details in the documentation of the tool, here we only need a quick overview. The readme file, which can be written in different markup languages such as Markdown <span class="citation" data-cites="LSHekw4z">[<a href="#ref-LSHekw4z" role="doc-biblioref">68</a>]</span> or reStructuredText <span class="citation" data-cites="RPjXsSC8">[<a href="#ref-RPjXsSC8" role="doc-biblioref">69</a>]</span>, will render automatically on the repository’s landing page, below the repository file structure. For improved clarity, you may consider creating a separate landing page: it will declutter the look by only showing the information of the readme file, and use improved esthetics. All hosting services offer simple ways to do this; in the case of Github, you can use Github pages <span class="citation" data-cites="nR2NTYy9">[<a href="#ref-nR2NTYy9" role="doc-biblioref">58</a>]</span>.</p>
<p>Adding a license to a repository is also a crucial step (Table <a href="#tbl:community-tools-1">1</a>). Licenses indicate how the code can be used: Is it free to use for any application? Can users modify the code as they please? Does it come with a warranty that it will work? Can it be used for profit? If a license is not provided, many researchers will choose not to use the code at all, as they won’t have an answer to these questions. For instance, academic users will not know if they are entitled to modify the code for their own research, and users from the industry will not know if the code can be used for profit. Many options exist for licensing code <span class="citation" data-cites="Lbp8MQSL">[<a href="#ref-Lbp8MQSL" role="doc-biblioref">59</a>]</span>, from permissive licenses that allow any kind of use with few or no conditions, like the Unlicense and MIT licenses, to more restrictive licenses that enforce disclosing the source and even requiring that any adaptation of the code uses the same license, like the GNU licenses. Our suggestion is to choose whatever suits your research group best; consider that as a rule of thumb, the more requirements you add, the less potential users you will have, but the more credit you will receive when users utilize your research for their own needs.</p>
<p>Working as a computational biologist, you will probably continue lines of work from scripts or software you have already been published. For instance, perhaps you improve performance of a given function, or add a new set of features entirely. Therefore, not only you should be interested in your code being accessible as a single instance, but instead provide ways of accessing numerous versions. To keep different versions of your code organized as you develop it, it is important to create and archive succeeding releases of your code (Table <a href="#tbl:community-tools-1">1</a>). A good way to do this is with Github releases <span class="citation" data-cites="146GfpGTF">[<a href="#ref-146GfpGTF" role="doc-biblioref">60</a>]</span>, which requires minimal effort and shows all versions you have created of your code. An even better way to do this is with research repositories like Zenodo <span class="citation" data-cites="Z6ni32OG">[<a href="#ref-Z6ni32OG" role="doc-biblioref">52</a>]</span> or Figshare <span class="citation" data-cites="1EMndz0j9">[<a href="#ref-1EMndz0j9" role="doc-biblioref">53</a>]</span>, which not only store your code and data, but also give you a digital object identifier (DOI) for each version, making different versions of your code citable, in case the corresponding publication is not publicly available yet, or enough time has passed since the publication so that the current version of the code differs widely from what was published. These repositories can even be combined with code repositories, e.g. Github has a Zenodo integration that will trigger a new archived version every time a new release is made from Github. Regardless of the solution chosen, we recommend keeping some logical order to the releases, using a standard such as semantic versioning <span class="citation" data-cites="13UfgtMhu">[<a href="#ref-13UfgtMhu" role="doc-biblioref">61</a>]</span>.</p>
<p>In most cases, it is probably enough to provide your code as an organized set of scripts and/or notebooks, for anyone to consult if they wish to reproduce and/or re-utilize your code. However, if you believe that your code would be partly or entirely used in a routine-fashion by other researchers, for instance for studying other organisms or other experimental conditions, you could consider packaging your code as a tool (Table <a href="#tbl:community-tools-1">1</a>). If you take this route, aim for using one of the main options for publishing tools as packages: Bioconda <span class="citation" data-cites="AvIrTFAn">[<a href="#ref-AvIrTFAn" role="doc-biblioref">65</a>]</span> is a catch-all solution, but language-specific options also exist, such as PyPI <span class="citation" data-cites="Mduzvxei">[<a href="#ref-Mduzvxei" role="doc-biblioref">62</a>]</span> if you work in Python, or CRAN <span class="citation" data-cites="rWFDdwIs">[<a href="#ref-rWFDdwIs" role="doc-biblioref">63</a>]</span> and Bioconductor <span class="citation" data-cites="10CJ2sJo8">[<a href="#ref-10CJ2sJo8" role="doc-biblioref">64</a>]</span> if you work in R. All of these options will allow you to reach a bigger audience, as packages can be easily searched and installed locally with minimal effort.</p>
<p>Another common situation in research projects is, instead of having code that can be used by others to analyze their data, having data that can be analyzed in many ways. As producing plots that display the data in every conceivable way is unreasonable, a good solution here is to develop an interactive web app (Table <a href="#tbl:community-tools-1">1</a>), also referred to as a data dashboard, that lets users interact with the data, by showing different sets of variables or changing parameter settings (e.g. the significance of a statistical test). Common options for this goal are Dash <span class="citation" data-cites="dk80YsaO">[<a href="#ref-dk80YsaO" role="doc-biblioref">66</a>]</span> for Python and R Shiny <span class="citation" data-cites="B0TdOflx">[<a href="#ref-B0TdOflx" role="doc-biblioref">67</a>]</span> for R.</p>
<h3 id="make-your-research-reproducible">3.2. Make your research reproducible</h3>
<p>Having your code/data accessible to anyone is only the first step when sharing software to the research community; you also need for any analysis you have conducted to be reproducible by anyone. We have already discussed the importance of reproducibility in science when working in a personal project (see <a href="#level-1-personal-research">Level 1: Personal Research</a>); however, when sharing your results with the academic community this becomes even more paramount, as anyone should be able to run your code, and obtain the same results. This is especially relevant in science, as there is a constant number of people that are new to the field (including undergraduates, postgraduates and senior researchers), many who come from very different backgrounds to try to understand how your code, developed for your own specific niche research area, works.</p>
<p>A cornerstone for reproducibility is to have a solid documentation: by explaining what your code does, users can understand the intended use of each function, which will guide them on how to achieve your same results. We can distinguish 4 different levels of documentation <span class="citation" data-cites="woYMAGdl">[<a href="#ref-woYMAGdl" role="doc-biblioref">70</a>]</span>:</p>
<ul>
<li><em>Tutorials:</em> A group of lessons that teach the reader how to become a user of your code.</li>
<li><em>How-to guides:</em> A set of documents that clarify to a user how to solve common problems/tasks.</li>
<li><em>Explanations:</em> Discussions that clarify particular topics related to your code.</li>
<li><em>References:</em> Technical descriptions of your code’s variables/classes/functions.</li>
</ul>
<p>The extent of documentation you write will depend on how many users you expect to have, and conversely will affect how many users you attract. If you foresee that your code has little usability outside of your research, perhaps a solid documentation of each function using docstrings <span class="citation" data-cites="IsDZd09v">[<a href="#ref-IsDZd09v" role="doc-biblioref">71</a>]</span> could already be enough. However, you might also want to add a tutorial for a beginner, a couple of how-to guides for frequently used routines, and even some explanations for clarifying the science behind your code. The latter has the added bonus that it can be re-used for the eventual manuscript of your publication. If you are not sure how many users you might get, air on the side of caution and prepare good documentation anyways: You will be surprised to see how often other researchers contact you for reproducing your results and/or using your code for other applications! Finally, to publish comprehensive documentation online, consider using 1) a standard documentation language such as reStructuredText <span class="citation" data-cites="RPjXsSC8">[<a href="#ref-RPjXsSC8" role="doc-biblioref">69</a>]</span> or Markdown <span class="citation" data-cites="LSHekw4z">[<a href="#ref-LSHekw4z" role="doc-biblioref">68</a>]</span>, and 2) a documentation platform such as Readthedocs <span class="citation" data-cites="9HZXe8sx">[<a href="#ref-9HZXe8sx" role="doc-biblioref">72</a>]</span> or Gitbook <span class="citation" data-cites="VjDIpAqq">[<a href="#ref-VjDIpAqq" role="doc-biblioref">73</a>]</span> (Table <a href="#tbl:community-tools-2">2</a>).</p>
<div id="tbl:community-tools-2" class="tablenos">
<table>
<caption><span>Table 2:</span> Tools for making your research reproducible. For each highlighted goal, different tools that achieve that goal, together with additional remarks, are shown. </caption>
<colgroup>
<col style="width: 18%" />
<col style="width: 43%" />
<col style="width: 37%" />
</colgroup>
<thead>
<tr class="header">
<th>Goal</th>
<th>Tool options</th>
<th>Additional remarks</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Document your code</td>
<td>• <strong>Readthedocs</strong> <span class="citation" data-cites="9HZXe8sx">[<a href="#ref-9HZXe8sx" role="doc-biblioref">72</a>]</span>: Uses <strong>reStructuredText</strong> <span class="citation" data-cites="RPjXsSC8">[<a href="#ref-RPjXsSC8" role="doc-biblioref">69</a>]</span>.<br/>• <strong>Gitbook</strong> <span class="citation" data-cites="VjDIpAqq">[<a href="#ref-VjDIpAqq" role="doc-biblioref">73</a>]</span>: Uses <strong>Markdown</strong> <span class="citation" data-cites="LSHekw4z">[<a href="#ref-LSHekw4z" role="doc-biblioref">68</a>]</span>.</td>
<td>Comprehensive documentation: from tutorials and how-to guides all the way down to function documentation based on all compiled <strong>docstrings</strong> <span class="citation" data-cites="IsDZd09v">[<a href="#ref-IsDZd09v" role="doc-biblioref">71</a>]</span>.</td>
</tr>
<tr class="even">
<td>Reproducible environments</td>
<td>• <strong>Environment managers</strong>: See Table XXX.<br/>• <strong>pip-tools</strong> <span class="citation" data-cites="dvOzCOyW">[<a href="#ref-dvOzCOyW" role="doc-biblioref">74</a>]</span>: Administer several environments in a single project.</td>
<td>As a recommendation, try having the minimum number of dependencies needed to reproduce your results.</td>
</tr>
<tr class="odd">
<td>Reproducible software</td>
<td>• <strong>Docker</strong> <span class="citation" data-cites="rmvJ2O0Q">[<a href="#ref-rmvJ2O0Q" role="doc-biblioref">75</a>]</span><br/>• <strong>Singularity</strong> <span class="citation" data-cites="99QkJBoF">[<a href="#ref-99QkJBoF" role="doc-biblioref">76</a>]</span></td>
<td>Package your research as a container ready to run in any computer.</td>
</tr>
<tr class="even">
<td>Reproducible commands</td>
<td>• <strong>Make</strong> <span class="citation" data-cites="5iKCi5fm">[<a href="#ref-5iKCi5fm" role="doc-biblioref">77</a>]</span></td>
<td>Build a program by following a series of steps in a single Makefile.</td>
</tr>
<tr class="odd">
<td>Reproducible workflows</td>
<td>• <strong>Make</strong> <span class="citation" data-cites="5iKCi5fm">[<a href="#ref-5iKCi5fm" role="doc-biblioref">77</a>]</span><br/>• <strong>Snakemake</strong> <span class="citation" data-cites="q2TJoDAO">[<a href="#ref-q2TJoDAO" role="doc-biblioref">78</a>]</span>: Uses a Python-based language.<br/>• <strong>Nextflow</strong> <span class="citation" data-cites="11OYUtIpR">[<a href="#ref-11OYUtIpR" role="doc-biblioref">79</a>]</span>: Uses <strong>Groovy</strong> <span class="citation" data-cites="CVNjbImf">[<a href="#ref-CVNjbImf" role="doc-biblioref">80</a>]</span>.</td>
<td>Run a pipeline of commands on NGS data in a reproducible way.</td>
</tr>
<tr class="even">
<td>Reproducible notebooks</td>
<td>• <strong>Binder</strong> <span class="citation" data-cites="1dqfprG7">[<a href="#ref-1dqfprG7" role="doc-biblioref">35</a>]</span><br/>• <strong>Colaboratory</strong> <span class="citation" data-cites="mr26BIpF">[<a href="#ref-mr26BIpF" role="doc-biblioref">36</a>]</span></td>
<td>Make your notebooks interactive and reproducible.</td>
</tr>
</tbody>
</table>
</div>
<p><strong>TODO:</strong> write about environments -&gt; depends on what is written in Level 1.</p>
<p>A tool worth using when defining each environment is pip-tools <span class="citation" data-cites="dvOzCOyW">[<a href="#ref-dvOzCOyW" role="doc-biblioref">74</a>]</span>, which allows defining several different environments for a single project, depending on who accesses the project. For instance, you might want users to use a wider number of dependencies for increased flexibility, but the CI tool to only have a minimal number of dependencies for improved efficiency.</p>
<p>Beyond dependency trackers, you might also want to ensure your tool behaves in the same way across computing environments, even between two different operative systems (e.g. Mac and Windows). The best solution for this is to use a container (Table <a href="#tbl:community-tools-2">2</a>), which is a standardized unit of software that runs using an isolated filesystem known as a container image. This image contains not only the needed dependencies, but also any configurations, binary files, environmental variables, etc. that the software needs for running, and can be built from a text file with all the instructions. The two main tools available for creating containers for free are Docker <span class="citation" data-cites="rmvJ2O0Q">[<a href="#ref-rmvJ2O0Q" role="doc-biblioref">75</a>]</span> and Singularity <span class="citation" data-cites="99QkJBoF">[<a href="#ref-99QkJBoF" role="doc-biblioref">76</a>]</span>.</p>
<p><strong>TODO:</strong> make, as any type of workflow -&gt; depends on what is written in Level 2.</p>
<p>There are certain research projects where particular resources for ensuring reproducibility are especially useful: as previously mentioned (see <a href="#level-2-collaboration">Level 2: Collaboration</a>) in the case you have developed a complex workflow, for example a pipeline for analyzing NGS data, tools like Snakemake <span class="citation" data-cites="q2TJoDAO">[<a href="#ref-q2TJoDAO" role="doc-biblioref">78</a>]</span> (which uses a Python-based language) or Nextflow <span class="citation" data-cites="11OYUtIpR">[<a href="#ref-11OYUtIpR" role="doc-biblioref">79</a>]</span> (which uses <strong>Groovy</strong> <span class="citation" data-cites="CVNjbImf">[<a href="#ref-CVNjbImf" role="doc-biblioref">80</a>]</span>) will certainly be helpful. Finally, in the case that you have prepared most of your code using Jupyter notebooks, tools like Binder <span class="citation" data-cites="1dqfprG7">[<a href="#ref-1dqfprG7" role="doc-biblioref">35</a>]</span> or Colaboratory <span class="citation" data-cites="mr26BIpF">[<a href="#ref-mr26BIpF" role="doc-biblioref">36</a>]</span> allow you to publish interactive reproducible notebooks that anyone can run from their own setup.</p>
<h3 id="make-your-research-sustainable">3.3. Make your research sustainable</h3>
<p>Now that your research can be accessed and reproduced by anyone, the final step is to keep it like that over time. This is especially relevant if you carry the research further by integrating new features, as you need to ensure no bugs appear in the rest of the code, and at the same time offer options for your user community to request new functionalities, in order to foster a strong community over time. However, even in the case in which your research is a self-contained project that will not be continued, it is still important to ensure that your user community has ways to contact you, in case bugs would be discovered in your code or, due to packages/dependencies updating, parts of your code would not work anymore over time (a common phenomenon referred to as “software rot” <span class="citation" data-cites="8tGcpIB0">[<a href="#ref-8tGcpIB0" role="doc-biblioref">81</a>]</span>). In the following we will therefore review useful techniques for making your code/software/research sustainable over time.</p>
<p>The first thing to consider when ensuring sustainability is to offer the users of your code a way to communicate with you (Table <a href="#tbl:community-tools-3">3</a>). You can achieve this with different tools, depending on the size of your userbase and the scope of the questions you receive. For smaller projects, where there are only questions once in a while, a single-channel solution like Gitter <span class="citation" data-cites="iXfPU3Cy">[<a href="#ref-iXfPU3Cy" role="doc-biblioref">82</a>]</span> is probably enough, as it offers a simple way for anyone in the community to ask questions, and any of the developers to answer in threads. For larger projects however, it could become unmanageable to have all discussions in the same channel, so multiple-channel solutions, i.e. forums, are better suited. Google groups <span class="citation" data-cites="6v98AMba">[<a href="#ref-6v98AMba" role="doc-biblioref">83</a>]</span> is a good example that works well for large pieces of software, as it allows for anyone to open separate threads for different issues. Recently, Github introduced Github Discussions <span class="citation" data-cites="ngGdYg2r">[<a href="#ref-ngGdYg2r" role="doc-biblioref">84</a>]</span>, also meant for keeping questions organized in different threads.</p>
<div id="tbl:community-tools-3" class="tablenos">
<table>
<caption><span>Table 3:</span> Tools for making your research sustainable. For each highlighted goal, different tools that achieve that goal, together with additional remarks, are shown. </caption>
<colgroup>
<col style="width: 20%" />
<col style="width: 33%" />
<col style="width: 46%" />
</colgroup>
<thead>
<tr class="header">
<th>Goal</th>
<th>Tool options</th>
<th>Additional remarks</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Tell users how to contact you</td>
<td>• Specific/shorter questions: <strong>Gitter</strong> <span class="citation" data-cites="iXfPU3Cy">[<a href="#ref-iXfPU3Cy" role="doc-biblioref">82</a>]</span>.<br/>• Larger issues / how-to’s: <strong>Google groups</strong> <span class="citation" data-cites="6v98AMba">[<a href="#ref-6v98AMba" role="doc-biblioref">83</a>]</span>, <strong>GitHub Discussions</strong> <span class="citation" data-cites="ngGdYg2r">[<a href="#ref-ngGdYg2r" role="doc-biblioref">84</a>]</span>.</td>
<td>Provide ways for users to contact you for questions, requests, etc. Remember to visit them periodically!</td>
</tr>
<tr class="even">
<td>Track to-do’s in your research</td>
<td>• <strong>Github Issues</strong> <span class="citation" data-cites="5HELQ2ms">[<a href="#ref-5HELQ2ms" role="doc-biblioref">85</a>]</span></td>
<td>Detail specific pending to-do’s in your research / allow others to request changes and/or highlight bugs.</td>
</tr>
<tr class="odd">
<td>Encourage user contributions</td>
<td>• <strong>Contribution guidelines</strong> <span class="citation" data-cites="ep7wwgWj">[<a href="#ref-ep7wwgWj" role="doc-biblioref">86</a>]</span>: How to open issues / contribute code.<br/>• <strong>Github Wikis</strong> <span class="citation" data-cites="pVl1NEXZ">[<a href="#ref-pVl1NEXZ" role="doc-biblioref">87</a>]</span>: More specific how-to guides.</td>
<td>Provide as much information as you can to guide your users. You can also include administrator guidelines.</td>
</tr>
<tr class="even">
<td>Foster a respectful community</td>
<td>• Smaller projects: <strong>Contributor Covenant</strong> <span class="citation" data-cites="115FgZQh6">[<a href="#ref-115FgZQh6" role="doc-biblioref">88</a>]</span>.<br/>• Larger projects: <strong>Citizen Code of Conduct</strong> <span class="citation" data-cites="qcPNBi0h">[<a href="#ref-qcPNBi0h" role="doc-biblioref">89</a>]</span>.</td>
<td>Essential when you would like researchers to contribute code.</td>
</tr>
<tr class="odd">
<td>Branch your repo sustainably</td>
<td>• <strong>Gitflow</strong> <span class="citation" data-cites="1Ei7XPHmQ">[<a href="#ref-1Ei7XPHmQ" role="doc-biblioref">90</a>]</span></td>
<td>Useful when several developers contribute code to the project. Allows users to get access to stable versions of your research in an ongoing project.</td>
</tr>
<tr class="even">
<td>Keep track of your issues</td>
<td>• <strong>Kanban flowcharts</strong> <span class="citation" data-cites="zcUBK27J">[<a href="#ref-zcUBK27J" role="doc-biblioref">91</a>]</span>: <strong>Github Projects</strong> <span class="citation" data-cites="ssoXtriL">[<a href="#ref-ssoXtriL" role="doc-biblioref">92</a>]</span>, <strong>GitKraken Boards</strong> <span class="citation" data-cites="QrHcaBPj">[<a href="#ref-QrHcaBPj" role="doc-biblioref">93</a>]</span>.<br/>• <strong>Scrum practices</strong> <span class="citation" data-cites="LLiWj74b">[<a href="#ref-LLiWj74b" role="doc-biblioref">94</a>]</span>: <strong>Zenhub</strong> <span class="citation" data-cites="rX7VZWos">[<a href="#ref-rX7VZWos" role="doc-biblioref">95</a>]</span>, <strong>Jira</strong> <span class="citation" data-cites="a6SmwGuV">[<a href="#ref-a6SmwGuV" role="doc-biblioref">96</a>]</span>.</td>
<td>Keep track of your pending tasks in different projects with <strong>Agile</strong> <span class="citation" data-cites="qOcBF2oS">[<a href="#ref-qOcBF2oS" role="doc-biblioref">97</a>]</span> software development practices. Especially useful if your research is split in many different repositories, each with multiple features/fixes to do.</td>
</tr>
<tr class="odd">
<td>Automate your repo</td>
<td>• <strong>bump2version</strong> <span class="citation" data-cites="1UVh4XJi">[<a href="#ref-1UVh4XJi" role="doc-biblioref">98</a>]</span>: Easier releasing.<br/>• <strong>Danger-CI</strong> <span class="citation" data-cites="SXI7Canc">[<a href="#ref-SXI7Canc" role="doc-biblioref">99</a>]</span>: Easier reviewing.</td>
<td>Do less, script more!</td>
</tr>
</tbody>
</table>
</div>
<p><strong>TODO:</strong> Issues (GitHub issues/milestones/labels/templates): Have a to-do list -&gt; based on what is written in Level 2</p>
<p>Now that your users know <em>where</em> to contact you, you should also tell them <em>how</em> to contact you. For this, it is essential to have a file with contribution guidelines <span class="citation" data-cites="ep7wwgWj">[<a href="#ref-ep7wwgWj" role="doc-biblioref">86</a>]</span> (Table <a href="#tbl:community-tools-3">3</a>), detailing how should users 1) open issues and 2) contribute with their own code changes, in form of pull requests. These guidelines are mainly intended for new users/contributors, so they should be written in the style of a <em>how-to</em> guide. However, they may also include additional instructions for the main developers, or even the administrator of the repository. Alternatively, those more detailed guidelines can be included in a supplemental wiki, which hosting services offer as part of the repository <span class="citation" data-cites="pVl1NEXZ">[<a href="#ref-pVl1NEXZ" role="doc-biblioref">87</a>]</span>.</p>
<p>Equally important to the contribution guidelines file is the code of conduct file (Table <a href="#tbl:community-tools-3">3</a>), which includes guidelines on how to behave when engaging with the community in the repository, and what to do if someone does not comply with these guidelines. It is an important statement to ensure users and developers respect one another. Several templates exist as code of conduct, such as the Contributor Covenant <span class="citation" data-cites="115FgZQh6">[<a href="#ref-115FgZQh6" role="doc-biblioref">88</a>]</span> for smaller projects, and the Citizen Code of Conduct <span class="citation" data-cites="qcPNBi0h">[<a href="#ref-qcPNBi0h" role="doc-biblioref">89</a>]</span> for larger projects.</p>
<p>Finally, ensuring sustainability of your project means not only allowing and encouraging users contacting you, but also smarter ways to develop and maintain your software as it grows in scope and number of users. This includes:</p>
<ol type="1">
<li><p><em>Branching System:</em> When many developers are involved in a project, you want to ensure that users can access functional versions of your code while you work on it. In this case, more advanced branching methods such as GitFlow <span class="citation" data-cites="1Ei7XPHmQ">[<a href="#ref-1Ei7XPHmQ" role="doc-biblioref">90</a>]</span> might be preferable (Table <a href="#tbl:community-tools-3">3</a>). With GitFlow, an additional development branch (often named as <code>devel</code>) is used as main branch for new branches to be based on, leaving the master branch as a separate branch only for stable versions of the code, and a merge from development to master always invoking a code release. This way, users can directly access the master branch for tested releases, whereas the latest additions of the code, perhaps not 100% tested, will be available for developers to further work on in the development branch. Additional branches can be added depending on the scope of the project, such as specific version branches in case further testing is needed, or hotfix branches, in case bugs are detected in master and a quick solution is required.</p></li>
<li><p><em>Project Management:</em> Although issues are fundamental for keeping track of what is there to do, they can become hard to organize and prioritize as they grow in number. Several project management tools exist to solve this (Table <a href="#tbl:community-tools-3">3</a>), all based on Agile <span class="citation" data-cites="qOcBF2oS">[<a href="#ref-qOcBF2oS" role="doc-biblioref">97</a>]</span> principles. The most simple one is to have a Kanban board <span class="citation" data-cites="zcUBK27J">[<a href="#ref-zcUBK27J" role="doc-biblioref">91</a>]</span>, where issues are organized in as many columns as necessary to have a clear layout and figure out what is the current state of a given task. Tools like Github Projects <span class="citation" data-cites="ssoXtriL">[<a href="#ref-ssoXtriL" role="doc-biblioref">92</a>]</span> or GitKraken Boards <span class="citation" data-cites="QrHcaBPj">[<a href="#ref-QrHcaBPj" role="doc-biblioref">93</a>]</span> use this approach. For Larger projects with either several collaborators and/or several repositories, a more structured approach such as a Scrum framework <span class="citation" data-cites="LLiWj74b">[<a href="#ref-LLiWj74b" role="doc-biblioref">94</a>]</span> might be needed, as issues are more easily prioritized by setting milestones, estimated by assigning points, and dependencies can be defined. Both Zenhub <span class="citation" data-cites="rX7VZWos">[<a href="#ref-rX7VZWos" role="doc-biblioref">95</a>]</span> and Jira <span class="citation" data-cites="a6SmwGuV">[<a href="#ref-a6SmwGuV" role="doc-biblioref">96</a>]</span> are great options for this.</p></li>
<li><p><em>Additional Automation:</em> As you develop your project, you will find that many aspects can be automated to improve efficiency. Want to make releasing faster to ensure all sections of your code get updated with the new release? bump2version <span class="citation" data-cites="1UVh4XJi">[<a href="#ref-1UVh4XJi" role="doc-biblioref">98</a>]</span> might help. Want to make sure contributors comply with certain standards in their pull requests? Look into Danger-CI <span class="citation" data-cites="SXI7Canc">[<a href="#ref-SXI7Canc" role="doc-biblioref">99</a>]</span>. Note here that we advice against implementing all of this from the start, but instead adding different tools as you realize you need them. If you find yourself performing a task in a routinely fashion, ask yourself if you could automate it. More often than not the answer is yes!</p></li>
</ol>
<p><strong>TODO:</strong> More automation options?</p>
<h2 id="case-studies">Case studies</h2>
<ul>
<li>Example of computational biology project 1: RNA-seq analysis (workflow)</li>
<li>Example of computational biology project 2: Genome-scale metabolic model (systems biology project)</li>
<li>Example of computational biology project 3: Software development (computational tool)</li>
</ul>
<h2 id="conclusion">Conclusion</h2>
<p>Pending</p>
<h2 id="acknowledgments">Acknowledgments</h2>
<p>Pending</p>
<h2 class="page_break_before" id="references">References</h2>
<!-- Explicitly insert bibliography here -->
<div id="refs" class="references hanging-indent" role="doc-bibliography">
<div id="ref-Xcr0K7I5">
<p>1. <strong>NIH working definition of bioinformatics and computational biology</strong> <br />
Huerta Michael, Downing Gregory, Haseltine Florence, Seto Belinda, Liu Yuan<br />
(2000-07-17) <a href="https://www.kennedykrieger.org/sites/default/files/library/documents/research/center-labs-cores/bioinformatics/bioinformatics-def.pdf">https://www.kennedykrieger.org/sites/default/files/library/documents/research/center-labs-cores/bioinformatics/bioinformatics-def.pdf</a></p>
</div>
<div id="ref-jn2NeK3H">
<p>2. <strong>What is bioinformatics? A proposed definition and overview of the field.</strong> <br />
NM Luscombe, D Greenbaum, M Gerstein<br />
<em>Methods of information in medicine</em> (2001) <a href="https://www.ncbi.nlm.nih.gov/pubmed/11552348">https://www.ncbi.nlm.nih.gov/pubmed/11552348</a> <br />
PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/11552348">11552348</a></p>
</div>
<div id="ref-k9JCaCug">
<p>3. <strong>Good enough practices in scientific computing</strong> <br />
Greg Wilson, Jennifer Bryan, Karen Cranston, Justin Kitzes, Lex Nederbragt, Tracy K. Teal<br />
<em>PLOS Computational Biology</em> (2017-06-22) <a href="https://doi.org/gbkbwp">https://doi.org/gbkbwp</a> <br />
DOI: <a href="https://doi.org/10.1371/journal.pcbi.1005510">10.1371/journal.pcbi.1005510</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/28640806">28640806</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5480810">PMC5480810</a></p>
</div>
<div id="ref-15pOpsdFR">
<p>4. <strong>Software engineering for scientific big data analysis</strong> <br />
Björn A Grüning, Samuel Lampa, Marc Vaudel, Daniel Blankenberg<br />
<em>GigaScience</em> (2019-05) <a href="https://doi.org/gf4f4m">https://doi.org/gf4f4m</a> <br />
DOI: <a href="https://doi.org/10.1093/gigascience/giz054">10.1093/gigascience/giz054</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31121028">31121028</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6532757">PMC6532757</a></p>
</div>
<div id="ref-SGTWzj9t">
<p>5. <strong>So you want to be a computational biologist?</strong> <br />
Nick Loman, Mick Watson<br />
<em>Nature Biotechnology</em> (2013-11-01) <a href="https://doi.org/p3j">https://doi.org/p3j</a> <br />
DOI: <a href="https://doi.org/10.1038/nbt.2740">10.1038/nbt.2740</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/24213777">24213777</a></p>
</div>
<div id="ref-kEX5dgzK">
<p>6. <strong>Ten Simple Rules for Taking Advantage of Git and GitHub</strong> <br />
Yasset Perez-Riverol, Laurent Gatto, Rui Wang, Timo Sachsenberg, Julian Uszkoreit, Felipe da Veiga Leprevost, Christian Fufezan, Tobias Ternent, Stephen J. Eglen, Daniel S. Katz, … Juan Antonio Vizcaíno<br />
<em>PLOS Computational Biology</em> (2016-07-14) <a href="https://doi.org/gbrb39">https://doi.org/gbrb39</a> <br />
DOI: <a href="https://doi.org/10.1371/journal.pcbi.1004947">10.1371/journal.pcbi.1004947</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/27415786">27415786</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4945047">PMC4945047</a></p>
</div>
<div id="ref-eYWjWdLd">
<p>7. <strong>Ten Simple Rules for Reproducible Research in Jupyter Notebooks</strong> <br />
Adam Rule, Amanda Birmingham, Cristal Zuniga, Ilkay Altintas, Shih-Cheng Huang, Rob Knight, Niema Moshiri, Mai H. Nguyen, Sara Brin Rosenthal, Fernando Pérez, Peter W. Rose<br />
<em>arXiv</em> (2018-10-13) <a href="https://arxiv.org/abs/1810.08055v1">https://arxiv.org/abs/1810.08055v1</a></p>
</div>
<div id="ref-1HhbGD4jJ">
<p>8. <strong>Streamlining Data-Intensive Biology With Workflow Systems</strong> <br />
Taylor Reiter, Phillip T. Brooks, Luiz Irber, Shannon E. K. Joslin, Charles M. Reid, Camille Scott, C. Titus Brown, N. Tessa Pierce<br />
<em>Cold Spring Harbor Laboratory</em> (2020-11-16) <a href="https://doi.org/gg353v">https://doi.org/gg353v</a> <br />
DOI: <a href="https://doi.org/10.1101/2020.06.30.178673">10.1101/2020.06.30.178673</a></p>
</div>
<div id="ref-16iU4PKI1">
<p>9. <strong>A Padawan Programmer’s Guide to Developing Software Libraries</strong> <br />
James T. Yurkovich, Benjamin J. Yurkovich, Andreas Dräger, Bernhard O. Palsson, Zachary A. King<br />
<em>Cell Systems</em> (2017-11) <a href="https://doi.org/gg8tqz">https://doi.org/gg8tqz</a> <br />
DOI: <a href="https://doi.org/10.1016/j.cels.2017.08.003">10.1016/j.cels.2017.08.003</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/28988801">28988801</a></p>
</div>
<div id="ref-lkZcKu7R">
<p>10. <strong>2018 Kaggle Machine Learning &amp; Data Science Survey</strong> <a href="https://kaggle.com/kaggle/kaggle-survey-2018">https://kaggle.com/kaggle/kaggle-survey-2018</a></p>
</div>
<div id="ref-1A4rxBI47">
<p>11. <strong>Modulecounts</strong> <a href="http://www.modulecounts.com/">http://www.modulecounts.com/</a></p>
</div>
<div id="ref-wu0sr8q2">
<p>12. <strong>Why scientists are turning to Rust</strong> <br />
Jeffrey M. Perkel<br />
<em>Nature</em> (2020-12-01) <a href="https://doi.org/ghqc7g">https://doi.org/ghqc7g</a> <br />
DOI: <a href="https://doi.org/10.1038/d41586-020-03382-2">10.1038/d41586-020-03382-2</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/33262490">33262490</a></p>
</div>
<div id="ref-j6HfMqoR">
<p>13. <strong>Home - Cookiecutter Data Science</strong> <a href="https://drivendata.github.io/cookiecutter-data-science/">https://drivendata.github.io/cookiecutter-data-science/</a></p>
</div>
<div id="ref-7D5DgTkH">
<p>14. <strong>Reproducible-Science-Curriculum/rr-init</strong> <br />
Reproducible Science Curriculum<br />
(2021-02-12) <a href="https://github.com/Reproducible-Science-Curriculum/rr-init">https://github.com/Reproducible-Science-Curriculum/rr-init</a></p>
</div>
<div id="ref-3Gx1ESE4">
<p>15. <strong>A Quick Guide to Organizing Computational Biology Projects</strong> <br />
William Stafford Noble<br />
<em>PLoS Computational Biology</em> (2009-07-31) <a href="https://doi.org/fbbpkn">https://doi.org/fbbpkn</a> <br />
DOI: <a href="https://doi.org/10.1371/journal.pcbi.1000424">10.1371/journal.pcbi.1000424</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/19649301">19649301</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2709440">PMC2709440</a></p>
</div>
<div id="ref-jq4WP878">
<p>16. <strong>Distribution and Reproducibility — Snakemake 5.32.2 documentation</strong> <a href="https://snakemake.readthedocs.io/en/stable/snakefiles/deployment.html">https://snakemake.readthedocs.io/en/stable/snakefiles/deployment.html</a></p>
</div>
<div id="ref-4PJ4Kllx">
<p>17. <strong>google/styleguide</strong> <br />
Google<br />
(2021-02-15) <a href="https://github.com/google/styleguide">https://github.com/google/styleguide</a></p>
</div>
<div id="ref-1D76el83m">
<p>18. <strong>PEP 8 – Style Guide for Python Code</strong> <br />
Python.org<br />
<a href="https://www.python.org/dev/peps/pep-0008/">https://www.python.org/dev/peps/pep-0008/</a></p>
</div>
<div id="ref-dmEZPbB4">
<p>19. <strong>GitHub: Where the world builds software</strong> <br />
GitHub<br />
<a href="https://github.com/">https://github.com/</a></p>
</div>
<div id="ref-16Oz9VFEk">
<p>20. <strong>DevOps Platform Delivered as a Single Application</strong> <br />
GitLab<br />
<a href="https://about.gitlab.com/">https://about.gitlab.com/</a></p>
</div>
<div id="ref-1E58VA9bS">
<p>21. <strong>Bitbucket | The Git solution for professional teams</strong> <br />
Atlassian<br />
<em>Bitbucket</em> <a href="https://bitbucket.org/product">https://bitbucket.org/product</a></p>
</div>
<div id="ref-qjg93S3G">
<p>22. <strong>Understanding the GitHub flow · GitHub Guides</strong> <a href="https://guides.github.com/introduction/flow/">https://guides.github.com/introduction/flow/</a></p>
</div>
<div id="ref-15kUlshEi">
<p>23. <strong>pytest: helps you write better programs — pytest documentation</strong> <a href="https://docs.pytest.org/en/stable/">https://docs.pytest.org/en/stable/</a></p>
</div>
<div id="ref-cEpTMlZc">
<p>24. <strong>Codecov - The Leading Code Coverage Solution</strong> <br />
Codecov<br />
<a href="https://about.codecov.io/">https://about.codecov.io/</a></p>
</div>
<div id="ref-1Bi82XR9D">
<p>25. <strong>Welcome to the tox automation project — tox 3.21.5.dev3 documentation</strong> <a href="https://tox.readthedocs.io/en/latest/">https://tox.readthedocs.io/en/latest/</a></p>
</div>
<div id="ref-SM1pqdjN">
<p>26. <strong>Travis CI - Test and Deploy with Confidence</strong> <a href="https://travis-ci.com/">https://travis-ci.com/</a></p>
</div>
<div id="ref-NUXbp429">
<p>27. <strong>Features • GitHub Actions</strong> <br />
GitHub<br />
<a href="https://github.com/features/actions">https://github.com/features/actions</a></p>
</div>
<div id="ref-RCh2gb01">
<p>28. <strong>Reviewing changes in pull requests - GitHub Docs</strong> <a href="https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/reviewing-changes-in-pull-requests">https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/reviewing-changes-in-pull-requests</a></p>
</div>
<div id="ref-6woNTKZi">
<p>29. <strong>Crucible Code Review Tool for Git, SVN, Perforce and More</strong> <br />
Atlassian<br />
<em>Atlassian</em> <a href="https://www.atlassian.com/software/crucible">https://www.atlassian.com/software/crucible</a></p>
</div>
<div id="ref-LJc9aT8R">
<p>30. <strong>Upsource: Code Review and Project Analytics by JetBrains</strong> <br />
JetBrains<br />
<a href="https://www.jetbrains.com/upsource/">https://www.jetbrains.com/upsource/</a></p>
</div>
<div id="ref-6DjakjNS">
<p>31. <strong>The FAIR Guiding Principles for scientific data management and stewardship</strong> <br />
Mark D. Wilkinson, Michel Dumontier, IJsbrand Jan Aalbersberg, Gabrielle Appleton, Myles Axton, Arie Baak, Niklas Blomberg, Jan-Willem Boiten, Luiz Bonino da Silva Santos, Philip E. Bourne, … Barend Mons<br />
<em>Scientific Data</em> (2016-03-15) <a href="https://doi.org/bdd4">https://doi.org/bdd4</a> <br />
DOI: <a href="https://doi.org/10.1038/sdata.2016.18">10.1038/sdata.2016.18</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/26978244">26978244</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4792175">PMC4792175</a></p>
</div>
<div id="ref-Gtnr8c2U">
<p>32. <strong>Tidy Data</strong> <br />
Hadley Wickham<br />
<em>Journal of Statistical Software</em> (2014) <a href="https://doi.org/gdm3p7">https://doi.org/gdm3p7</a> <br />
DOI: <a href="https://doi.org/10.18637/jss.v059.i10">10.18637/jss.v059.i10</a></p>
</div>
<div id="ref-JkrlcwLO">
<p>33. <strong>Data Version Control · DVC</strong> <a href="https://dvc.org/">https://dvc.org/</a></p>
</div>
<div id="ref-8Cn0lgXi">
<p>34. <strong>nbviewer</strong> <a href="https://nbviewer.jupyter.org/">https://nbviewer.jupyter.org/</a></p>
</div>
<div id="ref-1dqfprG7">
<p>35. <strong>The Binder Project</strong> <a href="https://mybinder.org/">https://mybinder.org/</a></p>
</div>
<div id="ref-mr26BIpF">
<p>36. <strong>Google Colaboratory</strong> <a href="https://colab.research.google.com/">https://colab.research.google.com/</a></p>
</div>
<div id="ref-12ovVgVju">
<p>37. <strong>GitHub Diffs &amp; Commenting for Jupyter Notebooks</strong> <a href="https://www.reviewnb.com/">https://www.reviewnb.com/</a></p>
</div>
<div id="ref-6YE0Zqqy">
<p>38. <strong>The WorkflowHub</strong> <a href="https://workflowhub.eu/">https://workflowhub.eu/</a></p>
</div>
<div id="ref-18tOsSTai">
<p>39. <strong>Google Docs: Free Online Documents for Personal Use</strong> <a href="https://www.google.com/docs/about/">https://www.google.com/docs/about/</a></p>
</div>
<div id="ref-sP8f9fNq">
<p>40. <strong>Microsoft 365 with Office apps | Microsoft 365</strong> <a href="https://www.microsoft.com/en-us/microsoft-365">https://www.microsoft.com/en-us/microsoft-365</a></p>
</div>
<div id="ref-UeFHiqcp">
<p>41. <strong>Open Research Collaboration and Publishing - Authorea</strong> <a href="https://www.authorea.com/">https://www.authorea.com/</a></p>
</div>
<div id="ref-J12A3elW">
<p>42. <strong>Overleaf, Online LaTeX Editor</strong> <a href="https://www.overleaf.com">https://www.overleaf.com</a></p>
</div>
<div id="ref-petulEc2">
<p>43. <strong>Manubot - Manuscripts, open and automated</strong> <a href="https://manubot.org">https://manubot.org</a></p>
</div>
<div id="ref-uRbOUixg">
<p>44. <strong>GitHub CLI</strong> <br />
GitHub CLI<br />
<a href="https://cli.github.com/">https://cli.github.com/</a></p>
</div>
<div id="ref-jIIrCDT3">
<p>45. <strong>About issues - GitHub Docs</strong> <a href="https://docs.github.com/en/github/managing-your-work-on-github/about-issues">https://docs.github.com/en/github/managing-your-work-on-github/about-issues</a></p>
</div>
<div id="ref-E7mOdOEw">
<p>46. <strong>Unit Testing for R</strong> <a href="https://testthat.r-lib.org/">https://testthat.r-lib.org/</a></p>
</div>
<div id="ref-ZNYSO1m0">
<p>47. <strong>Flake8: Your Tool For Style Guide Enforcement — flake8 3.8.4 documentation</strong> <a href="https://flake8.pycqa.org/en/latest/">https://flake8.pycqa.org/en/latest/</a></p>
</div>
<div id="ref-iFiB8Hir">
<p>48. <strong>Safety - Security for your Python dependencies</strong> <a href="https://pyup.io/safety/">https://pyup.io/safety/</a></p>
</div>
<div id="ref-T03Api6e">
<p>49. <strong>Continuous Integration and Delivery</strong> <br />
CircleCI<br />
<a href="https://circleci.com/">https://circleci.com/</a></p>
</div>
<div id="ref-FTO2RnTd">
<p>50. <strong>How to do a code review</strong> <br />
eng-practices<br />
<a href="https://google.github.io/eng-practices/review/reviewer/">https://google.github.io/eng-practices/review/reviewer/</a></p>
</div>
<div id="ref-HnmFW2rl">
<p>51. <strong>Code Review Guidelines for Humans</strong> <br />
Philipp Hauer<br />
<em>Philipp Hauer’s Blog</em> (2018-07-31) <a href="https://phauer.com/2018/code-review-guidelines/">https://phauer.com/2018/code-review-guidelines/</a></p>
</div>
<div id="ref-Z6ni32OG">
<p>52. <strong>Zenodo - Research. Shared.</strong> <a href="https://zenodo.org/">https://zenodo.org/</a></p>
</div>
<div id="ref-1EMndz0j9">
<p>53. <strong>About</strong> <br />
figshare<br />
<a href="https://figshare.com/about">https://figshare.com/about</a></p>
</div>
<div id="ref-11GtZ7icJ">
<p>54. <strong>Git Large File Storage</strong> <br />
Git Large File Storage<br />
<a href="https://git-lfs.github.com/">https://git-lfs.github.com/</a></p>
</div>
<div id="ref-varWEBV8">
<p>55. <strong>What is my disk quota? - GitHub Docs</strong> <a href="https://docs.github.com/en/github/managing-large-files/what-is-my-disk-quota">https://docs.github.com/en/github/managing-large-files/what-is-my-disk-quota</a></p>
</div>
<div id="ref-YuJbg3zO">
<p>56. <strong>Open collaborative writing with Manubot</strong> <br />
Daniel S. Himmelstein, Vincent Rubinetti, David R. Slochower, Dongbo Hu, Venkat S. Malladi, Casey S. Greene, Anthony Gitter<br />
<em>PLOS Computational Biology</em> (2019-06-24) <a href="https://doi.org/c7np">https://doi.org/c7np</a> <br />
DOI: <a href="https://doi.org/10.1371/journal.pcbi.1007128">10.1371/journal.pcbi.1007128</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31233491">31233491</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6611653">PMC6611653</a></p>
</div>
<div id="ref-16wWVekW1">
<p>57. <strong>Make a README</strong> <br />
Make a README<br />
<a href="https://www.makeareadme.com">https://www.makeareadme.com</a></p>
</div>
<div id="ref-nR2NTYy9">
<p>58. <strong>GitHub Pages</strong> <br />
GitHub Pages<br />
<a href="https://pages.github.com/">https://pages.github.com/</a></p>
</div>
<div id="ref-Lbp8MQSL">
<p>59. <strong>Licenses</strong> <br />
Choose a License<br />
<a href="https://choosealicense.com/licenses/">https://choosealicense.com/licenses/</a></p>
</div>
<div id="ref-146GfpGTF">
<p>60. <strong>Managing releases in a repository - GitHub Docs</strong> <a href="https://docs.github.com/en/github/administering-a-repository/managing-releases-in-a-repository">https://docs.github.com/en/github/administering-a-repository/managing-releases-in-a-repository</a></p>
</div>
<div id="ref-13UfgtMhu">
<p>61. <strong>Semantic Versioning 2.0.0</strong> <br />
Tom Preston-Werner<br />
<em>Semantic Versioning</em> <a href="https://semver.org/">https://semver.org/</a></p>
</div>
<div id="ref-Mduzvxei">
<p>62. <strong>PyPI · The Python Package Index</strong> <br />
PyPI<br />
<a href="https://pypi.org/">https://pypi.org/</a></p>
</div>
<div id="ref-rWFDdwIs">
<p>63. <strong>The Comprehensive R Archive Network</strong> <a href="https://cran.r-project.org/">https://cran.r-project.org/</a></p>
</div>
<div id="ref-10CJ2sJo8">
<p>64. <strong>Bioconductor - Home</strong> <a href="https://www.bioconductor.org/">https://www.bioconductor.org/</a></p>
</div>
<div id="ref-AvIrTFAn">
<p>65. <strong>News — Bioconda documentation</strong> <a href="https://bioconda.github.io/">https://bioconda.github.io/</a></p>
</div>
<div id="ref-dk80YsaO">
<p>66. <strong>Dash Overview</strong> <a href="/dash">/dash</a></p>
</div>
<div id="ref-B0TdOflx">
<p>67. <strong>Shiny</strong> <a href="https://shiny.rstudio.com/">https://shiny.rstudio.com/</a></p>
</div>
<div id="ref-LSHekw4z">
<p>68. <strong>Daring Fireball: Markdown Syntax Documentation</strong> <a href="https://daringfireball.net/projects/markdown/syntax">https://daringfireball.net/projects/markdown/syntax</a></p>
</div>
<div id="ref-RPjXsSC8">
<p>69. <strong>reStructuredText</strong> <a href="https://docutils.sourceforge.io/rst.html">https://docutils.sourceforge.io/rst.html</a></p>
</div>
<div id="ref-woYMAGdl">
<p>70. <strong>The documentation system — Documentation system documentation</strong> <a href="https://documentation.divio.com/">https://documentation.divio.com/</a></p>
</div>
<div id="ref-IsDZd09v">
<p>71. <strong>Python Docstrings</strong> <br />
GeeksforGeeks<br />
(2017-06-01) <a href="https://www.geeksforgeeks.org/python-docstrings/">https://www.geeksforgeeks.org/python-docstrings/</a></p>
</div>
<div id="ref-9HZXe8sx">
<p>72. <strong>Home | Read the Docs</strong> <a href="https://readthedocs.org/">https://readthedocs.org/</a></p>
</div>
<div id="ref-VjDIpAqq">
<p>73. <strong>GitBook - Document Everything!</strong> <a href="https://gitbook.com/">https://gitbook.com/</a></p>
</div>
<div id="ref-dvOzCOyW">
<p>74. <strong>jazzband/pip-tools</strong> <br />
Jazzband<br />
(2021-02-12) <a href="https://github.com/jazzband/pip-tools">https://github.com/jazzband/pip-tools</a></p>
</div>
<div id="ref-rmvJ2O0Q">
<p>75. <strong>Empowering App Development for Developers | Docker</strong> <a href="https://www.docker.com/">https://www.docker.com/</a></p>
</div>
<div id="ref-99QkJBoF">
<p>76. <strong>Home</strong> <br />
Sylabs.io<br />
<a href="https://sylabs.io/">https://sylabs.io/</a></p>
</div>
<div id="ref-5iKCi5fm">
<p>77. <strong>gnu.org</strong> <a href="https://www.gnu.org/software/make/">https://www.gnu.org/software/make/</a></p>
</div>
<div id="ref-q2TJoDAO">
<p>78. <strong>Snakemake — Snakemake 5.32.2 documentation</strong> <a href="https://snakemake.readthedocs.io/en/stable/">https://snakemake.readthedocs.io/en/stable/</a></p>
</div>
<div id="ref-11OYUtIpR">
<p>79. <strong>Nextflow - A DSL for parallel and scalable computational pipelines</strong> <a href="https://www.nextflow.io/">https://www.nextflow.io/</a></p>
</div>
<div id="ref-CVNjbImf">
<p>80. <strong>The Apache Groovy programming language</strong> <a href="https://groovy-lang.org/">https://groovy-lang.org/</a></p>
</div>
<div id="ref-8tGcpIB0">
<p>81. <strong>What is Software Rot? - Software Development Glossary by DeepSource</strong> <br />
DeepSource<br />
<a href="https://deepsource.io/glossary/software-rot/">https://deepsource.io/glossary/software-rot/</a></p>
</div>
<div id="ref-iXfPU3Cy">
<p>82. <strong>Gitter</strong> <a href="https://gitter.im/">https://gitter.im/</a></p>
</div>
<div id="ref-6v98AMba">
<p>83. <strong>Google Groups</strong> <a href="https://groups.google.com/forum/m/">https://groups.google.com/forum/m/</a></p>
</div>
<div id="ref-ngGdYg2r">
<p>84. <a href="https://docs.github.com/en/free-pro-team">https://docs.github.com/en/free-pro-team</a></p>
</div>
<div id="ref-5HELQ2ms">
<p>85. <strong>Mastering Issues · GitHub Guides</strong> <a href="https://guides.github.com/features/issues/">https://guides.github.com/features/issues/</a></p>
</div>
<div id="ref-ep7wwgWj">
<p>86. <strong>Setting guidelines for repository contributors - GitHub Docs</strong> <a href="https://docs.github.com/en/github/building-a-strong-community/setting-guidelines-for-repository-contributors">https://docs.github.com/en/github/building-a-strong-community/setting-guidelines-for-repository-contributors</a></p>
</div>
<div id="ref-pVl1NEXZ">
<p>87. <strong>About wikis - GitHub Docs</strong> <a href="https://docs.github.com/en/github/building-a-strong-community/about-wikis">https://docs.github.com/en/github/building-a-strong-community/about-wikis</a></p>
</div>
<div id="ref-115FgZQh6">
<p>88. <strong>Contributor Covenant: A Code of Conduct for Open Source Projects</strong> <a href="https://www.contributor-covenant.org/">https://www.contributor-covenant.org/</a></p>
</div>
<div id="ref-qcPNBi0h">
<p>89. <strong>stumpsyn/policies</strong> <br />
GitHub<br />
<a href="https://github.com/stumpsyn/policies">https://github.com/stumpsyn/policies</a></p>
</div>
<div id="ref-1Ei7XPHmQ">
<p>90. <strong>A successful Git branching model</strong> <br />
nvie.com<br />
<a href="http://nvie.com/posts/a-successful-git-branching-model/">http://nvie.com/posts/a-successful-git-branching-model/</a></p>
</div>
<div id="ref-zcUBK27J">
<p>91. <strong>Kanban - A brief introduction</strong> <br />
Atlassian<br />
<em>Atlassian</em> <a href="https://www.atlassian.com/agile/kanban">https://www.atlassian.com/agile/kanban</a></p>
</div>
<div id="ref-ssoXtriL">
<p>92. <strong>GitHub features: Integrated project management tools</strong> <br />
GitHub<br />
<a href="https://github.com/features/project-management">https://github.com/features/project-management</a></p>
</div>
<div id="ref-QrHcaBPj">
<p>93. <strong>Free Kanban Boards - Issue Tracker | GitKraken Boards</strong> <br />
GitKraken.com<br />
<a href="https://www.gitkraken.com/boards">https://www.gitkraken.com/boards</a></p>
</div>
<div id="ref-LLiWj74b">
<p>94. <strong>What is Scrum?</strong> <br />
Scrum.org<br />
<a href="https://www.scrum.org/resources/what-is-scrum">https://www.scrum.org/resources/what-is-scrum</a></p>
</div>
<div id="ref-rX7VZWos">
<p>95. <strong>ZenHub - Agile Project Management for GitHub</strong> <a href="https://www.zenhub.com/">https://www.zenhub.com/</a></p>
</div>
<div id="ref-a6SmwGuV">
<p>96. <strong>Jira | Issue &amp; Project Tracking Software</strong> <br />
Atlassian<br />
<em>Atlassian</em> <a href="https://www.atlassian.com/software/jira">https://www.atlassian.com/software/jira</a></p>
</div>
<div id="ref-qOcBF2oS">
<p>97. <strong>Manifesto for Agile Software Development</strong> <a href="https://agilemanifesto.org/">https://agilemanifesto.org/</a></p>
</div>
<div id="ref-1UVh4XJi">
<p>98. <strong>c4urself/bump2version</strong> <br />
Christian Verkerk<br />
(2021-02-13) <a href="https://github.com/c4urself/bump2version">https://github.com/c4urself/bump2version</a></p>
</div>
<div id="ref-SXI7Canc">
<p>99. <strong>Danger - Stop Saying “You Forgot To…” in Code Review</strong> <br />
Orta Therox<br />
<a href="http://danger.systems/ruby/index.html">http://danger.systems/ruby/index.html</a></p>
</div>
</div>
<!-- default theme -->

<style>
    /* import google fonts */
    @import url("https://fonts.googleapis.com/css?family=Open+Sans:400,600,700");
    @import url("https://fonts.googleapis.com/css?family=Source+Code+Pro");

    /* -------------------------------------------------- */
    /* global */
    /* -------------------------------------------------- */

    /* all elements */
    * {
        /* force sans-serif font unless specified otherwise */
        font-family: "Open Sans", "Helvetica", sans-serif;

        /* prevent text inflation on some mobile browsers */
        -webkit-text-size-adjust: none !important;
        -moz-text-size-adjust: none !important;
        -o-text-size-adjust: none !important;
        text-size-adjust: none !important;
    }

    @media only screen {
        /* "page" element */
        body {
            position: relative;
            box-sizing: border-box;
            font-size: 12pt;
            line-height: 1.5;
            max-width: 8.5in;
            margin: 20px auto;
            padding: 40px;
            border-radius: 5px;
            border: solid 1px #bdbdbd;
            box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
            background: #ffffff;
        }
    }

    /* when on screen < 8.5in wide */
    @media only screen and (max-width: 8.5in) {
        /* "page" element */
        body {
            padding: 20px;
            margin: 0;
            border-radius: 0;
            border: none;
            box-shadow: 0 0 20px rgba(0, 0, 0, 0.05) inset;
            background: none;
        }
    }

    /* -------------------------------------------------- */
    /* headings */
    /* -------------------------------------------------- */

    /* all headings */
    h1,
    h2,
    h3,
    h4,
    h5,
    h6 {
        margin: 20px 0;
        padding: 0;
        font-weight: bold;
    }

    /* biggest heading */
    h1 {
        margin: 40px 0;
        text-align: center;
    }

    /* second biggest heading */
    h2 {
        margin-top: 30px;
        padding-bottom: 5px;
        border-bottom: solid 1px #bdbdbd;
    }

    /* heading font sizes */
    h1 {
        font-size: 2em;
    }
    h2 {
        font-size: 1.5em;
    }
    h3{
        font-size: 1.35em;
    }
    h4 {
        font-size: 1.25em;
    }
    h5 {
        font-size: 1.15em;
    }
    h6 {
        font-size: 1em;
    }

    /* -------------------------------------------------- */
    /* manuscript header */
    /* -------------------------------------------------- */

    /* manuscript title */
    header > h1 {
        margin: 0;
    }

    /* manuscript title caption text (ie "automatically generated on") */
    header + p {
        text-align: center;
        margin-top: 10px;
    }

    /* -------------------------------------------------- */
    /* text elements */
    /* -------------------------------------------------- */

    /* links */
    a {
        color: #2196f3;
        overflow-wrap: break-word;
    }

    /* normal links (not empty, not button link, not syntax highlighting link) */
    a:not(:empty):not(.button):not(.sourceLine) {
        padding-left: 1px;
        padding-right: 1px;
    }

    /* superscripts and subscripts */
    sub,
    sup {
        /* prevent from affecting line height */
        line-height: 0;
    }

    /* unordered and ordered lists*/
    ul,
    ol {
        padding-left: 20px;
    }

    /* class for styling text semibold */
    .semibold {
        font-weight: 600;
    }

    /* class for styling elements horizontally left aligned */
    .left {
        display: block;
        text-align: left;
        margin-left: auto;
        margin-right: 0;
        justify-content: left;
    }

    /* class for styling elements horizontally centered */
    .center {
        display: block;
        text-align: center;
        margin-left: auto;
        margin-right: auto;
        justify-content: center;
    }

    /* class for styling elements horizontally right aligned */
    .right {
        display: block;
        text-align: right;
        margin-left: 0;
        margin-right: auto;
        justify-content: right;
    }

    /* -------------------------------------------------- */
    /* section elements */
    /* -------------------------------------------------- */

    /* horizontal divider line */
    hr {
        border: none;
        height: 1px;
        background: #bdbdbd;
    }

    /* paragraphs, horizontal dividers, figures, tables, code */
    p,
    hr,
    figure,
    table,
    pre {
        /* treat all as "paragraphs", with consistent vertical margins */
        margin-top: 20px;
        margin-bottom: 20px;
    }

    /* -------------------------------------------------- */
    /* figures */
    /* -------------------------------------------------- */

    /* figure */
    figure {
        max-width: 100%;
        margin-left: auto;
        margin-right: auto;
    }

    /* figure caption */
    figcaption {
        padding: 0;
        padding-top: 10px;
    }

    /* figure image element */
    figure > img,
    figure > svg {
        max-width: 100%;
        display: block;
        margin-left: auto;
        margin-right: auto;
    }

    /* figure auto-number */
    img + figcaption > span:first-of-type,
    svg + figcaption > span:first-of-type {
        font-weight: bold;
        margin-right: 5px;
    }

    /* -------------------------------------------------- */
    /* tables */
    /* -------------------------------------------------- */

    /* table */
    table {
        border-collapse: collapse;
        border-spacing: 0;
        width: 100%;
        margin-left: auto;
        margin-right: auto;
    }

    /* table cells */
    th,
    td {
        border: solid 1px #bdbdbd;
        padding: 10px;
        /* squash table if too wide for page by forcing line breaks */
        overflow-wrap: break-word;
        word-break: break-word;
    }

    /* header row and even rows */
    th,
    tr:nth-child(2n) {
        background-color: #fafafa;
    }

    /* odd rows */
    tr:nth-child(2n + 1) {
        background-color: #ffffff;
    }

    /* table caption */
    caption {
        text-align: left;
        padding: 0;
        padding-bottom: 10px;
    }

    /* table auto-number */
    table > caption > span:first-of-type,
    div.table_wrapper > table > caption > span:first-of-type {
        font-weight: bold;
        margin-right: 5px;
    }

    /* -------------------------------------------------- */
    /* code */
    /* -------------------------------------------------- */

    /* multi-line code block */
    pre {
        padding: 10px;
        background-color: #eeeeee;
        color: #000000;
        border-radius: 5px;
        break-inside: avoid;
        text-align: left;
    }

    /* inline code, ie code within normal text */
    :not(pre) > code {
        padding: 0 4px;
        background-color: #eeeeee;
        color: #000000;
        border-radius: 5px;
    }

    /* code text */
    /* apply all children, to reach syntax highlighting sub-elements */
    code,
    code * {
        /* force monospace font */
        font-family: "Source Code Pro", "Courier New", monospace;
    }

    /* -------------------------------------------------- */
    /* quotes */
    /* -------------------------------------------------- */

    /* quoted text */
    blockquote {
        margin: 0;
        padding: 0;
        border-left: 4px solid #bdbdbd;
        padding-left: 16px;
        break-inside: avoid;
    }

    /* -------------------------------------------------- */
    /* banners */
    /* -------------------------------------------------- */

    /* info banners */
    .banner {
        box-sizing: border-box;
        display: block;
        position: relative;
        width: 100%;
        margin-top: 20px;
        margin-bottom: 20px;
        padding: 20px;
        text-align: center;
    }

    /* paragraph in banner */
    .banner > p {
        margin: 0;
    }

    /* -------------------------------------------------- */
    /* highlight colors */
    /* -------------------------------------------------- */

    .white {
        background: #ffffff;
    }
    .lightgrey {
        background: #eeeeee;
    }
    .grey {
        background: #757575;
    }
    .darkgrey {
        background: #424242;
    }
    .black {
        background: #000000;
    }
    .lightred {
        background: #ffcdd2;
    }
    .lightyellow {
        background: #ffecb3;
    }
    .lightgreen {
        background: #dcedc8;
    }
    .lightblue {
        background: #e3f2fd;
    }
    .lightpurple {
        background: #f3e5f5;
    }
    .red {
        background: #f44336;
    }
    .orange {
        background: #ff9800;
    }
    .yellow {
        background: #ffeb3b;
    }
    .green {
        background: #4caf50;
    }
    .blue {
        background: #2196f3;
    }
    .purple {
        background: #9c27b0;
    }
    .white,
    .lightgrey,
    .lightred,
    .lightyellow,
    .lightgreen,
    .lightblue,
    .lightpurple,
    .orange,
    .yellow,
    .white a,
    .lightgrey a,
    .lightred a,
    .lightyellow a,
    .lightgreen a,
    .lightblue a,
    .lightpurple a,
    .orange a,
    .yellow a {
        color: #000000;
    }
    .grey,
    .darkgrey,
    .black,
    .red,
    .green,
    .blue,
    .purple,
    .grey a,
    .darkgrey a,
    .black a,
    .red a,
    .green a,
    .blue a,
    .purple a {
        color: #ffffff;
    }

    /* -------------------------------------------------- */
    /* buttons */
    /* -------------------------------------------------- */

    /* class for styling links like buttons */
    .button {
        display: inline-flex;
        justify-content: center;
        align-items: center;
        margin: 5px;
        padding: 10px 20px;
        font-size: 0.75em;
        font-weight: 600;
        text-transform: uppercase;
        text-decoration: none;
        letter-spacing: 1px;
        background: none;
        color: #2196f3;
        border: solid 1px #bdbdbd;
        border-radius: 5px;
    }

    /* buttons when hovered */
    .button:hover:not([disabled]),
    .icon_button:hover:not([disabled]) {
        cursor: pointer;
        background: #f5f5f5;
    }

    /* buttons when disabled */
    .button[disabled],
    .icon_button[disabled] {
        opacity: 0.35;
        pointer-events: none;
    }

    /* class for styling buttons containg only single icon */
    .icon_button {
        display: inline-flex;
        justify-content: center;
        align-items: center;
        text-decoration: none;
        margin: 0;
        padding: 0;
        background: none;
        border-radius: 5px;
        border: none;
        width: 20px;
        height: 20px;
        min-width: 20px;
        min-height: 20px;
    }

    /* icon button inner svg image */
    .icon_button > svg {
        height: 16px;
    }

    /* -------------------------------------------------- */
    /* icons */
    /* -------------------------------------------------- */

    /* class for styling icons inline with text */
    .inline_icon {
        height: 1em;
        position: relative;
        top: 0.125em;
    }

    /* -------------------------------------------------- */
    /* print control */
    /* -------------------------------------------------- */

    @media print {
        @page {
            /* suggested printing margin */
            margin: 0.5in;
        }

        /* document and "page" elements */
        html, body {
            margin: 0;
            padding: 0;
            width: 100%;
            height: 100%;
        }

        /* "page" element */
        body {
            font-size: 11pt !important;
            line-height: 1.35;
        }

        /* all headings */
        h1,
        h2,
        h3,
        h4,
        h5,
        h6 {
            margin: 15px 0;
        }

        /* figures and tables */
        figure, table {
            font-size: 0.85em;
        }

        /* table cells */
        th,
        td {
            padding: 5px;
        }

        /* shrink font awesome icons */
        i.fas,
        i.fab,
        i.far,
        i.fal {
            transform: scale(0.85);
        }

        /* decrease banner margins */
        .banner {
            margin-top: 15px;
            margin-bottom: 15px;
            padding: 15px;
        }

        /* class for centering an element vertically on its own page */
        .page_center {
            margin: auto;
            width: 100%;
            height: 100%;
            display: flex;
            align-items: center;
            vertical-align: middle;
            break-before: page;
            break-after: page;
        }

        /* always insert a page break before the element */
        .page_break_before {
            break-before: page;
        }

        /* always insert a page break after the element */
        .page_break_after {
            break-after: page;
        }

        /* avoid page break before the element */
        .page_break_before_avoid {
            break-before: avoid;
        }

        /* avoid page break after the element */
        .page_break_after_avoid {
            break-after: avoid;
        }

        /* avoid page break inside the element */
        .page_break_inside_avoid {
            break-inside: avoid;
        }
    }

    /* -------------------------------------------------- */
    /* override pandoc css quirks */
    /* -------------------------------------------------- */

    .sourceCode {
        /* prevent unsightly overflow in wide code blocks */
        overflow: auto !important;
    }

    div.sourceCode {
        /* prevent background fill on top-most code block  container */
        background: none !important;
    }

    .sourceCode * {
        /* force consistent line spacing */
        line-height: 1.5 !important;
    }

    div.sourceCode {
        /* style code block margins same as <pre> element */
        margin-top: 20px;
        margin-bottom: 20px;
    }

    /* -------------------------------------------------- */
    /* tablenos */
    /* -------------------------------------------------- */

    /* tablenos wrapper */
    .tablenos {
        /* show scrollbar on tables if necessary to prevent overflow */
        width: 100%;
        margin: 20px 0;
    }

    .tablenos > table {
        /* move margins from table to table_wrapper to allow margin collapsing */
        margin: 0;
    }

    @media only screen {
        /* tablenos wrapper */
        .tablenos {
            /* show scrollbar on tables if necessary to prevent overflow */
            overflow-x: auto !important;
        }

        .tablenos th,
        .tablenos td {
            overflow-wrap: unset !important;
            word-break: unset !important;
        }

        /* table in wrapper */
        .tablenos table,
        .tablenos table * {
            /* don't break table words */
            overflow-wrap: normal !important;
        }
    }

    /* -------------------------------------------------- */
    /* mathjax */
    /* -------------------------------------------------- */

    /* mathjax containers */
    .math.display > span:not(.MathJax_Preview) {
        /* turn inline element (no dimensions) into block (allows fixed width and thus scrolling) */
        display: flex !important;
        overflow-x: auto !important;
        overflow-y: hidden !important;
        justify-content: center;
        align-items: center;
        margin: 0 !important;
    }

    /* right click menu */
    .MathJax_Menu {
        border-radius: 5px !important;
        border: solid 1px #bdbdbd !important;
        box-shadow: none !important;
    }

    /* equation auto-number */
    span[id^="eq:"] > span.math.display + span {
        font-weight: 600;
    }

    /* equation */
    span[id^="eq:"] > span.math.display > span {
        /* nudge to make room for equation auto-number and anchor */
        margin-right: 60px !important;
    }

    /* -------------------------------------------------- */
    /* anchors plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* anchor button */
        .anchor {
            opacity: 0;
            margin-left: 5px;
        }

        /* anchor buttons within <h2>'s */
        h2 .anchor {
            margin-left: 10px;
        }

        /* anchor buttons when hovered/focused and anything containing an anchor button when hovered */
        *:hover > .anchor,
        .anchor:hover,
        .anchor:focus {
            opacity: 1;
        }

        /* anchor button when hovered */
        .anchor:hover {
            cursor: pointer;
        }
    }

    /* always show anchor button on devices with no mouse/hover ability */
    @media (hover: none) {
        .anchor {
            opacity: 1;
        }
    }

    /* always hide anchor button on print */
    @media only print {
        .anchor {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* accordion plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* accordion arrow button */
        .accordion_arrow {
            margin-right: 10px;
        }

        /* arrow icon when <h2> data-collapsed attribute true */
        h2[data-collapsed="true"] > .accordion_arrow > svg {
            transform: rotate(-90deg);
        }

        /* all elements (except <h2>'s) when data-collapsed attribute true */
        *:not(h2)[data-collapsed="true"] {
            display: none;
        }

        /* accordion arrow button when hovered and <h2>'s when hovered */
        .accordion_arrow:hover,
        h2[data-collapsed="true"]:hover,
        h2[data-collapsed="false"]:hover {
            cursor: pointer;
        }
    }

    /* always hide accordion arrow button on print */
    @media only print {
        .accordion_arrow {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* tooltips plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* tooltip container */
        #tooltip {
            position: absolute;
            width: 50%;
            min-width: 240px;
            max-width: 75%;
            z-index: 1;
        }

        /* tooltip content */
        #tooltip_content {
            margin-bottom: 5px;
            padding: 20px;
            border-radius: 5px;
            border: solid 1px #bdbdbd;
            box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
            background: #ffffff;
            overflow-wrap: break-word;
        }

        /* tooltip copy of paragraphs and figures */
        #tooltip_content > p,
        #tooltip_content > figure {
            margin: 0;
            max-height: 320px;
            overflow-y: auto;
        }

        /* tooltip copy of <img> */
        #tooltip_content > figure > img,
        #tooltip_content > figure > svg {
            max-height: 260px;
        }

        /* navigation bar */
        #tooltip_nav_bar {
            margin-top: 10px;
            text-align: center;
        }

        /* navigation bar previous/next buton */
        #tooltip_nav_bar > .icon_button {
            position: relative;
            top: 3px;
        }

        /* navigation bar previous button */
        #tooltip_nav_bar > .icon_button:first-of-type {
            margin-right: 5px;
        }

        /* navigation bar next button */
        #tooltip_nav_bar > .icon_button:last-of-type {
            margin-left: 5px;
        }
    }

    /* always hide tooltip on print */
    @media only print {
        #tooltip {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* jump to first plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* jump button */
        .jump_arrow {
            position: relative;
            top: 0.125em;
            margin-right: 5px;
        }
    }

    /* always hide jump button on print */
    @media only print {
        .jump_arrow {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* link highlight plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* anything with data-highlighted attribute true */
        [data-highlighted="true"] {
            background: #ffeb3b;
        }

        /* anything with data-selected attribute true */
        [data-selected="true"] {
            background: #ff8a65 !important;
        }

        /* animation definition for glow */
        @keyframes highlight_glow {
            0% {
                background: none;
            }
            10% {
                background: #bbdefb;
            }
            100% {
                background: none;
            }
        }

        /* anything with data-glow attribute true */
        [data-glow="true"] {
            animation: highlight_glow 2s;
        }
    }

    /* -------------------------------------------------- */
    /* table of contents plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* toc panel */
        #toc_panel {
            box-sizing: border-box;
            position: fixed;
            top: 0;
            left: 0;
            background: #ffffff;
            box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
            z-index: 2;
        }

        /* toc panel when closed */
        #toc_panel[data-open="false"] {
            min-width: 60px;
            width: 60px;
            height: 60px;
            border-right: solid 1px #bdbdbd;
            border-bottom: solid 1px #bdbdbd;
        }

        /* toc panel when open */
        #toc_panel[data-open="true"] {
            min-width: 260px;
            max-width: 480px;
            /* keep panel edge consistent distance away from "page" edge */
            width: calc(((100vw - 8.5in) / 2) - 30px - 40px);
            bottom: 0;
            border-right: solid 1px #bdbdbd;
        }

        /* toc panel header */
        #toc_header {
            box-sizing: border-box;
            display: flex;
            flex-direction: row;
            align-items: center;
            height: 60px;
            margin: 0;
            padding: 20px;
        }

        /* toc panel header when hovered */
        #toc_header:hover {
            cursor: pointer;
        }

        /* toc panel header when panel open */
        #toc_panel[data-open="true"] > #toc_header {
            border-bottom: solid 1px #bdbdbd;
        }

        /* toc open/close header button */
        #toc_button {
            margin-right: 20px;
        }

        /* hide toc list and header text when closed */
        #toc_panel[data-open="false"] > #toc_header > *:not(#toc_button),
        #toc_panel[data-open="false"] > #toc_list {
            display: none;
        }

        /* toc list of entries */
        #toc_list {
            box-sizing: border-box;
            width: 100%;
            padding: 20px;
            position: absolute;
            top: calc(60px + 1px);
            bottom: 0;
            overflow: auto;
        }

        /* toc entry, link to section in document */
        .toc_link {
            display: block;
            padding: 5px;
            position: relative;
            font-weight: 600;
            text-decoration: none;
        }

        /* toc entry when hovered or when "viewed" */
        .toc_link:hover,
        .toc_link[data-viewing="true"] {
            background: #f5f5f5;
        }

        /* toc entry, level 1 indentation */
        .toc_link[data-level="1"] {
            margin-left: 0;
        }

        /* toc entry, level 2 indentation */
        .toc_link[data-level="2"] {
            margin-left: 20px;
        }

        /* toc entry, level 3 indentation */
        .toc_link[data-level="3"] {
            margin-left: 40px;
        }

        /* toc entry, level 4 indentation */
        .toc_link[data-level="4"] {
            margin-left: 60px;
        }

        /* toc entry bullets */
        #toc_panel[data-bullets="true"] .toc_link[data-level]:before {
            position: absolute;
            left: -15px;
            top: -1px;
            font-size: 1.5em;
        }

        /* toc entry, level 2 bullet */
        #toc_panel[data-bullets="true"] .toc_link[data-level="2"]:before {
            content: "\2022";
        }

        /* toc entry, level 3 bullet */
        #toc_panel[data-bullets="true"] .toc_link[data-level="3"]:before {
            content: "\25AB";
        }

        /* toc entry, level 4 bullet */
        #toc_panel[data-bullets="true"] .toc_link[data-level="4"]:before {
            content: "-";
        }
    }

    /* when on screen < 8.5in wide */
    @media only screen and (max-width: 8.5in) {
        /* push <body> ("page") element down to make room for toc icon */
        .toc_body_nudge {
            padding-top: 60px;
        }

        /* toc icon when panel closed and not hovered */
        #toc_panel[data-open="false"]:not(:hover) {
            background: rgba(255, 255, 255, 0.75);
        }
    }

    /* always hide toc panel on print */
    @media only print {
        #toc_panel {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* lightbox plugin */
    /* -------------------------------------------------- */

    @media only screen {
        /* regular <img> in document when hovered */
        img.lightbox_document_img:hover {
            cursor: pointer;
        }

        .body_no_scroll {
            overflow: hidden !important;
        }

        /* screen overlay */
        #lightbox_overlay {
            display: flex;
            flex-direction: column;
            position: fixed;
            left: 0;
            top: 0;
            right: 0;
            bottom: 0;
            background: rgba(0, 0, 0, 0.75);
            z-index: 3;
        }

        /* middle area containing lightbox image */
        #lightbox_image_container {
            flex-grow: 1;
            display: flex;
            justify-content: center;
            align-items: center;
            overflow: hidden;
            position: relative;
            padding: 20px;
        }

        /* bottom area containing caption */
        #lightbox_bottom_container {
            display: flex;
            justify-content: center;
            align-items: center;
            height: 100px;
            min-height: 100px;
            max-height: 100px;
            background: rgba(0, 0, 0, 0.5);
        }

        /* image number info text box */
        #lightbox_number_info {
            position: absolute;
            color: #ffffff;
            font-weight: 600;
            left: 2px;
            top: 0;
            z-index: 4;
        }

        /* zoom info text box */
        #lightbox_zoom_info {
            position: absolute;
            color: #ffffff;
            font-weight: 600;
            right: 2px;
            top: 0;
            z-index: 4;
        }

        /* copy of image caption */
        #lightbox_caption {
            box-sizing: border-box;
            display: inline-block;
            width: 100%;
            max-height: 100%;
            padding: 10px 0;
            text-align: center;
            overflow-y: auto;
            color: #ffffff;
        }

        /* navigation previous/next button */
        .lightbox_button {
            width: 100px;
            height: 100%;
            min-width: 100px;
            min-height: 100%;
            color: #ffffff;
        }

        /* navigation previous/next button when hovered */
        .lightbox_button:hover {
            background: none !important;
        }

        /* navigation button icon */
        .lightbox_button > svg {
            height: 25px;
        }

        /* figure auto-number */
        #lightbox_caption > span:first-of-type {
            font-weight: bold;
            margin-right: 5px;
        }

        /* lightbox image when hovered */
        #lightbox_img:hover {
            cursor: grab;
        }

        /* lightbox image when grabbed */
        #lightbox_img:active {
            cursor: grabbing;
        }
    }

    /* when on screen < 480px wide */
    @media only screen and (max-width: 480px) {
        /* make navigation buttons skinnier on small screens to make more room for caption text */
        .lightbox_button {
            width: 50px;
            min-width: 50px;
        }
    }

    /* always hide lightbox on print */
    @media only print {
        #lightbox_overlay {
            display: none;
        }
    }

    /* -------------------------------------------------- */
    /* hypothesis (annotations) plugin */
    /* -------------------------------------------------- */

    /* hypothesis activation button */
    #hypothesis_button {
        box-sizing: border-box;
        position: fixed;
        top: 0;
        right: 0;
        width: 60px;
        height: 60px;
        background: #ffffff;
        border-radius: 0;
        border-left: solid 1px #bdbdbd;
        border-bottom: solid 1px #bdbdbd;
        box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
        z-index: 2;
    }

    /* hypothesis button svg */
    #hypothesis_button > svg {
        position: relative;
        top: -4px;
    }

    /* hypothesis annotation count */
    #hypothesis_count {
        position: absolute;
        left: 0;
        right: 0;
        bottom: 5px;
    }

    /* side panel */
    .annotator-frame {
        width: 280px !important;
    }

    /* match highlight color to rest of theme */
    .annotator-highlights-always-on .annotator-hl {
        background-color: #ffeb3b !important;
    }

    /* match focused color to rest of theme */
    .annotator-hl.annotator-hl-focused {
        background-color: #ff8a65 !important;
    }

    /* match bucket bar color to rest of theme */
    .annotator-bucket-bar {
        background: #f5f5f5 !important;
    }

    /* always hide button, toolbar, and tooltip on print */
    @media only print {
        #hypothesis_button {
            display: none;
        }

        .annotator-frame {
            display: none !important;
        }

        hypothesis-adder {
            display: none !important;
        }
    }
</style>
<!-- anchors plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin adds an anchor next to each of a certain type
        // of element that provides a human-readable url to that specific
        // item/position in the document (eg "manuscript.html#abstract"). It
        // also makes it such that scrolling out of view of a target removes
        // its identifier from the url.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'anchors';

        // default plugin options
        const options = {
            // which types of elements to add anchors next to, in
            // "document.querySelector" format
            typesQuery: 'h1, h2, h3, [id^="fig:"], [id^="tbl:"], [id^="eq:"]',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            // add anchor to each element of specified types
            const elements = document.querySelectorAll(options.typesQuery);
            for (const element of elements)
                addAnchor(element);

            // attach scroll listener to window
            window.addEventListener('scroll', onScroll);
        }

        // when window is scrolled
        function onScroll() {
            // if url has hash and user has scrolled out of view of hash
            // target, remove hash from url
            const tolerance = 100;
            const target = getHashTarget();
            if (target) {
                if (
                    target.getBoundingClientRect().top >
                        window.innerHeight + tolerance ||
                    target.getBoundingClientRect().bottom < 0 - tolerance
                )
                    history.pushState(null, null, ' ');
            }
        }

        // add anchor to element
        function addAnchor(element) {
            let addTo; // element to add anchor button to

            // if figure or table, modify withId and addTo to get expected
            // elements
            if (element.id.indexOf('fig:') === 0) {
                addTo = element.querySelector('figcaption');
            } else if (element.id.indexOf('tbl:') === 0) {
                addTo = element.querySelector('caption');
            } else if (element.id.indexOf('eq:') === 0) {
                addTo = element.querySelector('.eqnos-number');
            }

            addTo = addTo || element;
            const id = element.id || null;

            // do not add anchor if element doesn't have assigned id.
            // id is generated by pandoc and is assumed to be unique and
            // human-readable
            if (!id)
                return;

            // create anchor button
            const anchor = document.createElement('a');
            anchor.innerHTML = document.querySelector('.icon_link').innerHTML;
            anchor.title = 'Link to this part of the document';
            anchor.classList.add('icon_button', 'anchor');
            anchor.dataset.ignore = 'true';
            anchor.href = '#' + id;
            addTo.appendChild(anchor);
        }

        // get element that is target of link or url hash
        function getHashTarget() {
            const hash = window.location.hash;
            const id = hash.slice(1);
            let target = document.querySelector('[id="' + id + '"]');
            if (!target)
                return;

            // if figure or table, modify target to get expected element
            if (id.indexOf('fig:') === 0)
                target = target.querySelector('figure');
            if (id.indexOf('tbl:') === 0)
                target = target.querySelector('table');

            return target;
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- link icon -->

<template class="icon_link">
    <!-- modified from: https://fontawesome.com/icons/link -->
    <svg width="16" height="16" viewBox="0 0 512 512">
        <path
            fill="currentColor"
            d="M326.612 185.391c59.747 59.809 58.927 155.698.36 214.59-.11.12-.24.25-.36.37l-67.2 67.2c-59.27 59.27-155.699 59.262-214.96 0-59.27-59.26-59.27-155.7 0-214.96l37.106-37.106c9.84-9.84 26.786-3.3 27.294 10.606.648 17.722 3.826 35.527 9.69 52.721 1.986 5.822.567 12.262-3.783 16.612l-13.087 13.087c-28.026 28.026-28.905 73.66-1.155 101.96 28.024 28.579 74.086 28.749 102.325.51l67.2-67.19c28.191-28.191 28.073-73.757 0-101.83-3.701-3.694-7.429-6.564-10.341-8.569a16.037 16.037 0 0 1-6.947-12.606c-.396-10.567 3.348-21.456 11.698-29.806l21.054-21.055c5.521-5.521 14.182-6.199 20.584-1.731a152.482 152.482 0 0 1 20.522 17.197zM467.547 44.449c-59.261-59.262-155.69-59.27-214.96 0l-67.2 67.2c-.12.12-.25.25-.36.37-58.566 58.892-59.387 154.781.36 214.59a152.454 152.454 0 0 0 20.521 17.196c6.402 4.468 15.064 3.789 20.584-1.731l21.054-21.055c8.35-8.35 12.094-19.239 11.698-29.806a16.037 16.037 0 0 0-6.947-12.606c-2.912-2.005-6.64-4.875-10.341-8.569-28.073-28.073-28.191-73.639 0-101.83l67.2-67.19c28.239-28.239 74.3-28.069 102.325.51 27.75 28.3 26.872 73.934-1.155 101.96l-13.087 13.087c-4.35 4.35-5.769 10.79-3.783 16.612 5.864 17.194 9.042 34.999 9.69 52.721.509 13.906 17.454 20.446 27.294 10.606l37.106-37.106c59.271-59.259 59.271-155.699.001-214.959z"
        ></path>
    </svg>
</template>
<!-- accordion plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin allows sections of content under <h2> headings
        // to be collapsible.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'accordion';

        // default plugin options
        const options = {
            // whether to always start expanded ('false'), always start
            // collapsed ('true'), or start collapsed when screen small ('auto')
            startCollapsed: 'auto',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            // run through each <h2> heading
            const headings = document.querySelectorAll('h2');
            for (const heading of headings) {
                addArrow(heading);

                // start expanded/collapsed based on option
                if (
                    options.startCollapsed === 'true' ||
                    (options.startCollapsed === 'auto' && isSmallScreen())
                )
                    collapseHeading(heading);
                else
                    expandHeading(heading);
            }

            // attach hash change listener to window
            window.addEventListener('hashchange', onHashChange);
        }

        // when hash (eg manuscript.html#introduction) changes
        function onHashChange() {
            const target = getHashTarget();
            if (target)
                goToElement(target);
        }

        // add arrow to heading
        function addArrow(heading) {
            // add arrow button
            const arrow = document.createElement('button');
            arrow.innerHTML = document.querySelector(
                '.icon_angle_down'
            ).innerHTML;
            arrow.classList.add('icon_button', 'accordion_arrow');
            heading.insertBefore(arrow, heading.firstChild);

            // attach click listener to heading and button
            heading.addEventListener('click', onHeadingClick);
            arrow.addEventListener('click', onArrowClick);
        }

        // determine if on mobile-like device with small screen
        function isSmallScreen() {
            return Math.min(window.innerWidth, window.innerHeight) < 480;
        }

        // scroll to and focus element
        function goToElement(element, offset) {
            // expand accordion section if collapsed
            expandElement(element);
            const y =
                getRectInView(element).top -
                getRectInView(document.documentElement).top -
                (offset || 0);
            // trigger any function listening for "onscroll" event
            window.dispatchEvent(new Event('scroll'));
            window.scrollTo(0, y);
            document.activeElement.blur();
            element.focus();
        }

        // get element that is target of hash
        function getHashTarget(link) {
            const hash = link ? link.hash : window.location.hash;
            const id = hash.slice(1);
            let target = document.querySelector('[id="' + id + '"]');
            if (!target)
                return;

            // if figure or table, modify target to get expected element
            if (id.indexOf('fig:') === 0)
                target = target.querySelector('figure');
            if (id.indexOf('tbl:') === 0)
                target = target.querySelector('table');

            return target;
        }

        // when <h2> heading is clicked
        function onHeadingClick(event) {
            // only collapse if <h2> itself is target of click (eg, user did
            // not click on anchor within <h2>)
            if (event.target === this)
                toggleCollapse(this);
        }

        // when arrow button is clicked
        function onArrowClick() {
            toggleCollapse(this.parentNode);
        }

        // collapse section if expanded, expand if collapsed
        function toggleCollapse(heading) {
            if (heading.dataset.collapsed === 'false')
                collapseHeading(heading);
            else
                expandHeading(heading);
        }

        // elements to exclude from collapse, such as table of contents panel,
        // hypothesis panel, etc
        const exclude = '#toc_panel, div.annotator-frame, #lightbox_overlay';

        // collapse section
        function collapseHeading(heading) {
            heading.setAttribute('data-collapsed', 'true');
            const children = getChildren(heading);
            for (const child of children)
                child.setAttribute('data-collapsed', 'true');
        }

        // expand section
        function expandHeading(heading) {
            heading.setAttribute('data-collapsed', 'false');
            const children = getChildren(heading);
            for (const child of children)
                child.setAttribute('data-collapsed', 'false');
        }

        // get list of elements between this <h2> and next <h2> or <h1>
        // ("children" of the <h2> section)
        function getChildren(heading) {
            return nextUntil(heading, 'h2, h1', exclude);
        }

        // get position/dimensions of element or viewport
        function getRectInView(element) {
            let rect = {};
            rect.left = 0;
            rect.top = 0;
            rect.right = document.documentElement.clientWidth;
            rect.bottom = document.documentElement.clientHeight;
            let style = {};

            if (element instanceof HTMLElement) {
                rect = element.getBoundingClientRect();
                style = window.getComputedStyle(element);
            }

            const margin = {};
            margin.left = parseFloat(style.marginLeftWidth) || 0;
            margin.top = parseFloat(style.marginTopWidth) || 0;
            margin.right = parseFloat(style.marginRightWidth) || 0;
            margin.bottom = parseFloat(style.marginBottomWidth) || 0;

            const border = {};
            border.left = parseFloat(style.borderLeftWidth) || 0;
            border.top = parseFloat(style.borderTopWidth) || 0;
            border.right = parseFloat(style.borderRightWidth) || 0;
            border.bottom = parseFloat(style.borderBottomWidth) || 0;

            const newRect = {};
            newRect.left = rect.left + margin.left + border.left;
            newRect.top = rect.top + margin.top + border.top;
            newRect.right = rect.right + margin.right + border.right;
            newRect.bottom = rect.bottom + margin.bottom + border.bottom;
            newRect.width = newRect.right - newRect.left;
            newRect.height = newRect.bottom - newRect.top;

            return newRect;
        }

        // get list of elements after a start element up to element matching
        // query
        function nextUntil(element, query, exclude) {
            const elements = [];
            while (element = element.nextElementSibling, element) {
                if (element.matches(query))
                    break;
                if (!element.matches(exclude))
                    elements.push(element);
            }
            return elements;
        }

        // get closest element before specified element that matches query
        function firstBefore(element, query) {
            while (
                element &&
                element !== document.body &&
                !element.matches(query)
            )
                element = element.previousElementSibling || element.parentNode;

            return element;
        }

        // check if element is part of collapsed heading
        function isCollapsed(element) {
            while (element && element !== document.body) {
                if (element.dataset.collapsed === 'true')
                    return true;
                element = element.parentNode;
            }
            return false;
        }

        // expand heading containing element if necesary
        function expandElement(element) {
            if (isCollapsed(element)) {
                const heading = firstBefore(element, 'h2');
                if (heading)
                    heading.click();
            }
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- angle down icon -->

<template class="icon_angle_down">
    <!-- modified from: https://fontawesome.com/icons/angle-down -->
    <svg width="16" height="16" viewBox="0 0 448 512">
        <path
            fill="currentColor"
            d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
        ></path>
    </svg>
</template>
<!-- tooltips plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin makes it such that when the user hovers or
        // focuses a link to a citation or figure, a tooltip appears with a
        // preview of the reference content, along with arrows to navigate
        // between instances of the same reference in the document.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'tooltips';

        // default plugin options
        const options = {
            // whether user must click off to close tooltip instead of just
            // un-hovering
            clickClose: 'false',
            // delay (in ms) between opening and closing tooltip
            delay: '100',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            const links = getLinks();
            for (const link of links) {
                // attach hover and focus listeners to link
                link.addEventListener('mouseover', onLinkHover);
                link.addEventListener('mouseleave', onLinkUnhover);
                link.addEventListener('focus', onLinkFocus);
                link.addEventListener('touchend', onLinkTouch);
            }

            // attach mouse, key, and resize listeners to window
            window.addEventListener('mousedown', onClick);
            window.addEventListener('touchstart', onClick);
            window.addEventListener('keyup', onKeyUp);
            window.addEventListener('resize', onResize);
        }

        // when link is hovered
        function onLinkHover() {
            // function to open tooltip
            const delayOpenTooltip = function() {
                openTooltip(this);
            }.bind(this);

            // run open function after delay
            this.openTooltipTimer = window.setTimeout(
                delayOpenTooltip,
                options.delay
            );
        }

        // when mouse leaves link
        function onLinkUnhover() {
            // cancel opening tooltip
            window.clearTimeout(this.openTooltipTimer);

            // don't close on unhover if option specifies
            if (options.clickClose === 'true')
                return;

            // function to close tooltip
            const delayCloseTooltip = function() {
                // if tooltip open and if mouse isn't over tooltip, close
                const tooltip = document.getElementById('tooltip');
                if (tooltip && !tooltip.matches(':hover'))
                    closeTooltip();
            };

            // run close function after delay
            this.closeTooltipTimer = window.setTimeout(
                delayCloseTooltip,
                options.delay
            );
        }

        // when link is focused (tabbed to)
        function onLinkFocus(event) {
            openTooltip(this);
        }

        // when link is touched on touch screen
        function onLinkTouch(event) {
            // attempt to force hover state on first tap always, and trigger
            // regular link click (and navigation) on second tap
            if (event.target === document.activeElement)
                event.target.click();
            else {
                document.activeElement.blur();
                event.target.focus();
            }
            if (event.cancelable)
                event.preventDefault();
            event.stopPropagation();
            return false;
        }

        // when mouse is clicked anywhere in window
        function onClick(event) {
            closeTooltip();
        }

        // when key pressed
        function onKeyUp(event) {
            if (!event || !event.key)
                return;

            switch (event.key) {
                // trigger click of prev button
                case 'ArrowLeft':
                    const prevButton = document.getElementById(
                        'tooltip_prev_button'
                    );
                    if (prevButton)
                        prevButton.click();
                    break;
                // trigger click of next button
                case 'ArrowRight':
                    const nextButton = document.getElementById(
                        'tooltip_next_button'
                    );
                    if (nextButton)
                        nextButton.click();
                    break;
                // close on esc
                case 'Escape':
                    closeTooltip();
                    break;
            }
        }

        // when window is resized or zoomed
        function onResize() {
            closeTooltip();
        }

        // get all links of types we wish to handle
        function getLinks() {
            const queries = [];
            // exclude buttons, anchor links, toc links, etc
            const exclude =
                ':not(.button):not(.icon_button):not(.anchor):not(.toc_link)';
            queries.push('a[href^="#ref-"]' + exclude); // citation links
            queries.push('a[href^="#fig:"]' + exclude); // figure links
            const query = queries.join(', ');
            return document.querySelectorAll(query);
        }

        // get links with same target, get index of link in set, get total
        // same links
        function getSameLinks(link) {
            const sameLinks = [];
            const links = getLinks();
            for (const otherLink of links) {
                if (
                    otherLink.getAttribute('href') === link.getAttribute('href')
                )
                    sameLinks.push(otherLink);
            }

            return {
                elements: sameLinks,
                index: sameLinks.indexOf(link),
                total: sameLinks.length
            };
        }

        // open tooltip
        function openTooltip(link) {
            // delete tooltip if it exists, start fresh
            closeTooltip();

            // make tooltip element
            const tooltip = makeTooltip(link);

            // if source couldn't be found and tooltip not made, exit
            if (!tooltip)
                return;

            // make navbar elements
            const navBar = makeNavBar(link);
            if (navBar)
                tooltip.firstElementChild.appendChild(navBar);

            // attach tooltip to page
            document.body.appendChild(tooltip);

            // position tooltip
            const position = function() {
                positionTooltip(link);
            };
            position();

            // if tooltip contains images, position again after they've loaded
            const imgs = tooltip.querySelectorAll('img');
            for (const img of imgs)
                img.addEventListener('load', position);
        }

        // close (delete) tooltip
        function closeTooltip() {
            const tooltip = document.getElementById('tooltip');
            if (tooltip)
                tooltip.remove();
        }

        // make tooltip
        function makeTooltip(link) {
            // get target element that link points to
            const source = getSource(link);

            // if source can't be found, exit
            if (!source)
                return;

            // create new tooltip
            const tooltip = document.createElement('div');
            tooltip.id = 'tooltip';
            const tooltipContent = document.createElement('div');
            tooltipContent.id = 'tooltip_content';
            tooltip.appendChild(tooltipContent);

            // make copy of source node and put in tooltip
            const sourceCopy = makeCopy(source);
            tooltipContent.appendChild(sourceCopy);

            // attach mouse event listeners
            tooltip.addEventListener('click', onTooltipClick);
            tooltip.addEventListener('mousedown', onTooltipClick);
            tooltip.addEventListener('touchstart', onTooltipClick);
            tooltip.addEventListener('mouseleave', onTooltipUnhover);

            // (for interaction with lightbox plugin)
            // transfer click on tooltip copied img to original img
            const sourceImg = source.querySelector('img');
            const sourceCopyImg = sourceCopy.querySelector('img');
            if (sourceImg && sourceCopyImg) {
                const clickImg = function() {
                    sourceImg.click();
                    closeTooltip();
                };
                sourceCopyImg.addEventListener('click', clickImg);
            }

            return tooltip;
        }

        // make carbon copy of html dom element
        function makeCopy(source) {
            const sourceCopy = source.cloneNode(true);

            // delete elements marked with ignore (eg anchor and jump buttons)
            const deleteFromCopy = sourceCopy.querySelectorAll(
                '[data-ignore="true"]'
            );
            for (const element of deleteFromCopy)
                element.remove();

            // delete certain element attributes
            const attributes = [
                'id',
                'data-collapsed',
                'data-selected',
                'data-highlighted',
                'data-glow'
            ];
            for (const attribute of attributes) {
                sourceCopy.removeAttribute(attribute);
                const elements = sourceCopy.querySelectorAll(
                    '[' + attribute + ']'
                );
                for (const element of elements)
                    element.removeAttribute(attribute);
            }

            return sourceCopy;
        }

        // when tooltip is clicked
        function onTooltipClick(event) {
            // when user clicks on tooltip, stop click from transferring
            // outside of tooltip (eg, click off to close tooltip, or eg click
            // off to unhighlight same refs)
            event.stopPropagation();
        }

        // when tooltip is unhovered
        function onTooltipUnhover(event) {
            if (options.clickClose === 'true')
                return;

            // make sure new mouse/touch/focus no longer over tooltip or any
            // element within it
            const tooltip = document.getElementById('tooltip');
            if (!tooltip)
                return;
            if (this.contains(event.relatedTarget))
                return;

            closeTooltip();
        }

        // make nav bar to go betwen prev/next instances of same reference
        function makeNavBar(link) {
            // find other links to the same source
            const sameLinks = getSameLinks(link);

            // don't show nav bar when singular reference
            if (sameLinks.total <= 1)
                return;

            // find prev/next links with same target
            const prevLink = getPrevLink(link, sameLinks);
            const nextLink = getNextLink(link, sameLinks);

            // create nav bar
            const navBar = document.createElement('div');
            navBar.id = 'tooltip_nav_bar';
            const text = sameLinks.index + 1 + ' of ' + sameLinks.total;

            // create nav bar prev/next buttons
            const prevButton = document.createElement('button');
            const nextButton = document.createElement('button');
            prevButton.id = 'tooltip_prev_button';
            nextButton.id = 'tooltip_next_button';
            prevButton.title =
                'Jump to the previous occurence of this item in the document [←]';
            nextButton.title =
                'Jump to the next occurence of this item in the document [→]';
            prevButton.classList.add('icon_button');
            nextButton.classList.add('icon_button');
            prevButton.innerHTML = document.querySelector(
                '.icon_caret_left'
            ).innerHTML;
            nextButton.innerHTML = document.querySelector(
                '.icon_caret_right'
            ).innerHTML;
            navBar.appendChild(prevButton);
            navBar.appendChild(document.createTextNode(text));
            navBar.appendChild(nextButton);

            // attach click listeners to buttons
            prevButton.addEventListener('click', function() {
                onPrevNextClick(link, prevLink);
            });
            nextButton.addEventListener('click', function() {
                onPrevNextClick(link, nextLink);
            });

            return navBar;
        }

        // get previous link with same target
        function getPrevLink(link, sameLinks) {
            if (!sameLinks)
                sameLinks = getSameLinks(link);
            // wrap index to other side if < 1
            let index;
            if (sameLinks.index - 1 >= 0)
                index = sameLinks.index - 1;
            else
                index = sameLinks.total - 1;
            return sameLinks.elements[index];
        }

        // get next link with same target
        function getNextLink(link, sameLinks) {
            if (!sameLinks)
                sameLinks = getSameLinks(link);
            // wrap index to other side if > total
            let index;
            if (sameLinks.index + 1 <= sameLinks.total - 1)
                index = sameLinks.index + 1;
            else
                index = 0;
            return sameLinks.elements[index];
        }

        // get element that is target of link or url hash
        function getSource(link) {
            const hash = link ? link.hash : window.location.hash;
            const id = hash.slice(1);
            let target = document.querySelector('[id="' + id + '"]');
            if (!target)
                return;

            // if ref or figure, modify target to get expected element
            if (id.indexOf('ref-') === 0)
                target = target.querySelector('p');
            else if (id.indexOf('fig:') === 0)
                target = target.querySelector('figure');

            return target;
        }

        // when prev/next arrow button is clicked
        function onPrevNextClick(link, prevNextLink) {
            if (link && prevNextLink)
                goToElement(prevNextLink, window.innerHeight * 0.5);
        }

        // scroll to and focus element
        function goToElement(element, offset) {
            // expand accordion section if collapsed
            expandElement(element);
            const y =
                getRectInView(element).top -
                getRectInView(document.documentElement).top -
                (offset || 0);
            // trigger any function listening for "onscroll" event
            window.dispatchEvent(new Event('scroll'));
            window.scrollTo(0, y);
            document.activeElement.blur();
            element.focus();
        }

        // determine position to place tooltip based on link position in
        // viewport and tooltip size
        function positionTooltip(link, left, top) {
            const tooltipElement = document.getElementById('tooltip');
            if (!tooltipElement)
                return;

            // get convenient vars for position/dimensions of
            // link/tooltip/page/view
            link = getRectInPage(link);
            const tooltip = getRectInPage(tooltipElement);
            const view = getRectInPage();

            // horizontal positioning
            if (left)
                // use explicit value
                left = left;
            else if (link.left + tooltip.width < view.right)
                // fit tooltip to right of link
                left = link.left;
            else if (link.right - tooltip.width > view.left)
                // fit tooltip to left of link
                left = link.right - tooltip.width;
            // center tooltip in view
            else
                left = (view.right - view.left) / 2 - tooltip.width / 2;

            // vertical positioning
            if (top)
                // use explicit value
                top = top;
            else if (link.top - tooltip.height > view.top)
                // fit tooltip above link
                top = link.top - tooltip.height;
            else if (link.bottom + tooltip.height < view.bottom)
                // fit tooltip below link
                top = link.bottom;
            else {
                // center tooltip in view
                top = view.top + view.height / 2 - tooltip.height / 2;
                // nudge off of link to left/right if possible
                if (link.right + tooltip.width < view.right)
                    left = link.right;
                else if (link.left - tooltip.width > view.left)
                    left = link.left - tooltip.width;
            }

            tooltipElement.style.left = left + 'px';
            tooltipElement.style.top = top + 'px';
        }

        // get position/dimensions of element or viewport
        function getRectInView(element) {
            let rect = {};
            rect.left = 0;
            rect.top = 0;
            rect.right = document.documentElement.clientWidth;
            rect.bottom = document.documentElement.clientHeight;
            let style = {};

            if (element instanceof HTMLElement) {
                rect = element.getBoundingClientRect();
                style = window.getComputedStyle(element);
            }

            const margin = {};
            margin.left = parseFloat(style.marginLeftWidth) || 0;
            margin.top = parseFloat(style.marginTopWidth) || 0;
            margin.right = parseFloat(style.marginRightWidth) || 0;
            margin.bottom = parseFloat(style.marginBottomWidth) || 0;

            const border = {};
            border.left = parseFloat(style.borderLeftWidth) || 0;
            border.top = parseFloat(style.borderTopWidth) || 0;
            border.right = parseFloat(style.borderRightWidth) || 0;
            border.bottom = parseFloat(style.borderBottomWidth) || 0;

            const newRect = {};
            newRect.left = rect.left + margin.left + border.left;
            newRect.top = rect.top + margin.top + border.top;
            newRect.right = rect.right + margin.right + border.right;
            newRect.bottom = rect.bottom + margin.bottom + border.bottom;
            newRect.width = newRect.right - newRect.left;
            newRect.height = newRect.bottom - newRect.top;

            return newRect;
        }

        // get position of element relative to page
        function getRectInPage(element) {
            const rect = getRectInView(element);
            const body = getRectInView(document.body);

            const newRect = {};
            newRect.left = rect.left - body.left;
            newRect.top = rect.top - body.top;
            newRect.right = rect.right - body.left;
            newRect.bottom = rect.bottom - body.top;
            newRect.width = rect.width;
            newRect.height = rect.height;

            return newRect;
        }

        // (for interaction with accordion plugin)
        // get closest element before specified element that matches query
        function firstBefore(element, query) {
            while (
                element &&
                element !== document.body &&
                !element.matches(query)
            )
                element = element.previousElementSibling || element.parentNode;

            return element;
        }

        // (for interaction with accordion plugin)
        // check if element is part of collapsed heading
        function isCollapsed(element) {
            while (element && element !== document.body) {
                if (element.dataset.collapsed === 'true')
                    return true;
                element = element.parentNode;
            }
            return false;
        }

        // (for interaction with accordion plugin)
        // expand heading containing element if necesary
        function expandElement(element) {
            if (isCollapsed(element)) {
                const heading = firstBefore(element, 'h2');
                if (heading)
                    heading.click();
            }
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- caret left icon -->

<template class="icon_caret_left">
    <!-- modified from: https://fontawesome.com/icons/caret-left -->
    <svg width="16" height="16" viewBox="0 0 192 512">
        <path
            fill="currentColor"
            d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
        ></path>
    </svg>
</template>

<!-- caret right icon -->

<template class="icon_caret_right">
    <!-- modified from: https://fontawesome.com/icons/caret-right -->
    <svg width="16" height="16" viewBox="0 0 192 512">
        <path
            fill="currentColor"
            d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
        ></path>
    </svg>
</template>
<!-- jump to first plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin adds a button next to each reference entry,
        // figure, and table that jumps the page to the first occurrence of a
        // link to that item in the manuscript.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'jumpToFirst';

        // default plugin options
        const options = {
            // whether to add buttons next to reference entries
            references: 'true',
            // whether to add buttons next to figures
            figures: 'true',
            // whether to add buttons next to tables
            tables: 'true',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            if (options.references !== 'false')
                makeReferenceButtons();
            if (options.figures !== 'false')
                makeFigureButtons();
            if (options.tables !== 'false')
                makeTableButtons();
        }

        // when jump button clicked
        function onButtonClick() {
            const first = getFirstOccurrence(this.dataset.id);
            if (!first)
                return;

            // update url hash so navigating "back" in history will return
            // user to jump button
            window.location.hash = this.dataset.id;
            // scroll to link
            window.setTimeout(function() {
                goToElement(first, window.innerHeight * 0.5);
            }, 0);
        }

        // get first occurence of link to item in document
        function getFirstOccurrence(id) {
            let query = 'a';
            query += '[href="#' + id + '"]';
            // exclude buttons, anchor links, toc links, etc
            query +=
                ':not(.button):not(.icon_button):not(.anchor):not(.toc_link)';
            return document.querySelector(query);
        }

        // add button next to each reference entry
        function makeReferenceButtons() {
            const references = document.querySelectorAll('div[id^="ref-"]');
            for (const reference of references) {
                // get reference id and element to add button to
                const id = reference.id;
                const container = reference.firstElementChild;
                const first = getFirstOccurrence(id);

                // if can't find link to reference, ignore
                if (!first)
                    continue;

                // make jump button
                let button = document.createElement('button');
                button.classList.add('icon_button', 'jump_arrow');
                button.title =
                    'Jump to the first occurence of this reference in the document';
                button.innerHTML = document.querySelector(
                    '.icon_angle_double_up'
                ).innerHTML;
                button.dataset.id = id;
                button.dataset.ignore = 'true';
                container.innerHTML = button.outerHTML + container.innerHTML;
                button = container.firstElementChild;
                button.addEventListener('click', onButtonClick);
            }
        }

        // add button next to each figure
        function makeFigureButtons() {
            const figures = document.querySelectorAll('[id^="fig:"]');
            for (const figure of figures) {
                // get figure id and element to add button to
                const id = figure.id;
                const container = figure.querySelector('figcaption') || figure;
                const first = getFirstOccurrence(id);

                // if can't find link to figure, ignore
                if (!first)
                    continue;

                // make jump button
                const button = document.createElement('button');
                button.classList.add('icon_button', 'jump_arrow');
                button.title =
                    'Jump to the first occurence of this figure in the document';
                button.innerHTML = document.querySelector(
                    '.icon_angle_double_up'
                ).innerHTML;
                button.dataset.id = id;
                button.dataset.ignore = 'true';
                container.insertBefore(button, container.firstElementChild);
                button.addEventListener('click', onButtonClick);
            }
        }

        // add button next to each figure
        function makeTableButtons() {
            const tables = document.querySelectorAll('[id^="tbl:"]');
            for (const table of tables) {
                // get ref id and element to add button to
                const id = table.id;
                const container = table.querySelector('caption') || table;
                const first = getFirstOccurrence(id);

                // if can't find link to table, ignore
                if (!first)
                    continue;

                // make jump button
                const button = document.createElement('button');
                button.classList.add('icon_button', 'jump_arrow');
                button.title =
                    'Jump to the first occurence of this table in the document';
                button.innerHTML = document.querySelector(
                    '.icon_angle_double_up'
                ).innerHTML;
                button.dataset.id = id;
                button.dataset.ignore = 'true';
                container.insertBefore(button, container.firstElementChild);
                button.addEventListener('click', onButtonClick);
            }
        }

        // scroll to and focus element
        function goToElement(element, offset) {
            // expand accordion section if collapsed
            expandElement(element);
            const y =
                getRectInView(element).top -
                getRectInView(document.documentElement).top -
                (offset || 0);
            // trigger any function listening for "onscroll" event
            window.dispatchEvent(new Event('scroll'));
            window.scrollTo(0, y);
            document.activeElement.blur();
            element.focus();
        }

        // get position/dimensions of element or viewport
        function getRectInView(element) {
            let rect = {};
            rect.left = 0;
            rect.top = 0;
            rect.right = document.documentElement.clientWidth;
            rect.bottom = document.documentElement.clientHeight;
            let style = {};

            if (element instanceof HTMLElement) {
                rect = element.getBoundingClientRect();
                style = window.getComputedStyle(element);
            }

            const margin = {};
            margin.left = parseFloat(style.marginLeftWidth) || 0;
            margin.top = parseFloat(style.marginTopWidth) || 0;
            margin.right = parseFloat(style.marginRightWidth) || 0;
            margin.bottom = parseFloat(style.marginBottomWidth) || 0;

            const border = {};
            border.left = parseFloat(style.borderLeftWidth) || 0;
            border.top = parseFloat(style.borderTopWidth) || 0;
            border.right = parseFloat(style.borderRightWidth) || 0;
            border.bottom = parseFloat(style.borderBottomWidth) || 0;

            const newRect = {};
            newRect.left = rect.left + margin.left + border.left;
            newRect.top = rect.top + margin.top + border.top;
            newRect.right = rect.right + margin.right + border.right;
            newRect.bottom = rect.bottom + margin.bottom + border.bottom;
            newRect.width = newRect.right - newRect.left;
            newRect.height = newRect.bottom - newRect.top;

            return newRect;
        }

        // get closest element before specified element that matches query
        function firstBefore(element, query) {
            while (
                element &&
                element !== document.body &&
                !element.matches(query)
            )
                element = element.previousElementSibling || element.parentNode;

            return element;
        }

        // check if element is part of collapsed heading
        function isCollapsed(element) {
            while (element && element !== document.body) {
                if (element.dataset.collapsed === 'true')
                    return true;
                element = element.parentNode;
            }
            return false;
        }

        // (for interaction with accordion plugin)
        // expand heading containing element if necesary
        function expandElement(element) {
            if (isCollapsed(element)) {
                const heading = firstBefore(element, 'h2');
                if (heading)
                    heading.click();
            }
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- angle double up icon -->

<template class="icon_angle_double_up">
    <!-- modified from: https://fontawesome.com/icons/angle-double-up -->
    <svg width="16" height="16" viewBox="0 0 320 512">
        <path
            fill="currentColor"
            d="M177 255.7l136 136c9.4 9.4 9.4 24.6 0 33.9l-22.6 22.6c-9.4 9.4-24.6 9.4-33.9 0L160 351.9l-96.4 96.4c-9.4 9.4-24.6 9.4-33.9 0L7 425.7c-9.4-9.4-9.4-24.6 0-33.9l136-136c9.4-9.5 24.6-9.5 34-.1zm-34-192L7 199.7c-9.4 9.4-9.4 24.6 0 33.9l22.6 22.6c9.4 9.4 24.6 9.4 33.9 0l96.4-96.4 96.4 96.4c9.4 9.4 24.6 9.4 33.9 0l22.6-22.6c9.4-9.4 9.4-24.6 0-33.9l-136-136c-9.2-9.4-24.4-9.4-33.8 0z"
        ></path>
    </svg>
</template>
<!-- link highlight plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin makes it such that when a user hovers or
        // focuses a link, other links that have the same target will be
        // highlighted. It also makes it such that when clicking a link, the
        // target of the link (eg reference, figure, table) is briefly
        // highlighted.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'linkHighlight';

        // default plugin options
        const options = {
            // whether to also highlight links that go to external urls
            externalLinks: 'false',
            // whether user must click off to unhighlight instead of just
            // un-hovering
            clickUnhighlight: 'false',
            // whether to also highlight links that are unique
            highlightUnique: 'true',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            const links = getLinks();
            for (const link of links) {
                // attach mouse and focus listeners to link
                link.addEventListener('mouseenter', onLinkFocus);
                link.addEventListener('focus', onLinkFocus);
                link.addEventListener('mouseleave', onLinkUnhover);
            }

            // attach click and hash change listeners to window
            window.addEventListener('click', onClick);
            window.addEventListener('touchstart', onClick);
            window.addEventListener('hashchange', onHashChange);

            // run hash change on window load in case user has navigated
            // directly to hash
            onHashChange();
        }

        // when link is focused (tabbed to) or hovered
        function onLinkFocus() {
            highlight(this);
        }

        // when link is unhovered
        function onLinkUnhover() {
            if (options.clickUnhighlight !== 'true')
                unhighlightAll();
        }

        // when the mouse is clicked anywhere in window
        function onClick(event) {
            unhighlightAll();
        }

        // when hash (eg manuscript.html#introduction) changes
        function onHashChange() {
            const target = getHashTarget();
            if (target)
                glowElement(target);
        }

        // get element that is target of link or url hash
        function getHashTarget(link) {
            const hash = link ? link.hash : window.location.hash;
            const id = hash.slice(1);
            let target = document.querySelector('[id="' + id + '"]');
            if (!target)
                return;

            return target;
        }

        // start glow sequence on an element
        function glowElement(element) {
            const startGlow = function() {
                onGlowEnd();
                element.dataset.glow = 'true';
                element.addEventListener('animationend', onGlowEnd);
            };
            const onGlowEnd = function() {
                element.removeAttribute('data-glow');
                element.removeEventListener('animationend', onGlowEnd);
            };
            startGlow();
        }

        // highlight link and all others with same target
        function highlight(link) {
            // force unhighlight all to start fresh
            unhighlightAll();

            // get links with same target
            if (!link)
                return;
            const sameLinks = getSameLinks(link);

            // if link unique and option is off, exit and don't highlight
            if (sameLinks.length <= 1 && options.highlightUnique !== 'true')
                return;

            // highlight all same links, and "select" (special highlight) this
            // one
            for (const sameLink of sameLinks) {
                if (sameLink === link)
                    sameLink.setAttribute('data-selected', 'true');
                else
                    sameLink.setAttribute('data-highlighted', 'true');
            }
        }

        // unhighlight all links
        function unhighlightAll() {
            const links = getLinks();
            for (const link of links) {
                link.setAttribute('data-selected', 'false');
                link.setAttribute('data-highlighted', 'false');
            }
        }

        // get links with same target
        function getSameLinks(link) {
            const results = [];
            const links = getLinks();
            for (const otherLink of links) {
                if (
                    otherLink.getAttribute('href') === link.getAttribute('href')
                )
                    results.push(otherLink);
            }
            return results;
        }

        // get all links of types we wish to handle
        function getLinks() {
            let query = 'a';
            if (options.externalLinks !== 'true')
                query += '[href^="#"]';
            // exclude buttons, anchor links, toc links, etc
            query +=
                ':not(.button):not(.icon_button):not(.anchor):not(.toc_link)';
            return document.querySelectorAll(query);
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>
<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin provides a "table of contents" (toc) panel on
        // the side of the document that allows the user to conveniently
        // navigate between sections of the document.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'tableOfContents';

        // default plugin options
        const options = {
            // which types of elements to add links for, in
            // "document.querySelector" format
            typesQuery: 'h1, h2, h3',
            // whether toc starts open. use 'true' or 'false', or 'auto' to
            // use 'true' behavior when screen wide enough and 'false' when not
            startOpen: 'false',
            // whether toc closes when clicking on toc link. use 'true' or
            // 'false', or 'auto' to use 'false' behavior when screen wide
            // enough and 'true' when not
            clickClose: 'auto',
            // if list item is more than this many characters, text will be
            // truncated
            charLimit: '50',
            // whether or not to show bullets next to each toc item
            bullets: 'false',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            // make toc panel and populate with entries (links to document
            // sections)
            const panel = makePanel();
            if (!panel)
                return;
            makeEntries(panel);
            // attach panel to document after making entries, so 'toc' heading
            // in panel isn't included in toc
            document.body.insertBefore(panel, document.body.firstChild);

            // initial panel state
            if (
                options.startOpen === 'true' ||
                (options.startOpen === 'auto' && !isSmallScreen())
            )
                openPanel();
            else
                closePanel();

            // attach click, scroll, and hash change listeners to window
            window.addEventListener('click', onClick);
            window.addEventListener('scroll', onScroll);
            window.addEventListener('hashchange', onScroll);
            window.addEventListener('keyup', onKeyUp);
            onScroll();

            // add class to push document body down out of way of toc button
            document.body.classList.add('toc_body_nudge');
        }

        // determine if screen wide enough to fit toc panel
        function isSmallScreen() {
            // in default theme:
            // 816px = 8.5in = width of "page" (<body>) element
            // 260px = min width of toc panel (*2 for both sides of <body>)
            return window.innerWidth < 816 + 260 * 2;
        }

        // when mouse is clicked anywhere in window
        function onClick() {
            if (isSmallScreen())
                closePanel();
        }

        // when window is scrolled or hash changed
        function onScroll() {
            highlightViewed();
        }

        // when key pressed
        function onKeyUp(event) {
            if (!event || !event.key)
                return;

            // close on esc
            if (event.key === 'Escape')
                closePanel();
        }

        // find entry of currently viewed document section in toc and highlight
        function highlightViewed() {
            const firstId = getFirstInView(options.typesQuery);

            // get toc entries (links), unhighlight all, then highlight viewed
            const list = document.getElementById('toc_list');
            if (!firstId || !list)
                return;
            const links = list.querySelectorAll('a');
            for (const link of links)
                link.dataset.viewing = 'false';
            const link = list.querySelector('a[href="#' + firstId + '"]');
            if (!link)
                return;
            link.dataset.viewing = 'true';
        }

        // get first or previous toc listed element in top half of view
        function getFirstInView(query) {
            // get all elements matching query and with id
            const elements = document.querySelectorAll(query);
            const elementsWithIds = [];
            for (const element of elements) {
                if (element.id)
                    elementsWithIds.push(element);
            }


            // get first or previous element in top half of view
            for (let i = 0; i < elementsWithIds.length; i++) {
                const element = elementsWithIds[i];
                const prevElement = elementsWithIds[Math.max(0, i - 1)];
                if (element.getBoundingClientRect().top >= 0) {
                    if (
                        element.getBoundingClientRect().top <
                        window.innerHeight / 2
                    )
                        return element.id;
                    else
                        return prevElement.id;
                }
            }
        }

        // make panel
        function makePanel() {
            // create panel
            const panel = document.createElement('div');
            panel.id = 'toc_panel';
            if (options.bullets === 'true')
                panel.dataset.bullets = 'true';

            // create header
            const header = document.createElement('div');
            header.id = 'toc_header';

            // create toc button
            const button = document.createElement('button');
            button.id = 'toc_button';
            button.innerHTML = document.querySelector('.icon_th_list').innerHTML;
            button.title = 'Table of Contents';
            button.classList.add('icon_button');

            // create header text
            const text = document.createElement('h4');
            text.innerHTML = 'Table of Contents';

            // create container for toc list
            const list = document.createElement('div');
            list.id = 'toc_list';

            // attach click listeners
            panel.addEventListener('click', onPanelClick);
            header.addEventListener('click', onHeaderClick);
            button.addEventListener('click', onButtonClick);

            // attach elements
            header.appendChild(button);
            header.appendChild(text);
            panel.appendChild(header);
            panel.appendChild(list);

            return panel;
        }

        // create toc entries (links) to each element of the specified types
        function makeEntries(panel) {
            const elements = document.querySelectorAll(options.typesQuery);
            for (const element of elements) {
                // do not add link if element doesn't have assigned id
                if (!element.id)
                    continue;

                // create link/list item
                const link = document.createElement('a');
                link.classList.add('toc_link');
                switch (element.tagName.toLowerCase()) {
                    case 'h1':
                        link.dataset.level = '1';
                        break;
                    case 'h2':
                        link.dataset.level = '2';
                        break;
                    case 'h3':
                        link.dataset.level = '3';
                        break;
                    case 'h4':
                        link.dataset.level = '4';
                        break;
                }
                link.title = element.innerText;
                let text = element.innerText;
                if (text.length > options.charLimit)
                    text = text.slice(0, options.charLimit) + '...';
                link.innerHTML = text;
                link.href = '#' + element.id;
                link.addEventListener('click', onLinkClick);

                // attach link
                panel.querySelector('#toc_list').appendChild(link);
            }
        }

        // when panel is clicked
        function onPanelClick(event) {
            // stop click from propagating to window/document and closing panel
            event.stopPropagation();
        }

        // when header itself is clicked
        function onHeaderClick(event) {
            togglePanel();
        }

        // when button is clicked
        function onButtonClick(event) {
            togglePanel();
            // stop header underneath button from also being clicked
            event.stopPropagation();
        }

        // when link is clicked
        function onLinkClick(event) {
            if (
                options.clickClose === 'true' ||
                (options.clickClose === 'auto' && isSmallScreen())
            )
                closePanel();
            else
                openPanel();
        }

        // open panel if closed, close if opened
        function togglePanel() {
            const panel = document.getElementById('toc_panel');
            if (!panel)
                return;

            if (panel.dataset.open === 'true')
                closePanel();
            else
                openPanel();
        }

        // open panel
        function openPanel() {
            const panel = document.getElementById('toc_panel');
            if (panel)
                panel.dataset.open = 'true';
        }

        // close panel
        function closePanel() {
            const panel = document.getElementById('toc_panel');
            if (panel)
                panel.dataset.open = 'false';
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- th list icon -->

<template class="icon_th_list">
    <!-- modified from: https://fontawesome.com/icons/th-list -->
    <svg width="16" height="16" viewBox="0 0 512 512" tabindex="-1">
        <path
            fill="currentColor"
            d="M96 96c0 26.51-21.49 48-48 48S0 122.51 0 96s21.49-48 48-48 48 21.49 48 48zM48 208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm0 160c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm96-236h352c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h352c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h352c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
            tabindex="-1"
        ></path>
    </svg>
</template>
<!-- lightbox plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin makes it such that when a user clicks on an
        // image, the image fills the screen and the user can pan/drag/zoom
        // the image and navigate between other images in the document.

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'lightbox';

        // default plugin options
        const options = {
            // list of possible zoom/scale factors
            zoomSteps:
                '0.1, 0.25, 0.333333, 0.5, 0.666666, 0.75, 1,' +
                '1.25, 1.5, 1.75, 2, 2.5, 3, 3.5, 4, 5, 6, 7, 8',
            // whether to fit image to view ('fit'), display at 100% and shrink
            // if necessary ('shrink'), or always display at 100% ('100')
            defaultZoom: 'fit',
            // whether to zoom in/out toward center of view ('true') or mouse
            // ('false')
            centerZoom: 'false',
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            // run through each <img> element
            const imgs = document.querySelectorAll('figure > img');
            let count = 1;
            for (const img of imgs) {
                img.classList.add('lightbox_document_img');
                img.dataset.number = count;
                img.dataset.total = imgs.length;
                img.addEventListener('click', openLightbox);
                count++;
            }

            // attach mouse and key listeners to window
            window.addEventListener('mousemove', onWindowMouseMove);
            window.addEventListener('keyup', onKeyUp);
        }

        // when mouse is moved anywhere in window
        function onWindowMouseMove(event) {
            window.mouseX = event.clientX;
            window.mouseY = event.clientY;
        }

        // when key pressed
        function onKeyUp(event) {
            if (!event || !event.key)
                return;

            switch (event.key) {
                // trigger click of prev button
                case 'ArrowLeft':
                    const prevButton = document.getElementById(
                        'lightbox_prev_button'
                    );
                    if (prevButton)
                        prevButton.click();
                    break;
                // trigger click of next button
                case 'ArrowRight':
                    const nextButton = document.getElementById(
                        'lightbox_next_button'
                    );
                    if (nextButton)
                        nextButton.click();
                    break;
                // close on esc
                case 'Escape':
                    closeLightbox();
                    break;
            }
        }

        // open lightbox
        function openLightbox() {
            const lightbox = makeLightbox(this);
            if (!lightbox)
                return;

            blurBody(lightbox);
            document.body.appendChild(lightbox);
        }

        // make lightbox
        function makeLightbox(img) {
            // delete lightbox if it exists, start fresh
            closeLightbox();

            // create screen overlay containing lightbox
            const overlay = document.createElement('div');
            overlay.id = 'lightbox_overlay';

            // create image info boxes
            const numberInfo = document.createElement('div');
            const zoomInfo = document.createElement('div');
            numberInfo.id = 'lightbox_number_info';
            zoomInfo.id = 'lightbox_zoom_info';

            // create container for image
            const imageContainer = document.createElement('div');
            imageContainer.id = 'lightbox_image_container';
            const lightboxImg = makeLightboxImg(
                img,
                imageContainer,
                numberInfo,
                zoomInfo
            );
            imageContainer.appendChild(lightboxImg);

            // create bottom container for caption and navigation buttons
            const bottomContainer = document.createElement('div');
            bottomContainer.id = 'lightbox_bottom_container';
            const caption = makeCaption(img);
            const prevButton = makePrevButton(img);
            const nextButton = makeNextButton(img);
            bottomContainer.appendChild(prevButton);
            bottomContainer.appendChild(caption);
            bottomContainer.appendChild(nextButton);

            // attach top middle and bottom to overlay
            overlay.appendChild(numberInfo);
            overlay.appendChild(zoomInfo);
            overlay.appendChild(imageContainer);
            overlay.appendChild(bottomContainer);

            return overlay;
        }

        // make <img> object that is intuitively draggable and zoomable
        function makeLightboxImg(
            sourceImg,
            container,
            numberInfoBox,
            zoomInfoBox
        ) {
            // create copy of source <img>
            const img = sourceImg.cloneNode(true);
            img.classList.remove('lightbox_document_img');
            img.removeAttribute('id');
            img.removeAttribute('width');
            img.removeAttribute('height');
            img.style.position = 'unset';
            img.style.margin = '0';
            img.style.padding = '0';
            img.style.width = '';
            img.style.height = '';
            img.style.minWidth = '';
            img.style.minHeight = '';
            img.style.maxWidth = '';
            img.style.maxHeight = '';
            img.id = 'lightbox_img';

            // build sorted list of unique zoomSteps, always including a 100%
            let zoomSteps = [];
            const optionsZooms = options.zoomSteps.split(/[^0-9.]/);
            for (const optionZoom of optionsZooms) {
                const newZoom = parseFloat(optionZoom);
                if (newZoom && !zoomSteps.includes(newZoom))
                    zoomSteps.push(newZoom);
            }
            if (!zoomSteps.includes(1))
                zoomSteps.push(1);
            zoomSteps = zoomSteps.sort(function sortNumber(a, b) {
                return a - b;
            });

            // <img> object property variables
            let zoom = 1;
            let translateX = 0;
            let translateY = 0;
            let clickMouseX = undefined;
            let clickMouseY = undefined;
            let clickTranslateX = undefined;
            let clickTranslateY = undefined;

            updateNumberInfo();

            // update image numbers displayed in info box
            function updateNumberInfo() {
                numberInfoBox.innerHTML =
                    sourceImg.dataset.number + ' of ' + sourceImg.dataset.total;
            }

            // update zoom displayed in info box
            function updateZoomInfo() {
                let zoomInfo = zoom * 100;
                if (!Number.isInteger(zoomInfo))
                    zoomInfo = zoomInfo.toFixed(2);
                zoomInfoBox.innerHTML = zoomInfo + '%';
            }

            // move to closest zoom step above current zoom
            const zoomIn = function() {
                for (const zoomStep of zoomSteps) {
                    if (zoomStep > zoom) {
                        zoom = zoomStep;
                        break;
                    }
                }
                updateTransform();
            };

            // move to closest zoom step above current zoom
            const zoomOut = function() {
                zoomSteps.reverse();
                for (const zoomStep of zoomSteps) {
                    if (zoomStep < zoom) {
                        zoom = zoomStep;
                        break;
                    }
                }
                zoomSteps.reverse();

                updateTransform();
            };

            // update display of <img> based on scale/translate properties
            const updateTransform = function() {
                // set transform
                img.style.transform =
                    'translate(' +
                    (translateX || 0) +
                    'px,' +
                    (translateY || 0) +
                    'px) scale(' +
                    (zoom || 1) +
                    ')';

                // get new width/height after scale
                const rect = img.getBoundingClientRect();
                // limit translate
                translateX = Math.max(translateX, -rect.width / 2);
                translateX = Math.min(translateX, rect.width / 2);
                translateY = Math.max(translateY, -rect.height / 2);
                translateY = Math.min(translateY, rect.height / 2);

                // set transform
                img.style.transform =
                    'translate(' +
                    (translateX || 0) +
                    'px,' +
                    (translateY || 0) +
                    'px) scale(' +
                    (zoom || 1) +
                    ')';

                updateZoomInfo();
            };

            // fit <img> to container
            const fit = function() {
                // no x/y offset, 100% zoom by default
                translateX = 0;
                translateY = 0;
                zoom = 1;

                // widths of <img> and container
                const imgWidth = img.naturalWidth;
                const imgHeight = img.naturalHeight;
                const containerWidth = parseFloat(
                    window.getComputedStyle(container).width
                );
                const containerHeight = parseFloat(
                    window.getComputedStyle(container).height
                );

                // how much zooming is needed to fit <img> to container
                const xRatio = imgWidth / containerWidth;
                const yRatio = imgHeight / containerHeight;
                const maxRatio = Math.max(xRatio, yRatio);
                const newZoom = 1 / maxRatio;

                // fit <img> to container according to option
                if (options.defaultZoom === 'shrink') {
                    if (maxRatio > 1)
                        zoom = newZoom;
                } else if (options.defaultZoom === 'fit')
                    zoom = newZoom;

                updateTransform();
            };

            // when mouse wheel is rolled anywhere in container
            const onContainerWheel = function(event) {
                if (!event)
                    return;

                // let ctrl + mouse wheel to zoom behave as normal
                if (event.ctrlKey)
                    return;

                // prevent normal scroll behavior
                event.preventDefault();
                event.stopPropagation();

                // point around which to scale img
                const viewRect = container.getBoundingClientRect();
                const viewX = (viewRect.left + viewRect.right) / 2;
                const viewY = (viewRect.top + viewRect.bottom) / 2;
                const originX = options.centerZoom === 'true' ? viewX : mouseX;
                const originY = options.centerZoom === 'true' ? viewY : mouseY;

                // get point on image under origin
                const oldRect = img.getBoundingClientRect();
                const oldPercentX = (originX - oldRect.left) / oldRect.width;
                const oldPercentY = (originY - oldRect.top) / oldRect.height;

                // increment/decrement zoom
                if (event.deltaY < 0)
                    zoomIn();
                if (event.deltaY > 0)
                    zoomOut();

                // get offset between previous image point and origin
                const newRect = img.getBoundingClientRect();
                const offsetX =
                    originX - (newRect.left + newRect.width * oldPercentX);
                const offsetY =
                    originY - (newRect.top + newRect.height * oldPercentY);

                // translate image to keep image point under origin
                translateX += offsetX;
                translateY += offsetY;

                // perform translate
                updateTransform();
            };

            // when container is clicked
            function onContainerClick(event) {
                // if container itself is target of click, and not other
                // element above it
                if (event.target === this)
                    closeLightbox();
            }

            // when mouse button is pressed on image
            const onImageMouseDown = function(event) {
                // store original mouse position relative to image
                clickMouseX = window.mouseX;
                clickMouseY = window.mouseY;
                clickTranslateX = translateX;
                clickTranslateY = translateY;
                event.stopPropagation();
                event.preventDefault();
            };

            // when mouse button is released anywhere in window
            const onWindowMouseUp = function(event) {
                // reset original mouse position
                clickMouseX = undefined;
                clickMouseY = undefined;
                clickTranslateX = undefined;
                clickTranslateY = undefined;

                // remove global listener if lightbox removed from document
                if (!document.body.contains(container))
                    window.removeEventListener('mouseup', onWindowMouseUp);
            };

            // when mouse is moved anywhere in window
            const onWindowMouseMove = function(event) {
                if (
                    clickMouseX === undefined ||
                    clickMouseY === undefined ||
                    clickTranslateX === undefined ||
                    clickTranslateY === undefined
                )
                    return;

                // offset image based on original and current mouse position
                translateX = clickTranslateX + window.mouseX - clickMouseX;
                translateY = clickTranslateY + window.mouseY - clickMouseY;
                updateTransform();
                event.preventDefault();

                // remove global listener if lightbox removed from document
                if (!document.body.contains(container))
                    window.removeEventListener('mousemove', onWindowMouseMove);
            };

            // when window is resized
            const onWindowResize = function(event) {
                fit();

                // remove global listener if lightbox removed from document
                if (!document.body.contains(container))
                    window.removeEventListener('resize', onWindowResize);
            };

            // attach the necessary event listeners
            img.addEventListener('dblclick', fit);
            img.addEventListener('mousedown', onImageMouseDown);
            container.addEventListener('wheel', onContainerWheel);
            container.addEventListener('mousedown', onContainerClick);
            container.addEventListener('touchstart', onContainerClick);
            window.addEventListener('mouseup', onWindowMouseUp);
            window.addEventListener('mousemove', onWindowMouseMove);
            window.addEventListener('resize', onWindowResize);

            // run fit() after lightbox atttached to document and <img> Loaded
            // so needed container and img dimensions available
            img.addEventListener('load', fit);

            return img;
        }

        // make caption
        function makeCaption(img) {
            const caption = document.createElement('div');
            caption.id = 'lightbox_caption';
            const captionSource = img.nextElementSibling;
            if (captionSource.tagName.toLowerCase() === 'figcaption') {
                const captionCopy = makeCopy(captionSource);
                caption.innerHTML = captionCopy.innerHTML;
            }

            caption.addEventListener('touchstart', function(event) {
                event.stopPropagation();
            });

            return caption;
        }

        // make carbon copy of html dom element
        function makeCopy(source) {
            const sourceCopy = source.cloneNode(true);

            // delete elements marked with ignore (eg anchor and jump buttons)
            const deleteFromCopy = sourceCopy.querySelectorAll(
                '[data-ignore="true"]'
            );
            for (const element of deleteFromCopy)
                element.remove();

            // delete certain element attributes
            const attributes = [
                'id',
                'data-collapsed',
                'data-selected',
                'data-highlighted',
                'data-glow'
            ];
            for (const attribute of attributes) {
                sourceCopy.removeAttribute(attribute);
                const elements = sourceCopy.querySelectorAll(
                    '[' + attribute + ']'
                );
                for (const element of elements)
                    element.removeAttribute(attribute);
            }

            return sourceCopy;
        }

        // make button to jump to previous image in document
        function makePrevButton(img) {
            const prevButton = document.createElement('button');
            prevButton.id = 'lightbox_prev_button';
            prevButton.title = 'Jump to the previous image in the document [←]';
            prevButton.classList.add('icon_button', 'lightbox_button');
            prevButton.innerHTML = document.querySelector(
                '.icon_caret_left'
            ).innerHTML;

            // attach click listeners to button
            prevButton.addEventListener('click', function() {
                getPrevImg(img).click();
            });

            return prevButton;
        }

        // make button to jump to next image in document
        function makeNextButton(img) {
            const nextButton = document.createElement('button');
            nextButton.id = 'lightbox_next_button';
            nextButton.title = 'Jump to the next image in the document [→]';
            nextButton.classList.add('icon_button', 'lightbox_button');
            nextButton.innerHTML = document.querySelector(
                '.icon_caret_right'
            ).innerHTML;

            // attach click listeners to button
            nextButton.addEventListener('click', function() {
                getNextImg(img).click();
            });

            return nextButton;
        }

        // get previous image in document
        function getPrevImg(img) {
            const imgs = document.querySelectorAll('.lightbox_document_img');

            // find index of provided img
            let index;
            for (index = 0; index < imgs.length; index++) {
                if (imgs[index] === img)
                    break;
            }


            // wrap index to other side if < 1
            if (index - 1 >= 0)
                index--;
            else
                index = imgs.length - 1;
            return imgs[index];
        }

        // get next image in document
        function getNextImg(img) {
            const imgs = document.querySelectorAll('.lightbox_document_img');

            // find index of provided img
            let index;
            for (index = 0; index < imgs.length; index++) {
                if (imgs[index] === img)
                    break;
            }


            // wrap index to other side if > total
            if (index + 1 <= imgs.length - 1)
                index++;
            else
                index = 0;
            return imgs[index];
        }

        // close lightbox
        function closeLightbox() {
            focusBody();

            const lightbox = document.getElementById('lightbox_overlay');
            if (lightbox)
                lightbox.remove();
        }

        // make all elements behind lightbox non-focusable
        function blurBody(overlay) {
            const all = document.querySelectorAll('*');
            for (const element of all)
                element.tabIndex = -1;
            document.body.classList.add('body_no_scroll');
        }

        // make all elements focusable again
        function focusBody() {
            const all = document.querySelectorAll('*');
            for (const element of all)
                element.removeAttribute('tabIndex');
            document.body.classList.remove('body_no_scroll');
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>

<!-- caret left icon -->

<template class="icon_caret_left">
    <!-- modified from: https://fontawesome.com/icons/caret-left -->
    <svg width="16" height="16" viewBox="0 0 192 512">
        <path
            fill="currentColor"
            d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
        ></path>
    </svg>
</template>

<!-- caret right icon -->

<template class="icon_caret_right">
    <!-- modified from: https://fontawesome.com/icons/caret-right -->
    <svg width="16" height="16" viewBox="0 0 192 512">
        <path
            fill="currentColor"
            d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
        ></path>
    </svg>
</template>
<!-- attributes plugin -->

<script>
    (function() {
        // /////////////////////////
        // DESCRIPTION
        // /////////////////////////

        // This Manubot plugin allows arbitrary HTML attributes to be attached
        // to (almost) any element. Place an HTML comment inside or next to the
        // desired element in the format <!-- $attribute="value" -->

        // /////////////////////////
        // OPTIONS
        // /////////////////////////

        // plugin name prefix for url parameters
        const pluginName = 'attributes';

        // default plugin options
        const options = {
            // whether plugin is on or not
            enabled: 'true'
        };

        // change options above, or override with url parameter, eg:
        // 'manuscript.html?pluginName-enabled=false'

        // /////////////////////////
        // SCRIPT
        // /////////////////////////

        // start script
        function start() {
            // get list of comments in document
            const comments = findComments();

            for(const comment of comments)
                if (comment.parentElement)
                    addAttributes(
                        comment.parentElement,
                        comment.nodeValue.trim()
                    );
        }

        // add html attributes to specified element based on string of 
        // html attributes and values
        function addAttributes(element, text) {
            // regex's for finding attribute/value pairs in the format of
            // attribute="value" or attribute='value
            const regex2 = /\$([a-zA-Z\-]+)?=\"(.+?)\"/;
            const regex1 = /\$([a-zA-Z\-]+)?=\'(.+?)\'/;

            // loop through attribute/value pairs
            let match;
            while(match = text.match(regex2) || text.match(regex1)) {
                // get attribute and value from regex capture groups
                let attribute = match[1];
                let value = match[2];

                // remove from string
                text = text.substring(match.index + match[0].length);

                if (!attribute || !value)
                    break;

                // set attribute of parent element
                try {
                    element.setAttribute(attribute, value);
                } catch(error) {
                    console.log(error);
                }

                // special case for colspan
                if (attribute === 'colspan')
                    removeTableCells(element, value);
            }
        }

        // get list of comment elements in document
        function findComments() {
            const comments = [];

            // iterate over comment nodes in document
            function acceptNode(node) {
                return NodeFilter.FILTER_ACCEPT;
            }
            const iterator = document.createNodeIterator(
                document.body,
                NodeFilter.SHOW_COMMENT,
                acceptNode
            );
            let node;
            while(node = iterator.nextNode())
                comments.push(node);

            return comments;
        }

        // remove certain number of cells after specified cell
        function removeTableCells(cell, number) {
            number = parseInt(number);
            if (!number)
                return;

            // remove elements
            for(; number > 1; number--) {
                if (cell.nextElementSibling)
                    cell.nextElementSibling.remove();
            }
        }

        // load options from url parameters
        function loadOptions() {
            const url = window.location.search;
            const params = new URLSearchParams(url);
            for (const optionName of Object.keys(options)) {
                const paramName = pluginName + '-' + optionName;
                const param = params.get(paramName);
                if (param !== '' && param !== null)
                    options[optionName] = param;
            }
        }
        loadOptions();

        // start script when document is finished loading
        if (options.enabled === 'true')
            window.addEventListener('load', start);
    })();
</script>
<!-- mathjax plugin configuration -->

<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        "CommonHTML": { linebreaks: { automatic: true } },
        "HTML-CSS": { linebreaks: { automatic: true } },
        "SVG": { linebreaks: { automatic: true } },
        "fast-preview": { disabled: true }
  });
</script>

<!-- mathjax plugin -->

<script
    src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-MML-AM_CHTML"
    integrity="sha512-0doc9hKxR3PYwso42RD1p5ySZpzzuDiOwMrdCEh2WdJZCjcmFKc/wEnL+z8fBQrnHoiNWbo+3fiGkOYXBdQp4A=="
    crossorigin="anonymous"
>
    // /////////////////////////
    // DESCRIPTION
    // /////////////////////////

    // This third-party plugin 'MathJax' allows the proper rendering of
    // math/equations written in LaTeX.

    // https://www.mathjax.org/
</script>
<!-- annotations plugin -->

<script>
    // /////////////////////////
    // DESCRIPTION
    // /////////////////////////

    // This third-party plugin 'Hypothesis' allows public annotation of the
    // manuscript.

    // https://web.hypothes.is/

    // plugin configuration
    window.hypothesisConfig = function() {
        return {
            branding: {
                accentColor: '#2196f3',
                appBackgroundColor: '#f8f8f8',
                ctaBackgroundColor: '#f8f8f8',
                ctaTextColor: '#000000',
                selectionFontFamily: 'Open Sans, Helvetica, sans serif',
                annotationFontFamily: 'Open Sans, Helvetica, sans serif'
            }
        };
    };

    // hypothesis client script
    const embed = 'https://hypothes.is/embed.js';
    // hypothesis annotation count query url
    const query = 'https://api.hypothes.is/api/search?limit=0&url='

    
    // start script
    function start() {
        const button = makeButton();
        document.body.insertBefore(button, document.body.firstChild);
        insertCount(button);
    }

    // make button
    function makeButton() {
        // create button
        const button = document.createElement('button');
        button.id = 'hypothesis_button';
        button.innerHTML = document.querySelector('.icon_hypothesis').innerHTML;
        button.title = 'Hypothesis annotations';
        button.classList.add('icon_button');

        function onClick(event) {
            onButtonClick(event, button);
        }

        // attach click listeners
        button.addEventListener('click', onClick);

        return button;
    }

    // insert annotations count
    async function insertCount(button) {
        // get annotation count from Hypothesis based on url
        let count = '-';
        try {
            const canonical = document.querySelector('link[rel="canonical"]');
            const location = window.location;
            const url = encodeURIComponent((canonical || location).href);
            const response = await fetch(query + url);
            const json = await response.json();
            count = json.total || '-';
        } catch(error) {
            console.log(error);
        }
        
        // put count into button
        const counter = document.createElement('span');
        counter.id = 'hypothesis_count';
        counter.innerHTML = count;
        button.title = 'View ' + count + ' Hypothesis annotations';
        button.append(counter);
    }

    // when button is clicked
    function onButtonClick(event, button) {
        const script = document.createElement('script');
        script.src = embed;
        document.body.append(script);
        button.remove();
    }

    window.addEventListener('load', start);
</script>

<!-- hypothesis icon -->

<template class="icon_hypothesis">
    <!-- modified from: https://simpleicons.org/icons/hypothesis.svg / https://git.io/Jf1VB -->
    <svg width="16" height="16" viewBox="0 0 24 24" tabindex="-1">
        <path
            fill="currentColor"
            d="M3.43 0C2.5 0 1.72 .768 1.72 1.72V18.86C1.72 19.8 2.5 20.57 3.43 20.57H9.38L12 24L14.62 20.57H20.57C21.5 20.57 22.29 19.8 22.29 18.86V1.72C22.29 .77 21.5 0 20.57 0H3.43M5.14 3.43H7.72V9.43S8.58 7.72 10.28 7.72C12 7.72 13.74 8.57 13.74 11.24V17.14H11.16V12C11.16 10.61 10.28 10.07 9.43 10.29C8.57 10.5 7.72 11.41 7.72 13.29V17.14H5.14V3.43M18 13.72C18.95 13.72 19.72 14.5 19.72 15.42A1.71 1.71 0 0 1 18 17.13A1.71 1.71 0 0 1 16.29 15.42C16.29 14.5 17.05 13.71 18 13.71Z"
            tabindex="-1"
        ></path>
    </svg>
</template>
<!-- analytics plugin -->

<!-- copy and paste code from Google Analytics or similar service here -->
</body>
</html>
